<!DOCTYPE html>



  


<html class="theme-next mist use-motion" lang="zh-CN">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.4" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.4">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.4" color="#222">





  <meta name="keywords" content="WES,NA12878," />










<meta name="description" content="一直都比较有兴趣分析 human DNA 变异，参照 gatk 流程分析了 NA12878 的全外数据，正好有参考变异集，比较下分析准确性，也好参数调整优化之类，吧吧吧，下面就是一些详细步骤粘贴整理 NA12878.vcf 处理123456789101112131415# wget -c ftp://ftp-trace.ncbi.nih.gov/giab/ftp/data/NA12878/Garv">
<meta name="keywords" content="WES,NA12878">
<meta property="og:type" content="article">
<meta property="og:title" content="NA12878 全外显子数据分析结果比较">
<meta property="og:url" content="http://chengcz.github.io/2018/201806_compare_vcf/index.html">
<meta property="og:site_name" content="ChengCZ&#39;s Notebook">
<meta property="og:description" content="一直都比较有兴趣分析 human DNA 变异，参照 gatk 流程分析了 NA12878 的全外数据，正好有参考变异集，比较下分析准确性，也好参数调整优化之类，吧吧吧，下面就是一些详细步骤粘贴整理 NA12878.vcf 处理123456789101112131415# wget -c ftp://ftp-trace.ncbi.nih.gov/giab/ftp/data/NA12878/Garv">
<meta property="og:locale" content="zh-CN">
<meta property="og:image" content="http://chengcz.github.io/2018/201806_compare_vcf/images/2018/StrandBias_density.png">
<meta property="og:updated_time" content="2018-06-25T16:06:10.707Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="NA12878 全外显子数据分析结果比较">
<meta name="twitter:description" content="一直都比较有兴趣分析 human DNA 变异，参照 gatk 流程分析了 NA12878 的全外数据，正好有参考变异集，比较下分析准确性，也好参数调整优化之类，吧吧吧，下面就是一些详细步骤粘贴整理 NA12878.vcf 处理123456789101112131415# wget -c ftp://ftp-trace.ncbi.nih.gov/giab/ftp/data/NA12878/Garv">
<meta name="twitter:image" content="http://chengcz.github.io/2018/201806_compare_vcf/images/2018/StrandBias_density.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    version: '5.1.4',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: 'Author'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://chengcz.github.io/2018/201806_compare_vcf/"/>





  <title>NA12878 全外显子数据分析结果比较 | ChengCZ's Notebook</title>
  





  <script type="text/javascript">
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?a79f8c7f932a2d3ed822562170533a8d";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>




</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-CN">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">ChengCZ's Notebook</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            Home
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            About
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            Archives
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201806_compare_vcf/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">NA12878 全外显子数据分析结果比较</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              
              <time title="Post created" itemprop="dateCreated datePublished" datetime="2018-06-26T00:02:01+08:00">
                2018-06-26
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/bioinfor/" itemprop="url" rel="index">
                    <span itemprop="name">bioinfor</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <p>一直都比较有兴趣分析 human DNA 变异，参照 gatk 流程分析了 NA12878 的全外数据，正好有参考变异集，比较下分析准确性，也好参数调整优化之类，吧吧吧，下面就是一些详细步骤粘贴整理</p>
<h3 id="NA12878-vcf-处理"><a href="#NA12878-vcf-处理" class="headerlink" title="NA12878.vcf 处理"></a>NA12878.vcf 处理</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> wget -c ftp://ftp-trace.ncbi.nih.gov/giab/ftp/data/NA12878/Garvan_NA12878_HG001_HiSeq_Exome/project.NIST.hc.snps.indels.vcf</span><br><span class="line">hc_snp_indel=project.NIST.hc.snps.indels.vcf</span><br><span class="line"><span class="meta">#</span> wget -c ftp://ftp-trace.ncbi.nih.gov/giab/ftp/data/NA12878/Garvan_NA12878_HG001_HiSeq_Exome/nexterarapidcapture_expandedexome_targetedregions.bed</span><br><span class="line">intervals=nexterarapidcapture_expandedexome_targetedregions.bed</span><br><span class="line"></span><br><span class="line">java -jar GenomeAnalysisTK.jar \</span><br><span class="line">    -T SelectVariants \</span><br><span class="line">    -R $genome \</span><br><span class="line">    -V $hc_snp_indel \</span><br><span class="line">    -L $intervals \</span><br><span class="line">    -sn NIST7035 \</span><br><span class="line">    -select 'DP &gt; 5' \</span><br><span class="line">    --excludeNonVariants \</span><br><span class="line">    -o ref.NIST7035.hc.snps.indels.vcf.gz</span><br><span class="line"><span class="meta">#</span> 选择特定区间内 DP 大于5，样品 NIST7035 的变异位点</span><br></pre></td></tr></table></figure>
<h3 id="vcf-处理"><a href="#vcf-处理" class="headerlink" title="vcf 处理"></a>vcf 处理</h3><p>ref: <a href="https://software.broadinstitute.org/gatk/documentation/article.php?id=2806" target="_blank" rel="noopener">https://software.broadinstitute.org/gatk/documentation/article.php?id=2806</a></p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> SNP</span><br><span class="line">java -jar GenomeAnalysisTK.jar \</span><br><span class="line">    -T SelectVariants \</span><br><span class="line">    -R $genome \</span><br><span class="line">    -L $intervals \</span><br><span class="line">    -selectType SNP \</span><br><span class="line">    -V  NIST7035.vcf.gz \</span><br><span class="line">    -o NIST7035.snp.vcf.gz</span><br><span class="line">java -jar GenomeAnalysisTK.jar \</span><br><span class="line">    -T VariantFiltration \</span><br><span class="line">    -R $genome \</span><br><span class="line">    --filterExpression "DP &lt; 5" --filterName "LowDepth"  \</span><br><span class="line">    --filterExpression "FS &gt; 60.0 || SOR &gt; 3.0" --filterName "StrandBias"  \</span><br><span class="line">    --filterExpression "QD &lt; 2.0  || MQ &lt; 40.0 || MQRankSum &lt; -12.5" \</span><br><span class="line">    --filterName "snp_Filter"  \</span><br><span class="line">    --filterExpression "ReadPosRankSum &lt; -8.0" --filterName "endOfRead"  \</span><br><span class="line">    -V  NIST7035.snp.vcf.gz \</span><br><span class="line">    -o  NIST7035.snp.filter.vcf.gz</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> Indel</span><br><span class="line">java -jar GenomeAnalysisTK.jar \</span><br><span class="line">    -T SelectVariants \</span><br><span class="line">    -R $genome \</span><br><span class="line">    -L $intervals \</span><br><span class="line">    -selectType INDEL \</span><br><span class="line">    -V  NIST7035.vcf.gz \</span><br><span class="line">    -o NIST7035.indel.vcf.gz</span><br><span class="line">java -jar GenomeAnalysisTK.jar \</span><br><span class="line">    -T VariantFiltration \</span><br><span class="line">    -R $genome \</span><br><span class="line">    --filterExpression "DP &lt; 5" --filterName "LowDepth"  \</span><br><span class="line">    --filterExpression "QD &lt; 2.0 || FS &gt; 200.0" \</span><br><span class="line">    --filterName "InDels_Filter" \</span><br><span class="line">    --filterExpression "ReadPosRankSum &lt; -20.0" --filterName "endOfRead" \</span><br><span class="line">    -V NIST7035.indel.vcf.gz \</span><br><span class="line">    -o NIST7035.indel.filter.vcf.gz</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> MergeVcf</span><br><span class="line">java -jar picard.jar MergeVcfs \</span><br><span class="line">    I=NIST7035.indel.filter.vcf.gz \</span><br><span class="line">    I=NIST7035.snp.filter.vcf.gz \</span><br><span class="line">    O=NIST7035.filter.vcf.gz\</span><br><span class="line">    D=$genome.dict</span><br></pre></td></tr></table></figure>
<h3 id="compare-use-bcftools"><a href="#compare-use-bcftools" class="headerlink" title="compare use bcftools"></a>compare use bcftools</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">bcftools isec NIST7035.filter.vcf.gz ref.NIST7035.hc.snps.indels.vcf.gz -p bcftools</span><br><span class="line"><span class="meta">#</span> ref.vcf.gz, alt.vcf.gz 使用 bgzip 压缩，同时 tabix 进行 index</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> output: README.txt</span><br><span class="line"><span class="meta">#</span> This file was produced by vcfisec.</span><br><span class="line"><span class="meta">#</span> The command line was:   bcftools isec  -p bcftools NIST7035.filter.vcf.gz ref.NIST7035.hc.snps.indels.vcf.gz</span><br><span class="line"><span class="meta">#</span></span><br><span class="line"><span class="meta">#</span> Using the following file names:</span><br><span class="line"><span class="meta">#</span> bcftools/0000.vcf  for records private to  NIST7035.filter.vcf.gz</span><br><span class="line"><span class="meta">#</span> bcftools/0001.vcf  for records private to  ref.NIST7035.hc.snps.indels.vcf.gz</span><br><span class="line"><span class="meta">#</span> bcftools/0002.vcf  for records from NIST7035.filter.vcf.gz shared by both  NIST7035.filter.vcf.gz ref.NIST7035.hc.snps.indels.vcf.gz</span><br><span class="line"><span class="meta">#</span> bcftools/0003.vcf  for records from ref.NIST7035.hc.snps.indels.vcf.gz shared by both      NIST7035.filter.vcf.gz ref.NIST7035.hc.snps.indels.vcf.gz</span><br><span class="line"></span><br><span class="line">cat ref.NIST7035.hc.snps.indels.vcf.gz | grep -v '^#' | wc -l    # 54077</span><br><span class="line"></span><br><span class="line">cat 0000.vcf | grep -cv '^#'    # 3863, 假阳性</span><br><span class="line">cat 0001.vcf | grep -cv '^#'    # 2131, 假阴性</span><br><span class="line">cat 0002.vcf | grep -cv '^#'    # 51946</span><br><span class="line">cat 0003.vcf | grep -cv '^#'    # 51946</span><br><span class="line"></span><br><span class="line">cat 0000.vcf| grep -v '^#' | grep PASS | wc -l    # 1334</span><br><span class="line">cat 0002.vcf| grep -v '^#' | grep PASS | wc -l    # 47143</span><br></pre></td></tr></table></figure>
<p>不考虑snp、indels过滤情况下，计算假阳性、假阴性率。</p>
<blockquote>
<p>假阳性: 3863/54077 = 7.14 %</p>
<p>假阴性: 2131/54077 = 3.94 %</p>
</blockquote>
<p>哈哈哈哈，瞎了，没法看的样子。还是考虑下变异过滤结果吧</p>
<blockquote>
<p>假阳性: 1334/54077 = 2.46 %</p>
<p>假阴性: (2131+51946-47143)/54077 = 12.82 %</p>
</blockquote>
<p>更加让人没话说，这个假阴性太感人，过滤标准不太合适啊</p>
  <figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> 先看被过滤掉的这部分阴性</span><br><span class="line">cat 0002.vcf | grep -v '^#' | grep -v PASS | awk '&#123;n[$7]+=1&#125;END&#123;for (i in n)&#123;print i,n[i]&#125;&#125;'</span><br><span class="line"><span class="meta">#</span> LowDepth 11</span><br><span class="line"><span class="meta">#</span> StrandBias;snp_Filter 277</span><br><span class="line"><span class="meta">#</span> LowDepth;snp_Filter 2</span><br><span class="line"><span class="meta">#</span> InDels_Filter 265</span><br><span class="line"><span class="meta">#</span> StrandBias 3098</span><br><span class="line"><span class="meta">#</span> LowDepth;StrandBias 2</span><br><span class="line"><span class="meta">#</span> snp_Filter 1148</span><br></pre></td></tr></table></figure>
<p>过滤掉位点最严重的两项分别是<strong>StrandBias</strong>, <strong>snp_Filter</strong>，对应过滤指标分别为<code>FS &gt; 60.0 || SOR &gt; 3.0</code>, <code>QD &lt; 2.0  || MQ &lt; 40.0 || MQRankSum &lt; -12.5</code>，转换vcf文件到tab格式（脚本：<a href="https://github.com/chengcz/BioResource/tree/master/vcf2tab" target="_blank" rel="noopener">github</a>），对其中一些指标进行分布检查。</p>
<p>先看 <strong>StrandBias</strong>，对 <code>FS</code>,  <code>SOR</code> 两个参数分析，分布结果如下，可看出被过滤的位点基本上都是由于 <strong>SOR</strong> 参数（作图时FS&gt;60无显示），与 <strong>NA12878</strong> 变异集相比，需调整 <strong>SOR</strong> 临界值到更大水平。</p>
<p><img src="./images/2018/StrandBias_density.png" alt=""></p>
<p>再看 <strong>snp_Filter</strong> 对应的<code>QD &lt; 2.0  || MQ &lt; 40.0 || MQRankSum &lt; -12.5</code>，其中 MQRankSum 对应67个位点，QD 对应557个位点，MQ 对应 1168 个位点，看出 <strong>MQ</strong>, <strong>QD</strong> 标准需要进行调整。</p>
<p>以现在质控指标对文件<code>0001.vcf</code> <strong>2131</strong> 变异质控，<strong>QD</strong> 过滤掉949，<strong>MQRankSum</strong> 过滤掉14，<strong>MQ</strong> 过滤掉195，<strong>FS</strong> 过滤掉135，根据这部分参考变异集质量指标，同样 <strong>QD</strong>, <strong>MQ</strong>, <strong>FS</strong> 参数需要进行调整。</p>
<p>另，vcftools 也可以很方便进行两个 vcf 文件的比较，还没 bcftools 对 vcf 文件必须排序、index的要求，输出结果也相应不同，因为我需要看阳性、阴性位点是否被过滤以及过滤情况，就不再用 vcftools 整理了，具体命令可参考如下：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> http://vcftools.github.io/</span><br><span class="line">vcftools --vcf ref.vcf --diff alt.vcf --out prefix</span><br><span class="line"><span class="meta">#</span> ref.vcf.gz    # 则需要使用--gzvcf</span><br><span class="line"><span class="meta">#</span> alt.vcf.gz    # 则需要使用--gzdiff</span><br></pre></td></tr></table></figure>
<h3 id="质控参数调整思路"><a href="#质控参数调整思路" class="headerlink" title="质控参数调整思路"></a>质控参数调整思路</h3><p>根据质控结果比较来看，参控参数都还需要进行再调整，一个初步的参数调整思路是，对高质量变异集各项质控指标分布进行统计，然后进行阈值选择；或者对 <strong>NA12878</strong> 全基因组水平的变异进行 <strong><a href="https://software.broadinstitute.org/gatk/documentation/article.php?id=2805" target="_blank" rel="noopener">VQSR</a></strong> 分析，统计高质量变异位点质控指标分布</p>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/WES/" rel="tag"># WES</a>
          
            <a href="/tags/NA12878/" rel="tag"># NA12878</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2018/201806_ANNOVAR/" rel="next" title="vcf 注释 —— ANNOVAR">
                <i class="fa fa-chevron-left"></i> vcf 注释 —— ANNOVAR
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2018/201806_python_decode/" rel="prev" title="使用 python 正确读取不同编码方式文本">
                使用 python 正确读取不同编码方式文本 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">ChengCZ</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">35</span>
                  <span class="site-state-item-name">posts</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                
                  <span class="site-state-item-count">4</span>
                  <span class="site-state-item-name">categories</span>
                
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                
                  <span class="site-state-item-count">16</span>
                  <span class="site-state-item-name">tags</span>
                
              </div>
            

          </nav>

          

          

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-3"><a class="nav-link" href="#NA12878-vcf-处理"><span class="nav-number">1.</span> <span class="nav-text">NA12878.vcf 处理</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#vcf-处理"><span class="nav-number">2.</span> <span class="nav-text">vcf 处理</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#compare-use-bcftools"><span class="nav-number">3.</span> <span class="nav-text">compare use bcftools</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#质控参数调整思路"><span class="nav-number">4.</span> <span class="nav-text">质控参数调整思路</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2018</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">ChengCZ</span>

  
</div>


  <div class="powered-by">Powered by <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a></div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">Theme &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Mist</a> v5.1.4</div>




        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  
    <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>
  

  
  
    <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>
  


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.4"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.4"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.4"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.4"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.4"></script>



  


  




	





  





  












  





  

  

  

  
  

  

  

  

</body>
</html>
