<!DOCTYPE html>












  


<html class="theme-next mist use-motion" lang="en">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2"/>
<meta name="theme-color" content="#222">






  
  
    
    
  <script src="/lib/pace/pace.min.js?v=1.0.2"></script>
  <link href="/lib/pace/pace-theme-minimal.min.css?v=1.0.2" rel="stylesheet">







<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />






















<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=6.3.0" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=6.3.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=6.3.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=6.3.0">


  <link rel="mask-icon" href="/images/logo.svg?v=6.3.0" color="#222">









<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    version: '6.3.0',
    sidebar: {"position":"left","display":"remove","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta property="og:type" content="website">
<meta property="og:title" content="ChengCZ&#39;s Notebook">
<meta property="og:url" content="http://chengcz.github.io/index.html">
<meta property="og:site_name" content="ChengCZ&#39;s Notebook">
<meta property="og:locale" content="en">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="ChengCZ&#39;s Notebook">



  <link rel="alternate" href="/atom.xml" title="ChengCZ's Notebook" type="application/atom+xml" />




  <link rel="canonical" href="http://chengcz.github.io/"/>



<script type="text/javascript" id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>ChengCZ's Notebook</title>
  






  <script type="text/javascript">
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?a79f8c7f932a2d3ed822562170533a8d";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>




  <noscript>
  <style type="text/css">
    .use-motion .motion-element,
    .use-motion .brand,
    .use-motion .menu-item,
    .sidebar-inner,
    .use-motion .post-block,
    .use-motion .pagination,
    .use-motion .comments,
    .use-motion .post-header,
    .use-motion .post-body,
    .use-motion .collection-title { opacity: initial; }

    .use-motion .logo,
    .use-motion .site-title,
    .use-motion .site-subtitle {
      opacity: initial;
      top: initial;
    }

    .use-motion {
      .logo-line-before i { left: initial; }
      .logo-line-after i { right: initial; }
    }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="en">

  
  
    
  

  <div class="container sidebar-position-left 
  page-home">
    <div class="headband"></div>

    <a href="https://github.com/chengcz/" class="github-corner" aria-label="View source on Github"><svg width="80" height="80" viewBox="0 0 250 250" style="fill:#151513; color:#fff; position: absolute; top: 0; border: 0; left: 0; transform: scale(-1, 1);" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg><style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style></a>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">ChengCZ's Notebook</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="Toggle navigation bar">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home menu-item-active">
    <a href="/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-home"></i> <br />Home</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">
    <a href="/tags" rel="section">
      <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />Tags</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">
    <a href="/categories" rel="section">
      <i class="menu-item-icon fa fa-fw fa-th"></i> <br />Categories</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">
    <a href="/archives" rel="section">
      <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />Archives</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-about">
    <a href="/about" rel="section">
      <i class="menu-item-icon fa fa-fw fa-user"></i> <br />About</a>
  </li>

      
      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br />Search</a>
        </li>
      
    </ul>
  

  
    

  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="Searching..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



  



</div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201809_pysam/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201809_pysam/" itemprop="url">
                  python 高效处理 bam 等生物格式 —— pysam
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-09-15 17:05:10 / Modified: 17:15:20" itemprop="dateCreated datePublished" datetime="2018-09-15T17:05:10+08:00">2018-09-15</time>
            

            
              

              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/python/" itemprop="url" rel="index"><span itemprop="name">python</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>Pysam 打包 htslib C 代码， 处理 bam 等生物格式文件的 python 工具包，文档参考：<a href="https://pysam.readthedocs.io/en/latest/api.html" target="_blank" rel="noopener">https://pysam.readthedocs.io/en/latest/api.html</a></p>
<h3 id="install"><a href="#install" class="headerlink" title="install"></a>install</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">pip search pysam</span><br><span class="line"><span class="meta">#</span> conda search pysam</span><br><span class="line">pip install pysam</span><br><span class="line"><span class="meta">#</span> conda install pysam</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> pysam 打包了 htslib 提供了一个 python 接口用来处理 bam, vcf 等常用生物格式</span><br></pre></td></tr></table></figure>
<h3 id="解析sam"><a href="#解析sam" class="headerlink" title="解析sam"></a>解析sam</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> pysam</span><br><span class="line">fp = pysam.AlignmentFile(<span class="string">'bam'</span>, <span class="string">'rb)</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">fout = pysam.AlignmentFile("allpaired.bam", "wb", template=fp)</span></span><br><span class="line"><span class="string">for read in fp.fetch():</span></span><br><span class="line"><span class="string">    if read.is_paired:</span></span><br><span class="line"><span class="string">        fout.write(read)</span></span><br><span class="line"><span class="string">    # dir(read)</span></span><br><span class="line"><span class="string">    # 查看 read 对象的全部方法或属性</span></span><br><span class="line"><span class="string">                         </span></span><br><span class="line"><span class="string"># 特定位置 reads 读取，bam文件需要index</span></span><br><span class="line"><span class="string">for pileupcolumn in fp.fetch('</span>chr1<span class="string">', 100, 120):</span></span><br><span class="line"><span class="string">    for read in pileupcolumn.pileups:</span></span><br><span class="line"><span class="string">        print(read)</span></span><br><span class="line"><span class="string">        # ...</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">fout.close()</span></span><br><span class="line"><span class="string">fp.close()</span></span><br></pre></td></tr></table></figure>
<h3 id="解析fasta"><a href="#解析fasta" class="headerlink" title="解析fasta"></a>解析fasta</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">fp = pysam.FastaFile('fasta')</span><br><span class="line"><span class="meta">#</span> 提取序列, 0-based，fasta文件需要index</span><br><span class="line">seq = fp.fetch(reference=chro, start=start, end=end)</span><br><span class="line">fp.close()</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> 也可以使用with ... as ... 上下文管理</span><br><span class="line">with pysam.FastaFile('fasta') as fp:</span><br><span class="line">    fp.fetch(reference=chro, start=start, end=end)</span><br></pre></td></tr></table></figure>
<h3 id="解析VCF"><a href="#解析VCF" class="headerlink" title="解析VCF"></a>解析VCF</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">fp = pysam.VariantFile('vcf')</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> 获得vcf包含的所有sample lst</span><br><span class="line">samples = list((fp.header.samples))</span><br><span class="line"></span><br><span class="line">for rec in bcf_in.fetch():</span><br><span class="line"><span class="meta">#</span> fp.fetch('chr1', 100000, 200000) 获取特定位置区间内的vairant，index后的vcf文件可用</span><br><span class="line">    print(rec)</span><br><span class="line"><span class="meta">	#</span> dir(rec) 查看 rec 的所有方法属性</span><br><span class="line">fp.colse()</span><br></pre></td></tr></table></figure>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201809_tRFs/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201809_tRFs/" itemprop="url">
                  identification of tRFs from sRNA deep-sequencing data
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-09-02 00:43:24 / Modified: 00:43:53" itemprop="dateCreated datePublished" datetime="2018-09-02T00:43:24+08:00">2018-09-02</time>
            

            
              

              
            
          </span>

          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h3 id="methods"><a href="#methods" class="headerlink" title="methods"></a>methods</h3><p>参考 <a href="http://genome.bioch.virginia.edu/trfdb/method.php" target="_blank" rel="noopener">http://genome.bioch.virginia.edu/trfdb/method.php</a>，详细方法描述如下：</p>
<p>tRNA 基因信息从 <a href="http://gtrnadb.ucsc.edu/" target="_blank" rel="noopener">Genomic tRNA database</a> 或者 NCBI 下载，根据注释基因组版本，tRNA 基因链关系，在成熟 rRNA 基因上下游分别延申 100 bp 提取序列</p>
<blockquote>
<p>Information about the tRNA genes for species R. sphaeroides (Bacteria; ATCC_17025), S. pombe (schiPomb1), Droshophila (dm3), C. elegans (ce6), Xenopus (xenTro3), zebra fish (Zv9), mouse (mm9) and human (hg19) was either downloaded from the “<strong>Genomic tRNA database</strong>“ (<a href="http://gtrnadb.ucsc.edu/" target="_blank" rel="noopener">http://gtrnadb.ucsc.edu/</a>) or “ <strong>NCBI</strong>“ (<a href="http://www.ncbi.nlm.nih.gov/)" target="_blank" rel="noopener">http://www.ncbi.nlm.nih.gov/)</a>. We extracted sequences that spanned 100bp upstream and 100bp downstream to the mature tRNA genes from the same genome assembly on which the tRNA gene coordinates were built. The genomic sequences were extracted based on the strand information of tRNA gene transcription. </p>
</blockquote>
<p>物种特异的 blastdb 被构建，每一个 sRNA 通过 blastn 比对到对应的物种数据库。通常只有 100% 比对到上 db 数据的 sRNA read 进入后续分析步骤。</p>
<blockquote>
<p>A species-specific tRNAdb  blast database was built to query the small RNA sequences. To find the tRNA-related RNA sequences in each library, the small RNAs were mapped to the species-specific tRNAdb, using BLASTn (Altschul, Madden et al. 1997). In general we considered only those alignments where the query sequence (small RNA) was mapped to the database sequence (tRNA) along 100% of its length. </p>
</blockquote>
<p>blast 输出结果通过 in-house 脚本得到 sRNA 比对的 tRNA 基因上的位置。Blastdb 数据库考虑了基因链信息，因此只有 blast 比对到相同链的 read 被考虑</p>
<blockquote>
<p>Blast output file was parsed using an in-house developed script to get the information of mapped positions of small RNA on tRNA genes. Since custom tRNA database was built considering the strand information of tRNA gene therefore only those blast alignments were considered where queries mapped on to the positive strand of the subject sequence. </p>
</blockquote>
<p>根据比对从 tRNA 第一个剪辑到最后一个碱基提取 sRNA 信息，最多允许1个错配。在 tRNA 成熟过程中 tRNA 核苷酸转移酶在 tRNA 3’末端添加了“CCA”，因此 tRNA 3‘ 末端的 sRNA 特异的允许不超过3个碱基的错配</p>
<blockquote>
<p>We extracted the information for small RNAs aligned from its first to the last base with tRNA sequences, allowing either one or no mismatch. Since “CCA” is added at the 3’ end of tRNA by tRNA nucleotidyltransferase during maturation of tRNA (Xiong and Steitz 2006) therefore a special exception for the small RNA mapping to the 3’ ends of tRNAs in the tRNAdb was devised allowing a terminal mismatch of &lt;=3 bases. </p>
</blockquote>
<p>为了消除假阳性，比对到 tRNA blastdb 的 sRNA read，再次使用 blast 进行全基因组搜索，排除不在 tRNA 位点的 read。最后，只有那些只定位于 tRNA 位点的 read 被认定为可能的 tRFs</p>
<blockquote>
<p>To eliminate any false positives, the small RNAs that mapped on to the “tRNAdb” were again searched against the whole genome database using blast search excluding the tRNA loci. Finally only those small RNAs were qualified as probable tRFs that mapped exclusively on tRNA loci.</p>
</blockquote>
<p> 映射在tRNA基因上的sRNA的末端用于获得tRNA上任何映射的sRNA的显着富集</p>
<blockquote>
<p>The ends of the small RNA mapped on tRNA genes was used to access the significant enrichment of any mapped small RNA on tRNA. </p>
</blockquote>
<p>如果 tRF 是 tRNA 转录物随机降解的产物，则tRF的末端将沿着tRNA基因的长度以相当的频率均等分布。然而，大多数 sRNA（在单个 tRNA 的总比对读数的90％以上）映射在三个特定区域：成熟 tRNA 的5’末端（tRF-5），3’末端（tRF-3）和 3’ tRNA基因前体的区域（tRF-1）。因此，仅考虑映射到这些特定位置的tRF用于进一步分析。 </p>
<blockquote>
<p>If the tRFs were a result of the random degradation of tRNA transcript, the ends of the tRFs would be equally distributed along the lengths of the tRNA genes with comparable frequency. However, the small RNA mostly (&gt;90% of total mapped reads on individual tRNA) mapped on three specific regions: extreme 5’ end (tRF-5), extreme 3’ end (tRF-3) of mature tRNA and 3’ trailer region (tRF-1) of primary tRNA genes. Therefore tRFs mapped only to these specific locations were considered for further analysis. </p>
</blockquote>
<p>如下图所示，对于每个 tRF 都存在一种丰富特别高的 sRNA read（例如tRF-5“GCATTGGTGGTTCAGTGGTAGA”以8258读数/百万份进行测序），占到映射到该位点的读数的80％以上。这将主 tRF 与 其他低丰度的核酸酶降解产物区分开，或者可能来自 tRNA 和 tRF 的随机降解。最后，数据库中仅包含高度丰富的tRF（至少一个库中每百万&gt; 20个读数）。</p>
<blockquote>
<p>As shown in Figure for each tRF there is one most abundant RNA sequenced (example the tRF-5 “GCATTGGTGGTTCAGTGGTAGA” was sequenced at 8258 reads per million) accounting for more than 80% of the reads mapping to that site. This distinguishes the main tRF from other low abundance products created by nucleases digesting the main tRF, or possibly from random degradation of tRNAs and tRFs. Finally, only highly abundant tRFs (&gt;20 reads per million in at least one library) were included in the database. </p>
</blockquote>
<p><img src="http://genome.bioch.virginia.edu/trfdb/Fig2.jpg" alt=""></p>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201809_ShellNote2/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201809_ShellNote2/" itemprop="url">
                  shell 命令使用笔记2
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-09-02 00:38:30 / Modified: 01:01:02" itemprop="dateCreated datePublished" datetime="2018-09-02T00:38:30+08:00">2018-09-02</time>
            

            
              

              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/linux/" itemprop="url" rel="index"><span itemprop="name">linux</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <ol>
<li><p>一个 shell 脚本模版</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span>!/usr/bin/env bash    # 指定 shell 执行程序</span><br><span class="line"><span class="meta">#</span>!/bin/bash    # option B</span><br><span class="line"></span><br><span class="line">set -x    # 打印出执行的命令，</span><br><span class="line">set -e    # cmd 执行出错时，退出 shell</span><br></pre></td></tr></table></figure>
</li>
<li><p>多列对文件进行排序</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sort -k1,1 -k2,2n sort_demo.txt    # 一定是 -k1,1 形式</span><br></pre></td></tr></table></figure>
</li>
<li><p>解包</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">head -n 5 hg19-tRNAs_map.txt</span><br><span class="line"><span class="meta">#</span> chr1.trna1	tRNA-Val-CAC-13-1</span><br><span class="line"><span class="meta">#</span> chr1.trna10	tRNA-Asn-GTT-16-1</span><br><span class="line"><span class="meta">#</span> chr1.trna100	tRNA-Asn-GTT-6-1</span><br><span class="line"><span class="meta">#</span> chr1.trna101	tRNA-Asn-GTT-22-1</span><br><span class="line"><span class="meta">#</span> chr1.trna102	tRNA-Glu-TTC-7-1</span><br><span class="line"></span><br><span class="line">cat hg19-tRNAs_name_map.txt | while read pos name; do echo "Position: $pos    Name: $name"; done | head -n 5</span><br><span class="line"><span class="meta">#</span> Position: chr1.trna1    Name: tRNA-Val-CAC-13-1</span><br><span class="line"><span class="meta">#</span> Position: chr1.trna10    Name: tRNA-Asn-GTT-16-1</span><br><span class="line"><span class="meta">#</span> Position: chr1.trna100    Name: tRNA-Asn-GTT-6-1</span><br><span class="line"><span class="meta">#</span> Position: chr1.trna101    Name: tRNA-Asn-GTT-22-1</span><br><span class="line"><span class="meta">#</span> Position: chr1.trna102    Name: tRNA-Glu-TTC-7-1</span><br></pre></td></tr></table></figure>
</li>
</ol>
<p>4. </p>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201809_parseXML/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201809_parseXML/" itemprop="url">
                  parse blast XML use python
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-09-02 00:05:10 / Modified: 01:10:21" itemprop="dateCreated datePublished" datetime="2018-09-02T00:05:10+08:00">2018-09-02</time>
            

            
              

              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/bioformat/" itemprop="url" rel="index"><span itemprop="name">bioformat</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>2017.05</p>
<hr>
<p>blast 输出 <strong>xml</strong> 格式结果，包括最完整的比对信息，经常需要自己从  blast.xml 结果中提取想要的信息，最开始都自己循环 xml 一行行匹配，一点都不 pythonic，陆续发现了一些新的 xml 解析实现，整理如下：</p>
<h3 id="re-search"><a href="#re-search" class="headerlink" title="re.search"></a>re.search</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">parse_BLAST_XML</span><span class="params">(infile)</span>:</span></span><br><span class="line">    get = <span class="keyword">lambda</span> x: re.search(<span class="string">'&gt;.*?&lt;'</span>, x).group(<span class="number">0</span>)[<span class="number">1</span>:<span class="number">-1</span>]</span><br><span class="line"></span><br><span class="line">    r_query_id = re.compile(<span class="string">"&lt;Iteration_query-ID&gt;"</span>)</span><br><span class="line">    r_query_def = re.compile(<span class="string">"&lt;Iteration_query-def&gt;"</span>)</span><br><span class="line">    r_query_len = re.compile(<span class="string">"&lt;Iteration_query-len&gt;"</span>)</span><br><span class="line">    r_hit_num = re.compile(<span class="string">"&lt;Hit_num&gt;"</span>)</span><br><span class="line">    r_hit_id = re.compile(<span class="string">"&lt;Hit_id&gt;"</span>)</span><br><span class="line">    r_hit_def = re.compile(<span class="string">"&lt;Hit_def&gt;"</span>)</span><br><span class="line">    r_hit_accession = re.compile(<span class="string">"&lt;Hit_accession&gt;"</span>)</span><br><span class="line">    r_hit_len = re.compile(<span class="string">"&lt;Hit_len&gt;"</span>)</span><br><span class="line">    r_hsp_num = re.compile(<span class="string">"&lt;Hsp_num&gt;"</span>)</span><br><span class="line">    r_hsp_bit_score = re.compile(<span class="string">"&lt;Hsp_bit-score&gt;"</span>)</span><br><span class="line">    r_hsp_score = re.compile(<span class="string">"&lt;Hsp_score&gt;"</span>)</span><br><span class="line">    r_hsp_evalue = re.compile(<span class="string">"&lt;Hsp_evalue&gt;"</span>)</span><br><span class="line">    r_hsp_query_from = re.compile(<span class="string">"&lt;Hsp_query-from&gt;"</span>)</span><br><span class="line">    r_hsp_query_to = re.compile(<span class="string">"&lt;Hsp_query-to&gt;"</span>)</span><br><span class="line">    r_hsp_hit_from = re.compile(<span class="string">"&lt;Hsp_hit-from&gt;"</span>)</span><br><span class="line">    r_hsp_hit_to = re.compile(<span class="string">"&lt;Hsp_hit-to&gt;"</span>)</span><br><span class="line">    r_hsp_query_frame = re.compile(<span class="string">"&lt;Hsp_query-frame&gt;"</span>)</span><br><span class="line">    r_hsp_hit_frame = re.compile(<span class="string">"&lt;Hsp_hit-frame&gt;"</span>)</span><br><span class="line">    r_hsp_identity = re.compile(<span class="string">"&lt;Hsp_identity&gt;"</span>)</span><br><span class="line">    r_hsp_positive = re.compile(<span class="string">"&lt;Hsp_positive&gt;"</span>)</span><br><span class="line">    r_hsp_gaps = re.compile(<span class="string">"&lt;Hsp_gaps&gt;"</span>)</span><br><span class="line">    r_hsp_align_len = re.compile(<span class="string">"&lt;Hsp_align-len&gt;"</span>)</span><br><span class="line">    r_hsp_qseq = re.compile(<span class="string">"&lt;Hsp_qseq&gt;"</span>)</span><br><span class="line">    r_hsp_hseq = re.compile(<span class="string">"&lt;Hsp_hseq&gt;"</span>)</span><br><span class="line">    r_hsp_midline = re.compile(<span class="string">"&lt;Hsp_midline&gt;"</span>)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> line <span class="keyword">in</span> open(infile,<span class="string">'r'</span>):</span><br><span class="line">        <span class="keyword">if</span> r_query_id.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            query_id =  get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_query_def.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            query_def = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_query_len.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            query_len = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hit_num.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hit_num = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hit_id.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hit_id = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hit_def.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hit_def = get(line)</span><br><span class="line">   </span><br><span class="line">        <span class="keyword">elif</span> r_hit_accession.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hit_accession = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hit_len.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hit_len = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_num.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_num = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_bit_score.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_bit_score = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_score.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_score = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_evalue.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_evalue = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_query_from.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_query_from = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_query_to.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_query_to = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_hit_from.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_hit_from = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_hit_to.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_hit_to = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_query_frame.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_query_frame = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_hit_frame.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_hit_frame = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_identity.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_identity = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_positive.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_positive = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_gaps.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_gaps = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_align_len.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_align_len = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_qseq.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_qseq = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_hseq.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_hseq = get(line)</span><br><span class="line">        <span class="keyword">elif</span> r_hsp_midline.search(line) <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            hsp_midline = get(line)</span><br><span class="line"></span><br><span class="line">            percent_ident = <span class="string">"%5.2f"</span> %  (((float(hsp_identity)/float(hsp_align_len))*<span class="number">100</span>))</span><br><span class="line">            mismatch = str(int(hsp_align_len) - int(hsp_identity))</span><br><span class="line">            <span class="comment"># ['query id', 'query name', 'subject id', 'identity(%)', 'alignment length', 'mismatches', 'gap opens', 'q. start', 'q. end', 's. start', 's. end', 'evalue', 'bit score', 'subject description']</span></span><br><span class="line">            <span class="keyword">yield</span> [query_id, query_def, hit_id, percent_ident, hsp_align_len, mismatch, hsp_gaps, hsp_query_from, hsp_query_to, hsp_hit_from, hsp_hit_to, hsp_evalue, hsp_bit_score, hit_def]</span><br></pre></td></tr></table></figure>
<h3 id="Bio-Blast-NCBIXML-parse"><a href="#Bio-Blast-NCBIXML-parse" class="headerlink" title="Bio.Blast.NCBIXML.parse"></a>Bio.Blast.NCBIXML.parse</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> __future__ <span class="keyword">import</span> division</span><br><span class="line"><span class="keyword">from</span> Bio.Blast <span class="keyword">import</span> NCBIXML</span><br><span class="line"><span class="comment"># biopython</span></span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">parse_blast_xml</span><span class="params">(result, threshold=<span class="number">1e-5</span>)</span>:</span></span><br><span class="line">	<span class="string">'''</span></span><br><span class="line"><span class="string">		result: opened file handle of xml format</span></span><br><span class="line"><span class="string">	'''</span></span><br><span class="line">    <span class="keyword">for</span> blastrecords <span class="keyword">in</span> NCBIXML.parse(result):</span><br><span class="line">        query_id, _, query_descri = blastrecords.query.partition(<span class="string">' '</span>)</span><br><span class="line">        query_len = blastrecords.query_length</span><br><span class="line">        <span class="keyword">for</span> align <span class="keyword">in</span> blastrecords.alignments:</span><br><span class="line">            sbjct_title = align.title</span><br><span class="line">            sbjct_accession = align.accession</span><br><span class="line">            sbjct_descri = align.hit_def</span><br><span class="line">            sbjct_len = align.length</span><br><span class="line">            <span class="keyword">for</span> hsp <span class="keyword">in</span> align.hsps:</span><br><span class="line">                hsp_evalue = hsp.expect</span><br><span class="line">                <span class="keyword">if</span> hsp_evalue &gt; threshold:</span><br><span class="line">                    <span class="keyword">continue</span></span><br><span class="line">                hsp_score = hsp.score</span><br><span class="line">                hsp_gap = hsp.gaps</span><br><span class="line">                hsp_len = hsp.align_length</span><br><span class="line">                hsp_match = hsp.identities</span><br><span class="line">                hsp_frame = hsp.frame    <span class="comment"># nucl blast to protein database</span></span><br><span class="line">                hsp_strand = hsp.strand    <span class="comment"># nucl2prot</span></span><br><span class="line">                query_seq = hsp.query</span><br><span class="line">                query_start = hsp.query_start</span><br><span class="line">                query_end = hsp.query_end</span><br><span class="line">                sbjct_seq = hsp.sbjct</span><br><span class="line">                sbjct_start = hsp.sbjct_start</span><br><span class="line">                sbjct_end = hsp.sbjct_end</span><br><span class="line">                tmp = [</span><br><span class="line">                    query_id, </span><br><span class="line">                    query_descri, </span><br><span class="line">                    query_seq, </span><br><span class="line">                    query_len, </span><br><span class="line">                    <span class="comment"># hsp_frame, </span></span><br><span class="line">                    <span class="comment"># hsp_strand,</span></span><br><span class="line">                    sbjct_accession,    </span><br><span class="line">                    <span class="comment"># sbjct_title,</span></span><br><span class="line">                    sbjct_descri,</span><br><span class="line">                    sbjct_seq,</span><br><span class="line">                    sbjct_len,</span><br><span class="line">                    hsp_match / hsp_len,</span><br><span class="line">                    hsp_len,</span><br><span class="line">                    hsp_len - hsp_match,</span><br><span class="line">                    hsp_gap,</span><br><span class="line">                    query_start, </span><br><span class="line">                    query_end, </span><br><span class="line">                    sbjct_start, </span><br><span class="line">                    sbjct_end,</span><br><span class="line">                    hsp_evalue,</span><br><span class="line">                    hsp_score,</span><br><span class="line">                ]</span><br><span class="line">                <span class="comment"># qaccver saccver pident length mismatch gapopen qstart qend sstart send evalue bitscore</span></span><br><span class="line">                <span class="keyword">yield</span> tmp</span><br></pre></td></tr></table></figure>
<h3 id="other-python-modules"><a href="#other-python-modules" class="headerlink" title="other python modules"></a>other python modules</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># http://pythonguidecn.readthedocs.io/zh/latest/scenarios/xml.html</span></span><br><span class="line"><span class="comment"># http://www.runoob.com/python/python-xml.html</span></span><br><span class="line"><span class="comment"># 懒得再去看各种文档，挖坑...</span></span><br></pre></td></tr></table></figure>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201808-MALBAC-md/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201808-MALBAC-md/" itemprop="url">
                  Whole Genome Amplification Technologies:MALBAC
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-08-04 22:43:24" itemprop="dateCreated datePublished" datetime="2018-08-04T22:43:24+08:00">2018-08-04</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2018-08-05 14:39:31" itemprop="dateModified" datetime="2018-08-05T14:39:31+08:00">2018-08-05</time>
              
            
          </span>

          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h3 id="Technical-principles"><a href="#Technical-principles" class="headerlink" title="Technical principles"></a>Technical principles</h3><p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/MALBAC.png" alt=""></p>
<ol>
<li><p>A single cell is picked and lysed. First, genomic DNA of the single cell is melted into single-stranded DNA molecules at 94°C. </p>
<blockquote>
<p>Picograms of DNA fragments (~10 to 100 kb)     </p>
</blockquote>
</li>
<li><p>MALBAC primers then anneal randomly to single-stranded DNA molecules at 0°C and are extended by a polymerase with displacement activity at elevated temperatures, creating semiamplicons.</p>
<blockquote>
<p>random primers, each having a common <strong>27-nucleotide sequence</strong> and <strong>8 variable nucleotides</strong>     </p>
</blockquote>
</li>
<li><p>In the following five temperature cycles, after the step of looping the full amplicons, singlestranded amplicons and the genomic DNA are used as a template to produce full amplicons and additional semiamplicons, respectively. </p>
</li>
<li><p>For full amplicons, the 3′ end is complementary to the sequence on the 5′ end. The two ends hybridize to form looped DNA, which can efficiently prevent the full amplicon from being used as a template, therefore warranting a close-to-linear amplification. </p>
</li>
<li><p>After the five cycles of linear preamplification, only the full amplicons can be exponentially amplified in the following PCR <strong>using the common 27- nucleotide sequence as the primer</strong>. PCR reaction will generate microgram level of DNA material for sequencing experiments.    </p>
</li>
</ol>
<h3 id="Advantages"><a href="#Advantages" class="headerlink" title="Advantages"></a>Advantages</h3><ul>
<li>quasi-linear amplification    </li>
<li>accuracy in CNV detection, especially after signal normalization with a reference cell    </li>
</ul>
<p>###Disadvantages </p>
<ul>
<li>MALBAC is not free from sequence-dependent bias    </li>
<li>MALBAC is that it has a high false positive rate for SNV detection    </li>
<li>underamplified regions of the genome are sometimes lost during amplification    </li>
</ul>
<h3 id="ref"><a href="#ref" class="headerlink" title="ref"></a>ref</h3><ul>
<li><a href="https://mp.weixin.qq.com/s?src=11&amp;timestamp=1533173344&amp;ver=1035&amp;signature=ECOZFKSfsDQvziR6S3RxFBSCZ-NjivGnu20HtxSG*CNeeR7VVjOtGJNiMxV9kGw14wQvuTtTJDYJ50k3c-1ABKJcq2DTmW4CDKutvp7lK5h5IzYfMh-ne8vpwLlYLm0b&amp;new=1" target="_blank" rel="noopener">单细胞全基因组扩增技术: MALBAC文章全文翻译</a></li>
<li>Zong, Chenghang, et al. “<a href="https://www.ncbi.nlm.nih.gov/pubmed/23258894/" target="_blank" rel="noopener">Genome-Wide Detection of Single-Nucleotide and Copy-Number Variations of a Single Human Cell.</a>” <em>Science</em>, vol. 338, no. 6114, 2012, pp. 1622–1626. </li>
<li>L, Huang, et al. “<a href="https://www.ncbi.nlm.nih.gov/pubmed/26077818/" target="_blank" rel="noopener">Single-Cell Whole-Genome Amplification and Sequencing: Methodology and Applications.</a>” <em>Annual Review of Genomics and Human Genetics</em>, vol. 16, no. 1, 2015, pp. 79–102. </li>
</ul>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201808_ScISOr-Seq/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201808_ScISOr-Seq/" itemprop="url">
                  Single-cell isoform RNA sequencing (ScISOr-Seq) across thousands of cells reveals isoforms of cerebellar cell types
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-07-31 19:18:19 / Modified: 19:31:19" itemprop="dateCreated datePublished" datetime="2018-07-31T19:18:19+08:00">2018-07-31</time>
            

            
              

              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/read/" itemprop="url" rel="index"><span itemprop="name">read</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>三代平台喜获单细胞测序技能</p>
<p><a href="https://www.biorxiv.org/content/early/2018/07/08/364950" target="_blank" rel="noopener"><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/ScISOr-Seq.png" alt=""></a></p>
<h2 id="Methods"><a href="#Methods" class="headerlink" title="Methods"></a>Methods</h2><ul>
<li>we employ microfluidics to produce amplified <strong>full-length cDNAs barcoded</strong> for their cell of origin</li>
<li>one pool for 3’ sequencing to measure gene expression(<strong>Pacific</strong>, <strong>Nanopore</strong>)</li>
<li>another pool for long-read sequencing and isoform expression(<strong>10xGenomics</strong>)</li>
</ul>
<h2 id="Results"><a href="#Results" class="headerlink" title="Results"></a>Results</h2><h3 id="a-expression"><a href="#a-expression" class="headerlink" title="a.expression"></a>a.expression</h3><ul>
<li>We sequence a mean of 17,885 reads per cell</li>
<li>After filtering cells and considering only reads confidently mapped to genes, we have 3,875 unique molecular identifiers (UMIs) and 1,448 genes per cell during 3’end sequencing</li>
<li>6,627 cells were clustered into 17 groups    </li>
<li>To assess stability of clusters, we tripled Illumina sequencing depth for rep2    </li>
</ul>
<h3 id="b-Full-length-cDNAs"><a href="#b-Full-length-cDNAs" class="headerlink" title="b.Full-length cDNAs"></a>b.Full-length cDNAs</h3><ul>
<li>generate ~5.2 million PacBio circular consensus reads    </li>
<li>cellular barcodes are located close to the polyA-tail, we first searched for polyA-tails    </li>
<li>chimeras, which were removed from further analysis    </li>
<li>for 58.0% (compared to 74.0% for 10x-3’seq) of the polyA-tail-containing CCS, we identified a perfect-match 16mer cellular barcode     </li>
</ul>
<h3 id="c-ScISOr-Seq-using-Nanopore-sequencing"><a href="#c-ScISOr-Seq-using-Nanopore-sequencing" class="headerlink" title="c.ScISOr-Seq using Nanopore sequencing"></a>c.ScISOr-Seq using Nanopore sequencing</h3><ul>
<li>Using 1µg of barcoded cDNA</li>
<li>~31.4% (1D) and ~35.2% (passed 1D2) of Nanopore reads have a 30bp window with &gt;=25 Ts   </li>
<li>nanopore 需要比 pacbio 更高的样品起始量，nanopore 数据错误率更高</li>
</ul>
<h2 id="Question"><a href="#Question" class="headerlink" title="Question"></a>Question</h2><ol>
<li>三代数据仍然只是用于转录本组装、注释，需要二代数据定量。分析逻辑约等于三代注释，cell特异转录本，外加二代scRNAseq</li>
<li>single cell full-length cDNAs 带 barcode 扩增原理？？？</li>
<li></li>
</ol>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201807_transvar/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201807_transvar/" itemprop="url">
                  从 p.HGVS 到基因组变异位置
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-07-19 19:25:21" itemprop="dateCreated datePublished" datetime="2018-07-19T19:25:21+08:00">2018-07-19</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2018-07-31 19:36:05" itemprop="dateModified" datetime="2018-07-31T19:36:05+08:00">2018-07-31</time>
              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/bioinfor/" itemprop="url" rel="index"><span itemprop="name">bioinfor</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h3 id="transcvar-web-版"><a href="#transcvar-web-版" class="headerlink" title="transcvar web 版"></a>transcvar web 版</h3><p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/transvar_web.png" alt=""></p>
<h3 id="transcvar-本地化"><a href="#transcvar-本地化" class="headerlink" title="transcvar 本地化"></a>transcvar 本地化</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> https://github.com/zwdzwd/transvar</span><br><span class="line">pip search transvar</span><br><span class="line">pip install transvar --user</span><br><span class="line"></span><br><span class="line">transvar config --download_ref --refversion hg19</span><br><span class="line">transvar config --download_anno --refversion hg19</span><br></pre></td></tr></table></figure>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/transvar.png" alt=""></p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">head HGVS_cds</span><br><span class="line"><span class="meta">#</span> MLH1:c.100G&gt;A</span><br><span class="line"><span class="meta">#</span> MLH1:c.790+1G&gt;A</span><br><span class="line"><span class="meta">#</span> VHL:c.1_17del17</span><br><span class="line"><span class="meta">#</span> VHL:c.136G&gt;T</span><br><span class="line"></span><br><span class="line">transvar canno --refversion hg19 --refseq -l HGVS_cds --oneline &gt;HGVS_cds.transvar.txt</span><br><span class="line"></span><br><span class="line">head HGVS_cds.transvar.txt</span><br><span class="line"><span class="meta">#</span> input	transcript	gene	strand	coordinates(gDNA/cDNA/protein)	region	info</span><br><span class="line"><span class="meta">#</span> MLH1:c.100G&gt;A	NM_001167618 (protein_coding)	MLH1	+	chr3:g.37059029G&gt;A/c.100G&gt;A/p.A34T	inside_[cds_in_exon_10]	CSQN=Missense;reference_codon=GCC;alternative_codon=ACC;dbxref=GeneID:4292,HGNC:7127,HPRD:00390,MIM:120436;aliases=NP_001161090;source=RefSeq</span><br><span class="line"><span class="meta">#</span> MLH1:c.790+1G&gt;A	NM_001258271 (protein_coding)	MLH1	+	chr3:g.37056036G&gt;A/c.790+1G&gt;A/.	inside_[intron_between_exon_9_and_10]	CSQN=IntronicSNV;dbxref=GeneID:4292,HGNC:7127,HPRD:00390,MIM:120436;aliases=NP_001245200;source=RefSeq</span><br><span class="line"><span class="meta">#</span> VHL:c.1_17del17	NM_198156 (protein_coding)	VHL	+	chr3:g.10183532_10183548del17/c.1_17del17/.	inside_[cds_in_exon_1]	CSQN=CdsStartDeletion;left_align_gDNA=g.10183529_10183545del17;unaligned_gDNA=g.10183532_10183548del17;left_align_cDNA=c.1-3_14del17;unalign_cDNA=c.1_17del17;cds_start_at_chr3:10183532_lost;dbxref=GeneID:7428,HGNC:12687,HPRD:01905,MIM:608537;aliases=NP_937799;source=RefSeq</span><br></pre></td></tr></table></figure>
<p>类似的也可以从 c.HGVS、g.HGVS 开始获得完整注释</p>
<h3 id="本地化数据库配置"><a href="#本地化数据库配置" class="headerlink" title="本地化数据库配置"></a>本地化数据库配置</h3><p>软件提供数据库路径为：<a href="http://transvar.info/transvar_user/annotations/" target="_blank" rel="noopener">http://transvar.info/transvar_user/annotations/</a>，本地数据库配置详细如下：</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/transvar_config.png" alt=""></p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/transvar_config2.png" alt=""></p>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201807_lnc_classification/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201807_lnc_classification/" itemprop="url">
                  lncRNA 根据染色体位置进行分类
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-07-19 19:15:52" itemprop="dateCreated datePublished" datetime="2018-07-19T19:15:52+08:00">2018-07-19</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2018-07-31 19:30:07" itemprop="dateModified" datetime="2018-07-31T19:30:07+08:00">2018-07-31</time>
              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/RNAseq/" itemprop="url" rel="index"><span itemprop="name">RNAseq</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>1/1/2018 </p>
<hr>
<p>根据与基因的相对位置，lncRNA可以分为<strong>Intergenic LncRNAs(lincRNA)</strong>, <strong>Bidirectional LncRNAs</strong>, <strong>Intronic LncRNAs</strong>, <strong>Antisense LncRNAs</strong>, <strong>Sense-overlapping LncRNAs</strong>五类，每类详细定义规则如下：</p>
<h3 id="定义"><a href="#定义" class="headerlink" title="定义"></a>定义</h3><p>ref: <a href="https://www.arraystar.com/reviews/v30-lncrna-classification/" target="_blank" rel="noopener">https://www.arraystar.com/reviews/v30-lncrna-classification/</a></p>
<ol>
<li><p>Intergenic LncRNAs</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/lnc_intergenic.jpeg" alt=""></p>
<p>Intergenic LncRNAs are long non-coding RNAs which locate between annotated protein-coding genes and are at least 1 kb away from the nearest protein-coding genes. They are named according to their 3-protein-coding genes nearby. Gene expression patterns have implicated these LincRNAs in diverse biological processes, including cell-cycle regulation, immune surveillance and embryonic stem cell pluripotency. LincRNAs collaborate with chromatin modifying protein (PRC2, CoREST and SCMX) to regulate gene expression at specific loci.</p>
</li>
<li><p>Bidirectional LncRNAs</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/lnc_bidirectional.jpeg" alt=""></p>
<p>A Bidirectional LncRNA is oriented head to head with a protein-coding gene within 1kb. A Bidirectional LncRNA transcript exhibits a similar expression pattern to its protein-coding counterpart which suggests that they may be subject to share regulatory pressures. However, the discordant expression relationships between bidirectional LncRNAs and protein coding gene pairs have also been found, challenging the assertion that LncRNA transcription occurs solely to “open” chromatin to promote the expression of neighboring coding genes.</p>
</li>
<li><p>Intronic LncRNAs</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/lnc_intronic.jpeg" alt=""></p>
<p>Intronic LncRNAs are RNA molecules that overlap with the intron of annotated coding genes in either sense or antisense orientation. Most of the Intronic LncRNAs have the same tissue expression patterns as the corresponding coding genes, and may stabilize protein-coding transcripts or regulate their alternative splicing.</p>
</li>
<li><p>Antisense LncRNAs</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/lnc_antisense.jpeg" alt=""></p>
<p>Antisense LncRNAs are RNA molecules that are transcribed from the antisense strand and overlap in part with well-defined spliced sense or intronless sense RNAs. Antisense-overlapping LncRNAs have a tendency to undergo fewer splicing events and typically show lower abundance than sense transcripts. The basal expression levels of antisense-overlapping LncRNAs and sense mRNAs in different tissues and cell lines can be either positively or negatively regulated. Antisense-overlapping LncRNAs are frequently functional and use diverse transcriptional and post-transcriptional gene regulatory mechanisms to carry out a wide variety of biological roles.</p>
</li>
<li><p>Sense-overlapping LncRNAs</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/lnc_sense.jpeg" alt=""></p>
<p>These LncRNAs can be considered transcript variants of protein-coding mRNAs, as they overlap with a known annotated gene on the same genomic strand. The majority of these LncRNAs lack substantial open reading frames (ORFs) for protein translation, while others contain an open reading frame that shares the same start codon as a protein-coding transcript for that gene, but unlikely encode a protein for several reasons, including non-sense mediated decay (NMD) issues that limits the translation of mRNAs with premature termination stop codons and trigger NMD-mediated destruction of the mRNA, or an upstream alternative open reading frame which inhibits the translation of the predicted ORF.</p>
</li>
</ol>
<h3 id="实现"><a href="#实现" class="headerlink" title="实现"></a>实现</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> step0 prepare</span><br><span class="line">awk -F "\t" 'BEGIN&#123;OFS="\t"&#125;&#123;if ($3=="transcript") &#123;print $1,$4,$5,$9,$7&#125;&#125;' lnc.gtf | sed -e 's/gene_id.*transcript_id "//g' -e 's/".*\t/\t.\t/g' &gt;lnc.transcript.bed</span><br><span class="line">awk -F "\t" 'BEGIN&#123;OFS="\t"&#125;&#123;if ($3=="transcript") &#123;print $1,$4,$5,$9,$7&#125;&#125;' mRNA.gtf | sed -e 's/gene_id.*transcript_id "//g' -e 's/".*\t/\t.\t/g' &gt;mRNA.transcript.bed</span><br><span class="line">awk -F "\t" 'BEGIN&#123;OFS="\t"&#125;&#123;if ($3=="exon") &#123;print $1,$4,$5,$9,$7&#125;&#125;' mRNA.gtf | sed -e 's/gene_id.*transcript_id "//g' -e 's/".*\t/\t.\t/g' &gt;mRNA.exon.bed</span><br><span class="line"></span><br><span class="line">head mRNA.transcript.bed</span><br><span class="line"><span class="meta">#</span> chr1	65419	71585	ENST00000641515	.	+</span><br><span class="line"><span class="meta">#</span> chr1	69055	70108	ENST00000335137	.	+</span><br><span class="line"><span class="meta">#</span> chr1	450703	451697	ENST00000426406	.	-</span><br><span class="line"><span class="meta">#</span> chr1	685679	686673	ENST00000332831	.	-</span><br><span class="line"><span class="meta">#</span> chr1	923928	939291	ENST00000420190	.	+</span><br><span class="line"><span class="meta">#</span> chr1	925150	935793	ENST00000437963	.	+</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> step1 获得 Intergenic lncRNA</span><br><span class="line"><span class="meta">#</span> -v 取反，获得不和任何 coding-protein 基因（包括基因上下游 1000 bp范围）重合的 lncRNA</span><br><span class="line">bedtools window -a lnc.transcript.bed -b mRNA.transcript.bed -v &gt;lnc.Intergenic.bed</span><br><span class="line"><span class="meta">#</span>############## lnc.Intergenic.bed ###############</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> step2 基因 1000 bp 区间内的 lncRNA</span><br><span class="line"><span class="meta">#</span> 获得和 coding-protein 有重叠的 lncRNA ，lnc_gene.bed</span><br><span class="line">bedtools intersect -a lnc.transcript.bed -b mRNA.transcript.bed -wa | sort -u &gt; lnc_gene.bed</span><br><span class="line"><span class="meta">#</span> 获得 coding-protein 基因上下游 1000 bp 范围内的 lnc ，lnc_gene_1k.bed</span><br><span class="line">python -c 'with open("lnc.transcript.bed") as a, open("lnc_gene.bed") as m, open("lnc.Intergenic.bed") as i, open("lnc_gene_1k.bed", "w") as k1: k1.write("".join(set(a.readlines())-set(m.readlines())-set(i.readlines())))'</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> step2b 根据转录方向进行分类</span><br><span class="line">bedtools window -a lnc_gene_1k.bed -b mRNA.transcript.bed -u -Sm | sort -u &gt;lnc.Bidirectional.bed</span><br><span class="line"><span class="meta">#</span>############## lnc.Bidirectional.bed ###############</span><br><span class="line">bedtools window -a lnc_gene_1k.bed -b mRNA.transcript.bed -u -sm | sort -u &gt;lnc.Enhancer.bed</span><br><span class="line"><span class="meta">#</span>############## lnc.Enhancer.bed ###############</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> step3 </span><br><span class="line"><span class="meta">#</span> 和 coding-protein 重叠，不和外显子重叠，得到内含子区域的lnc</span><br><span class="line">bedtools intersect -a lnc_gene.bed -b mRNA.exon.bed -v &gt; lnc.Intronic.bed</span><br><span class="line"><span class="meta">#</span>############## lnc.Intronic.bed ###############</span><br><span class="line"><span class="meta">#</span> 取相同链的 lnc</span><br><span class="line">bedtools intersect -a lnc_gene.bed -b mRNA.exon.bed -wa -s | sort -u &gt; lnc.Sense.bed</span><br><span class="line"><span class="meta">#</span>############## lnc.Sense.bed ###############</span><br><span class="line"><span class="meta">#</span> 取反向链的 lnc</span><br><span class="line">bedtools intersect -a lnc_gene.bed -b mRNA.exon.bed -wa -S | sort -u &gt; lnc.Antisense.bed</span><br><span class="line"><span class="meta">#</span>############## lnc.Antisense.bed ###############</span><br></pre></td></tr></table></figure>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201807_FamilyGene/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201807_FamilyGene/" itemprop="url">
                  家族基因分析的一般套路
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-07-19 19:07:27" itemprop="dateCreated datePublished" datetime="2018-07-19T19:07:27+08:00">2018-07-19</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2018-07-31 19:14:38" itemprop="dateModified" datetime="2018-07-31T19:14:38+08:00">2018-07-31</time>
              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/bioinfor/" itemprop="url" rel="index"><span itemprop="name">bioinfor</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>7/2016 </p>
<hr>
<ol>
<li><p>分析前准备</p>
<p>分析前准备主要是两点，找文章，找序列。</p>
<ul>
<li>找文章，该家族基因是否在研究物种或其他物种中报告，都进行了什么分析和试验；</li>
<li>找序列，获取该家族代表性序列进行后续分析</li>
</ul>
</li>
<li><p>家族基因成员鉴定</p>
<p>主要 blast 和 hmmer 两种方式，一般讲来使用两个方式的并集，然后根据基因特征再一步确认过滤</p>
<ul>
<li>blast</li>
<li>hmmer</li>
</ul>
</li>
<li><p>家族基因特征描述</p>
<ul>
<li>染色体分布</li>
<li>基因结构</li>
<li>motif分析</li>
<li>启动子序列分析</li>
</ul>
</li>
<li><p>多物种进化关系分析</p>
<ul>
<li>构建分子进化树</li>
<li>Ka/Ks计算等等</li>
</ul>
</li>
<li><p>基因分类命名</p>
<p>一般根据进化关系、或者分析特征进行分类，然后染色体顺序命名</p>
</li>
<li><p>基因表达分析</p>
<ul>
<li>多组织材料表达分析</li>
<li>不同处理样品表达分析</li>
<li>共表达分析</li>
</ul>
</li>
<li><p>候选基因表达、功能验证</p>
</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">graph LR</span><br><span class="line">A[FamilyGene.seq] -- blast --&gt; B[FamilyGene.list]</span><br><span class="line">A[FamilyGene.hmm] -- hmmer --&gt; B[FamilyGene.list]</span><br><span class="line">B[FamilyGene.list] --&gt; C[FeatureAnalysis]</span><br><span class="line">C[FeatureAnalysis] --&gt; Ca[染色体分布]</span><br><span class="line">C[FeatureAnalysis] --&gt; Cb[基因结构]</span><br><span class="line">C[FeatureAnalysis] --&gt; Cc[motif分析]</span><br><span class="line">C[FeatureAnalysis] --&gt; Cd[promotor seq analysis]</span><br><span class="line">B[FamilyGene.list] --&gt; D[phylogenetics]</span><br><span class="line">B[FamilyGene.list] --&gt; E[Group,Name]</span><br><span class="line">B[FamilyGene.list] --&gt; F[Expression]</span><br><span class="line">F[Expression] --&gt; G[表达, 功能验证]</span><br></pre></td></tr></table></figure>
          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201807_singleTcell_NSCLC/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201807_singleTcell_NSCLC/" itemprop="url">
                  Global characterization of T cells in non-small-cell lung cancer by single-cell sequencing
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              

              
                
              

              <time title="Created: 2018-07-16 19:18:19" itemprop="dateCreated datePublished" datetime="2018-07-16T19:18:19+08:00">2018-07-16</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Edited on</span>
                
                <time title="Modified: 2018-07-31 19:30:40" itemprop="dateModified" datetime="2018-07-31T19:30:40+08:00">2018-07-31</time>
              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/read/" itemprop="url" rel="index"><span itemprop="name">read</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>肺癌T细胞 scRNAseq 分析</p>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/29942094" target="_blank" rel="noopener"><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/singleTcell_NSCLC.png" alt=""></a></p>
<h3 id="Methods"><a href="#Methods" class="headerlink" title="Methods"></a>Methods</h3><ol>
<li><p>we performed deep single-cell RNA sequencing on T cells isolated from <strong>tumor</strong>, <strong>adjacent normal tissues</strong> and <strong>peripheral blood</strong> for 14 treatment-naïve patients, including <strong>11 adenocarcinomas</strong> and <strong>three squamous cell carcinomas</strong>    </p>
</li>
<li><p>After confirming the existence of T cell infiltration in NSCLC tumors (Supplementary Fig. 1a), we sorted various T cell subtypes based on <strong>cell surface markers CD3/ CD4/CD8/CD25 by fluorescence-activated cell sorting</strong>    </p>
</li>
<li><p>A total of 12,346 cells were sequenced at an <strong>average depth of 1.04 million uniquely-mapped read pairs</strong> per cell</p>
<blockquote>
<p>The final expression table contained <strong>12,415 genes</strong> and <strong>11,769 cells</strong>    </p>
</blockquote>
</li>
<li><p>Dimensional reduction analysis (t-SNE) applied to the expression data revealed that T cells clustered primarily based on their tissue origins and subtypes      </p>
</li>
<li><p>To further address the intrinsic T cell heterogeneity, we applied unsupervised clustering based on t-SNE+ densityClust and identified seven CD8 and nine CD4 clusters    </p>
</li>
<li><p>To gain insight into the clonal relationship among individual T cells, we reconstructed their T cell receptor (TCR) sequences    </p>
<blockquote>
<p>We obtained full-length TCRs with both alpha and beta chains for 8,038 cells of the 16 clusters, including 5,015 harboring unique TCRs and 3,023 harboring repeated use of TCRs, indicative of clonal expansion    </p>
</blockquote>
</li>
<li><p>Detection of Treg marker genes    </p>
</li>
<li><p>Survival analysis. </p>
</li>
</ol>
<h3 id="Results"><a href="#Results" class="headerlink" title="Results"></a>Results</h3><ol>
<li>As well as tumor-infiltrating CD8+ T cells undergoing exhaustion, we observed two clusters of cells exhibiting states preceding exhaustion, and a high ratio of “pre-exhausted” to exhausted T cells was associated with better prognosis of lung adenocarcinoma.</li>
<li><p>we observed further heterogeneity within the tumor regulatory T cells (Tregs), characterized by the bimodal distribution of TNFRSF9, an activation marker for antigenspecific Tregs. The gene signature of those activated tumor Tregs, which included IL1R2, correlated with poor prognosis in lung adenocarcinoma. </p>
</li>
<li><p>…</p>
</li>
</ol>
<h3 id="points"><a href="#points" class="headerlink" title="points"></a>points</h3><ul>
<li>通常，CX3CR1在正常组织中表达高于肿瘤组织</li>
</ul>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  
  <nav class="pagination">
    <span class="page-number current">1</span><a class="page-number" href="/page/2/">2</a><span class="space">&hellip;</span><a class="page-number" href="/page/5/">5</a><a class="extend next" rel="next" href="/page/2/"><i class="fa fa-angle-right" aria-label="Next page"></i></a>
  </nav>



          </div>
          

        </div>
        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="powered-by">
  <i class="fa fa-user-md"></i>
  <span id="busuanzi_container_site_uv">
    本站访客数:<span id="busuanzi_value_site_uv"></span>
  </span>
  <span id="busuanzi_container_site_pv">
    本站总访问量:<span id="busuanzi_value_site_pv"></span>
  </span>
</div>

<div class="copyright">&copy; <span itemprop="copyrightYear">2018</span>
  <span class="with-love" id="animate">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">ChengCZ</span>

  

  
</div>



<!--

  <div class="powered-by">Powered by <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> v3.7.1</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">Theme &mdash; <a class="theme-link" target="_blank" href="https://theme-next.org">NexT.Mist</a> v6.3.0</div>


-->

        








        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    
	
    

    
  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>


























  
  
    <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  


  


  <script type="text/javascript" src="/js/src/utils.js?v=6.3.0"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=6.3.0"></script>



  
  

  

  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=6.3.0"></script>



  



  










  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  

  

  
  

  

  

  

  

  

</body>
</html>
