<!DOCTYPE html>












  


<html class="theme-next mist use-motion" lang="zh-CN">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2"/>
<meta name="theme-color" content="#222">






  
  
    
    
  <script src="/lib/pace/pace.min.js?v=1.0.2"></script>
  <link href="/lib/pace/pace-theme-minimal.min.css?v=1.0.2" rel="stylesheet">







<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />






















<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=6.3.0" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=6.3.0">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=6.3.0">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=6.3.0">


  <link rel="mask-icon" href="/images/logo.svg?v=6.3.0" color="#222">









<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    version: '6.3.0',
    sidebar: {"position":"left","display":"remove","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: false,
    fastclick: false,
    lazyload: false,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>


  




  <meta property="og:type" content="website">
<meta property="og:title" content="ChengCZ&#39;s Notebook">
<meta property="og:url" content="http://chengcz.github.io/index.html">
<meta property="og:site_name" content="ChengCZ&#39;s Notebook">
<meta property="og:locale" content="zh-CN">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="ChengCZ&#39;s Notebook">



  <link rel="alternate" href="/atom.xml" title="ChengCZ's Notebook" type="application/atom+xml" />




  <link rel="canonical" href="http://chengcz.github.io/"/>



<script type="text/javascript" id="page.configurations">
  CONFIG.page = {
    sidebar: "",
  };
</script>

  <title>ChengCZ's Notebook</title>
  






  <script type="text/javascript">
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?a79f8c7f932a2d3ed822562170533a8d";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>




  <noscript>
  <style type="text/css">
    .use-motion .motion-element,
    .use-motion .brand,
    .use-motion .menu-item,
    .sidebar-inner,
    .use-motion .post-block,
    .use-motion .pagination,
    .use-motion .comments,
    .use-motion .post-header,
    .use-motion .post-body,
    .use-motion .collection-title { opacity: initial; }

    .use-motion .logo,
    .use-motion .site-title,
    .use-motion .site-subtitle {
      opacity: initial;
      top: initial;
    }

    .use-motion {
      .logo-line-before i { left: initial; }
      .logo-line-after i { right: initial; }
    }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-CN">

  
  
    
  

  <div class="container sidebar-position-left 
  page-home">
    <div class="headband"></div>

    <a href="https://github.com/chengcz/" class="github-corner" aria-label="View source on Github"><svg width="80" height="80" viewBox="0 0 250 250" style="fill:#151513; color:#fff; position: absolute; top: 0; border: 0; left: 0; transform: scale(-1, 1);" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg><style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style></a>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">ChengCZ's Notebook</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
    
  </div>

  <div class="site-nav-toggle">
    <button aria-label="切换导航栏">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>



<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        
        
        
          
          <li class="menu-item menu-item-home menu-item-active">
    <a href="/" rel="section">
      <i class="menu-item-icon fa fa-fw fa-home"></i> <br />首页</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-tags">
    <a href="/tags" rel="section">
      <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />标签</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-categories">
    <a href="/categories" rel="section">
      <i class="menu-item-icon fa fa-fw fa-th"></i> <br />分类</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-archives">
    <a href="/archives" rel="section">
      <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />归档</a>
  </li>
        
        
        
          
          <li class="menu-item menu-item-about">
    <a href="/about" rel="section">
      <i class="menu-item-icon fa fa-fw fa-user"></i> <br />关于</a>
  </li>

      
      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br />搜索</a>
        </li>
      
    </ul>
  

  
    

  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="搜索..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



  



</div>
    </header>

    


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201808-MALBAC-md/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201808-MALBAC-md/" itemprop="url">
                  Whole Genome Amplification Technologies:MALBAC
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2018-08-04 22:43:24 / 修改时间：22:49:14" itemprop="dateCreated datePublished" datetime="2018-08-04T22:43:24+08:00">2018-08-04</time>
            

            
              

              
            
          </span>

          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h3 id="Technical-principles"><a href="#Technical-principles" class="headerlink" title="Technical principles"></a>Technical principles</h3><p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/MALBAC.png" alt=""></p>
<ol>
<li><p>A single cell is picked and lysed. First, genomic DNA of the single cell is melted into single-stranded DNA molecules at 94°C. </p>
<blockquote>
<p>Picograms of DNA fragments (~10 to 100 kb)     </p>
</blockquote>
</li>
<li><p>MALBAC primers then anneal randomly to single-stranded DNA molecules at 0°C and are extended by a polymerase with displacement activity at elevated temperatures, creating semiamplicons.</p>
<blockquote>
<p>random primers, each having a common <strong>27-nucleotide sequence</strong> and <strong>8 variable nucleotides</strong>     </p>
</blockquote>
</li>
<li><p>In the following five temperature cycles, after the step of looping the full amplicons, singlestranded amplicons and the genomic DNA are used as a template to produce full amplicons and additional semiamplicons, respectively. </p>
</li>
<li><p>For full amplicons, the 3′ end is complementary to the sequence on the 5′ end. The two ends hybridize to form looped DNA, which can efficiently prevent the full amplicon from being used as a template, therefore warranting a close-to-linear amplification. </p>
</li>
<li><p>After the five cycles of linear preamplification, only the full amplicons can be exponentially amplified in the following PCR <strong>using the common 27- nucleotide sequence as the primer</strong>. PCR reaction will generate microgram level of DNA material for sequencing experiments.    </p>
</li>
</ol>
<h3 id="Advantages"><a href="#Advantages" class="headerlink" title="Advantages"></a>Advantages</h3><ul>
<li>quasi-linear amplification    </li>
<li>accuracy in CNV detection, especially after signal normalization with a reference cell    </li>
</ul>
<p>###Disadvantages </p>
<ul>
<li>MALBAC is not free from sequence-dependent bias    </li>
<li>MALBAC is that it has a high false positive rate for SNV detection    </li>
<li>underamplified regions of the genome are sometimes lost during amplification    </li>
</ul>
<h3 id="ref"><a href="#ref" class="headerlink" title="ref"></a>ref</h3><ul>
<li><a href="https://mp.weixin.qq.com/s?src=11&amp;timestamp=1533173344&amp;ver=1035&amp;signature=ECOZFKSfsDQvziR6S3RxFBSCZ-NjivGnu20HtxSG*CNeeR7VVjOtGJNiMxV9kGw14wQvuTtTJDYJ50k3c-1ABKJcq2DTmW4CDKutvp7lK5h5IzYfMh-ne8vpwLlYLm0b&amp;new=1" target="_blank" rel="noopener">单细胞全基因组扩增技术: MALBAC文章全文翻译</a></li>
<li>Zong, Chenghang, et al. “<a href="https://www.ncbi.nlm.nih.gov/pubmed/23258894/" target="_blank" rel="noopener">Genome-Wide Detection of Single-Nucleotide and Copy-Number Variations of a Single Human Cell.</a>” <em>Science</em>, vol. 338, no. 6114, 2012, pp. 1622–1626. </li>
<li>L, Huang, et al. “<a href="https://www.ncbi.nlm.nih.gov/pubmed/26077818/" target="_blank" rel="noopener">Single-Cell Whole-Genome Amplification and Sequencing: Methodology and Applications.</a>” <em>Annual Review of Genomics and Human Genetics</em>, vol. 16, no. 1, 2015, pp. 79–102. </li>
</ul>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201808_ScISOr-Seq/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201808_ScISOr-Seq/" itemprop="url">
                  Single-cell isoform RNA sequencing (ScISOr-Seq) across thousands of cells reveals isoforms of cerebellar cell types
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2018-07-31 19:18:19 / 修改时间：19:31:19" itemprop="dateCreated datePublished" datetime="2018-07-31T19:18:19+08:00">2018-07-31</time>
            

            
              

              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/read/" itemprop="url" rel="index"><span itemprop="name">read</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>三代平台喜获单细胞测序技能</p>
<p><a href="https://www.biorxiv.org/content/early/2018/07/08/364950" target="_blank" rel="noopener"><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/ScISOr-Seq.png" alt=""></a></p>
<h2 id="Methods"><a href="#Methods" class="headerlink" title="Methods"></a>Methods</h2><ul>
<li>we employ microfluidics to produce amplified <strong>full-length cDNAs barcoded</strong> for their cell of origin</li>
<li>one pool for 3’ sequencing to measure gene expression(<strong>Pacific</strong>, <strong>Nanopore</strong>)</li>
<li>another pool for long-read sequencing and isoform expression(<strong>10xGenomics</strong>)</li>
</ul>
<h2 id="Results"><a href="#Results" class="headerlink" title="Results"></a>Results</h2><h3 id="a-expression"><a href="#a-expression" class="headerlink" title="a.expression"></a>a.expression</h3><ul>
<li>We sequence a mean of 17,885 reads per cell</li>
<li>After filtering cells and considering only reads confidently mapped to genes, we have 3,875 unique molecular identifiers (UMIs) and 1,448 genes per cell during 3’end sequencing</li>
<li>6,627 cells were clustered into 17 groups    </li>
<li>To assess stability of clusters, we tripled Illumina sequencing depth for rep2    </li>
</ul>
<h3 id="b-Full-length-cDNAs"><a href="#b-Full-length-cDNAs" class="headerlink" title="b.Full-length cDNAs"></a>b.Full-length cDNAs</h3><ul>
<li>generate ~5.2 million PacBio circular consensus reads    </li>
<li>cellular barcodes are located close to the polyA-tail, we first searched for polyA-tails    </li>
<li>chimeras, which were removed from further analysis    </li>
<li>for 58.0% (compared to 74.0% for 10x-3’seq) of the polyA-tail-containing CCS, we identified a perfect-match 16mer cellular barcode     </li>
</ul>
<h3 id="c-ScISOr-Seq-using-Nanopore-sequencing"><a href="#c-ScISOr-Seq-using-Nanopore-sequencing" class="headerlink" title="c.ScISOr-Seq using Nanopore sequencing"></a>c.ScISOr-Seq using Nanopore sequencing</h3><ul>
<li>Using 1µg of barcoded cDNA</li>
<li>~31.4% (1D) and ~35.2% (passed 1D2) of Nanopore reads have a 30bp window with &gt;=25 Ts   </li>
<li>nanopore 需要比 pacbio 更高的样品起始量，nanopore 数据错误率更高</li>
</ul>
<h2 id="Question"><a href="#Question" class="headerlink" title="Question"></a>Question</h2><ol>
<li>三代数据仍然只是用于转录本组装、注释，需要二代数据定量。分析逻辑约等于三代注释，cell特异转录本，外加二代scRNAseq</li>
<li>single cell full-length cDNAs 带 barcode 扩增原理？？？</li>
<li></li>
</ol>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201807_transvar/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201807_transvar/" itemprop="url">
                  从 p.HGVS 到基因组变异位置
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2018-07-19 19:25:21" itemprop="dateCreated datePublished" datetime="2018-07-19T19:25:21+08:00">2018-07-19</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">更新于</span>
                
                <time title="修改时间：2018-07-31 19:36:05" itemprop="dateModified" datetime="2018-07-31T19:36:05+08:00">2018-07-31</time>
              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/bioinfor/" itemprop="url" rel="index"><span itemprop="name">bioinfor</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h3 id="transcvar-web-版"><a href="#transcvar-web-版" class="headerlink" title="transcvar web 版"></a>transcvar web 版</h3><p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/transvar_web.png" alt=""></p>
<h3 id="transcvar-本地化"><a href="#transcvar-本地化" class="headerlink" title="transcvar 本地化"></a>transcvar 本地化</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> https://github.com/zwdzwd/transvar</span><br><span class="line">pip search transvar</span><br><span class="line">pip install transvar --user</span><br><span class="line"></span><br><span class="line">transvar config --download_ref --refversion hg19</span><br><span class="line">transvar config --download_anno --refversion hg19</span><br></pre></td></tr></table></figure>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/transvar.png" alt=""></p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">head HGVS_cds</span><br><span class="line"><span class="meta">#</span> MLH1:c.100G&gt;A</span><br><span class="line"><span class="meta">#</span> MLH1:c.790+1G&gt;A</span><br><span class="line"><span class="meta">#</span> VHL:c.1_17del17</span><br><span class="line"><span class="meta">#</span> VHL:c.136G&gt;T</span><br><span class="line"></span><br><span class="line">transvar canno --refversion hg19 --refseq -l HGVS_cds --oneline &gt;HGVS_cds.transvar.txt</span><br><span class="line"></span><br><span class="line">head HGVS_cds.transvar.txt</span><br><span class="line"><span class="meta">#</span> input	transcript	gene	strand	coordinates(gDNA/cDNA/protein)	region	info</span><br><span class="line"><span class="meta">#</span> MLH1:c.100G&gt;A	NM_001167618 (protein_coding)	MLH1	+	chr3:g.37059029G&gt;A/c.100G&gt;A/p.A34T	inside_[cds_in_exon_10]	CSQN=Missense;reference_codon=GCC;alternative_codon=ACC;dbxref=GeneID:4292,HGNC:7127,HPRD:00390,MIM:120436;aliases=NP_001161090;source=RefSeq</span><br><span class="line"><span class="meta">#</span> MLH1:c.790+1G&gt;A	NM_001258271 (protein_coding)	MLH1	+	chr3:g.37056036G&gt;A/c.790+1G&gt;A/.	inside_[intron_between_exon_9_and_10]	CSQN=IntronicSNV;dbxref=GeneID:4292,HGNC:7127,HPRD:00390,MIM:120436;aliases=NP_001245200;source=RefSeq</span><br><span class="line"><span class="meta">#</span> VHL:c.1_17del17	NM_198156 (protein_coding)	VHL	+	chr3:g.10183532_10183548del17/c.1_17del17/.	inside_[cds_in_exon_1]	CSQN=CdsStartDeletion;left_align_gDNA=g.10183529_10183545del17;unaligned_gDNA=g.10183532_10183548del17;left_align_cDNA=c.1-3_14del17;unalign_cDNA=c.1_17del17;cds_start_at_chr3:10183532_lost;dbxref=GeneID:7428,HGNC:12687,HPRD:01905,MIM:608537;aliases=NP_937799;source=RefSeq</span><br></pre></td></tr></table></figure>
<p>类似的也可以从 c.HGVS、g.HGVS 开始获得完整注释</p>
<h3 id="本地化数据库配置"><a href="#本地化数据库配置" class="headerlink" title="本地化数据库配置"></a>本地化数据库配置</h3><p>软件提供数据库路径为：<a href="http://transvar.info/transvar_user/annotations/" target="_blank" rel="noopener">http://transvar.info/transvar_user/annotations/</a>，本地数据库配置详细如下：</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/transvar_config.png" alt=""></p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/transvar_config2.png" alt=""></p>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201807_lnc_classification/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201807_lnc_classification/" itemprop="url">
                  lncRNA 根据染色体位置进行分类
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2018-07-19 19:15:52" itemprop="dateCreated datePublished" datetime="2018-07-19T19:15:52+08:00">2018-07-19</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">更新于</span>
                
                <time title="修改时间：2018-07-31 19:30:07" itemprop="dateModified" datetime="2018-07-31T19:30:07+08:00">2018-07-31</time>
              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/RNAseq/" itemprop="url" rel="index"><span itemprop="name">RNAseq</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>1/1/2018 </p>
<hr>
<p>根据与基因的相对位置，lncRNA可以分为<strong>Intergenic LncRNAs(lincRNA)</strong>, <strong>Bidirectional LncRNAs</strong>, <strong>Intronic LncRNAs</strong>, <strong>Antisense LncRNAs</strong>, <strong>Sense-overlapping LncRNAs</strong>五类，每类详细定义规则如下：</p>
<h3 id="定义"><a href="#定义" class="headerlink" title="定义"></a>定义</h3><p>ref: <a href="https://www.arraystar.com/reviews/v30-lncrna-classification/" target="_blank" rel="noopener">https://www.arraystar.com/reviews/v30-lncrna-classification/</a></p>
<ol>
<li><p>Intergenic LncRNAs</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/lnc_intergenic.jpeg" alt=""></p>
<p>Intergenic LncRNAs are long non-coding RNAs which locate between annotated protein-coding genes and are at least 1 kb away from the nearest protein-coding genes. They are named according to their 3-protein-coding genes nearby. Gene expression patterns have implicated these LincRNAs in diverse biological processes, including cell-cycle regulation, immune surveillance and embryonic stem cell pluripotency. LincRNAs collaborate with chromatin modifying protein (PRC2, CoREST and SCMX) to regulate gene expression at specific loci.</p>
</li>
<li><p>Bidirectional LncRNAs</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/lnc_bidirectional.jpeg" alt=""></p>
<p>A Bidirectional LncRNA is oriented head to head with a protein-coding gene within 1kb. A Bidirectional LncRNA transcript exhibits a similar expression pattern to its protein-coding counterpart which suggests that they may be subject to share regulatory pressures. However, the discordant expression relationships between bidirectional LncRNAs and protein coding gene pairs have also been found, challenging the assertion that LncRNA transcription occurs solely to “open” chromatin to promote the expression of neighboring coding genes.</p>
</li>
<li><p>Intronic LncRNAs</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/lnc_intronic.jpeg" alt=""></p>
<p>Intronic LncRNAs are RNA molecules that overlap with the intron of annotated coding genes in either sense or antisense orientation. Most of the Intronic LncRNAs have the same tissue expression patterns as the corresponding coding genes, and may stabilize protein-coding transcripts or regulate their alternative splicing.</p>
</li>
<li><p>Antisense LncRNAs</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/lnc_antisense.jpeg" alt=""></p>
<p>Antisense LncRNAs are RNA molecules that are transcribed from the antisense strand and overlap in part with well-defined spliced sense or intronless sense RNAs. Antisense-overlapping LncRNAs have a tendency to undergo fewer splicing events and typically show lower abundance than sense transcripts. The basal expression levels of antisense-overlapping LncRNAs and sense mRNAs in different tissues and cell lines can be either positively or negatively regulated. Antisense-overlapping LncRNAs are frequently functional and use diverse transcriptional and post-transcriptional gene regulatory mechanisms to carry out a wide variety of biological roles.</p>
</li>
<li><p>Sense-overlapping LncRNAs</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/lnc_sense.jpeg" alt=""></p>
<p>These LncRNAs can be considered transcript variants of protein-coding mRNAs, as they overlap with a known annotated gene on the same genomic strand. The majority of these LncRNAs lack substantial open reading frames (ORFs) for protein translation, while others contain an open reading frame that shares the same start codon as a protein-coding transcript for that gene, but unlikely encode a protein for several reasons, including non-sense mediated decay (NMD) issues that limits the translation of mRNAs with premature termination stop codons and trigger NMD-mediated destruction of the mRNA, or an upstream alternative open reading frame which inhibits the translation of the predicted ORF.</p>
</li>
</ol>
<h3 id="实现"><a href="#实现" class="headerlink" title="实现"></a>实现</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> step0 prepare</span><br><span class="line">awk -F "\t" 'BEGIN&#123;OFS="\t"&#125;&#123;if ($3=="transcript") &#123;print $1,$4,$5,$9,$7&#125;&#125;' lnc.gtf | sed -e 's/gene_id.*transcript_id "//g' -e 's/".*\t/\t.\t/g' &gt;lnc.transcript.bed</span><br><span class="line">awk -F "\t" 'BEGIN&#123;OFS="\t"&#125;&#123;if ($3=="transcript") &#123;print $1,$4,$5,$9,$7&#125;&#125;' mRNA.gtf | sed -e 's/gene_id.*transcript_id "//g' -e 's/".*\t/\t.\t/g' &gt;mRNA.transcript.bed</span><br><span class="line">awk -F "\t" 'BEGIN&#123;OFS="\t"&#125;&#123;if ($3=="exon") &#123;print $1,$4,$5,$9,$7&#125;&#125;' mRNA.gtf | sed -e 's/gene_id.*transcript_id "//g' -e 's/".*\t/\t.\t/g' &gt;mRNA.exon.bed</span><br><span class="line"></span><br><span class="line">head mRNA.transcript.bed</span><br><span class="line"><span class="meta">#</span> chr1	65419	71585	ENST00000641515	.	+</span><br><span class="line"><span class="meta">#</span> chr1	69055	70108	ENST00000335137	.	+</span><br><span class="line"><span class="meta">#</span> chr1	450703	451697	ENST00000426406	.	-</span><br><span class="line"><span class="meta">#</span> chr1	685679	686673	ENST00000332831	.	-</span><br><span class="line"><span class="meta">#</span> chr1	923928	939291	ENST00000420190	.	+</span><br><span class="line"><span class="meta">#</span> chr1	925150	935793	ENST00000437963	.	+</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> step1 获得 Intergenic lncRNA</span><br><span class="line"><span class="meta">#</span> -v 取反，获得不和任何 coding-protein 基因（包括基因上下游 1000 bp范围）重合的 lncRNA</span><br><span class="line">bedtools window -a lnc.transcript.bed -b mRNA.transcript.bed -v &gt;lnc.Intergenic.bed</span><br><span class="line"><span class="meta">#</span>############## lnc.Intergenic.bed ###############</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> step2 基因 1000 bp 区间内的 lncRNA</span><br><span class="line"><span class="meta">#</span> 获得和 coding-protein 有重叠的 lncRNA ，lnc_gene.bed</span><br><span class="line">bedtools intersect -a lnc.transcript.bed -b mRNA.transcript.bed -wa | sort -u &gt; lnc_gene.bed</span><br><span class="line"><span class="meta">#</span> 获得 coding-protein 基因上下游 1000 bp 范围内的 lnc ，lnc_gene_1k.bed</span><br><span class="line">python -c 'with open("lnc.transcript.bed") as a, open("lnc_gene.bed") as m, open("lnc.Intergenic.bed") as i, open("lnc_gene_1k.bed", "w") as k1: k1.write("".join(set(a.readlines())-set(m.readlines())-set(i.readlines())))'</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> step2b 根据转录方向进行分类</span><br><span class="line">bedtools window -a lnc_gene_1k.bed -b mRNA.transcript.bed -u -Sm | sort -u &gt;lnc.Bidirectional.bed</span><br><span class="line"><span class="meta">#</span>############## lnc.Bidirectional.bed ###############</span><br><span class="line">bedtools window -a lnc_gene_1k.bed -b mRNA.transcript.bed -u -sm | sort -u &gt;lnc.Enhancer.bed</span><br><span class="line"><span class="meta">#</span>############## lnc.Enhancer.bed ###############</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> step3 </span><br><span class="line"><span class="meta">#</span> 和 coding-protein 重叠，不和外显子重叠，得到内含子区域的lnc</span><br><span class="line">bedtools intersect -a lnc_gene.bed -b mRNA.exon.bed -v &gt; lnc.Intronic.bed</span><br><span class="line"><span class="meta">#</span>############## lnc.Intronic.bed ###############</span><br><span class="line"><span class="meta">#</span> 取相同链的 lnc</span><br><span class="line">bedtools intersect -a lnc_gene.bed -b mRNA.exon.bed -wa -s | sort -u &gt; lnc.Sense.bed</span><br><span class="line"><span class="meta">#</span>############## lnc.Sense.bed ###############</span><br><span class="line"><span class="meta">#</span> 取反向链的 lnc</span><br><span class="line">bedtools intersect -a lnc_gene.bed -b mRNA.exon.bed -wa -S | sort -u &gt; lnc.Antisense.bed</span><br><span class="line"><span class="meta">#</span>############## lnc.Antisense.bed ###############</span><br></pre></td></tr></table></figure>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201807_FamilyGene/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201807_FamilyGene/" itemprop="url">
                  家族基因分析的一般套路
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2018-07-19 19:07:27" itemprop="dateCreated datePublished" datetime="2018-07-19T19:07:27+08:00">2018-07-19</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">更新于</span>
                
                <time title="修改时间：2018-07-31 19:14:38" itemprop="dateModified" datetime="2018-07-31T19:14:38+08:00">2018-07-31</time>
              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/bioinfor/" itemprop="url" rel="index"><span itemprop="name">bioinfor</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>7/2016 </p>
<hr>
<ol>
<li><p>分析前准备</p>
<p>分析前准备主要是两点，找文章，找序列。</p>
<ul>
<li>找文章，该家族基因是否在研究物种或其他物种中报告，都进行了什么分析和试验；</li>
<li>找序列，获取该家族代表性序列进行后续分析</li>
</ul>
</li>
<li><p>家族基因成员鉴定</p>
<p>主要 blast 和 hmmer 两种方式，一般讲来使用两个方式的并集，然后根据基因特征再一步确认过滤</p>
<ul>
<li>blast</li>
<li>hmmer</li>
</ul>
</li>
<li><p>家族基因特征描述</p>
<ul>
<li>染色体分布</li>
<li>基因结构</li>
<li>motif分析</li>
<li>启动子序列分析</li>
</ul>
</li>
<li><p>多物种进化关系分析</p>
<ul>
<li>构建分子进化树</li>
<li>Ka/Ks计算等等</li>
</ul>
</li>
<li><p>基因分类命名</p>
<p>一般根据进化关系、或者分析特征进行分类，然后染色体顺序命名</p>
</li>
<li><p>基因表达分析</p>
<ul>
<li>多组织材料表达分析</li>
<li>不同处理样品表达分析</li>
<li>共表达分析</li>
</ul>
</li>
<li><p>候选基因表达、功能验证</p>
</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">graph LR</span><br><span class="line">A[FamilyGene.seq] -- blast --&gt; B[FamilyGene.list]</span><br><span class="line">A[FamilyGene.hmm] -- hmmer --&gt; B[FamilyGene.list]</span><br><span class="line">B[FamilyGene.list] --&gt; C[FeatureAnalysis]</span><br><span class="line">C[FeatureAnalysis] --&gt; Ca[染色体分布]</span><br><span class="line">C[FeatureAnalysis] --&gt; Cb[基因结构]</span><br><span class="line">C[FeatureAnalysis] --&gt; Cc[motif分析]</span><br><span class="line">C[FeatureAnalysis] --&gt; Cd[promotor seq analysis]</span><br><span class="line">B[FamilyGene.list] --&gt; D[phylogenetics]</span><br><span class="line">B[FamilyGene.list] --&gt; E[Group,Name]</span><br><span class="line">B[FamilyGene.list] --&gt; F[Expression]</span><br><span class="line">F[Expression] --&gt; G[表达, 功能验证]</span><br></pre></td></tr></table></figure>
          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201807_singleTcell_NSCLC/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201807_singleTcell_NSCLC/" itemprop="url">
                  Global characterization of T cells in non-small-cell lung cancer by single-cell sequencing
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2018-07-16 19:18:19" itemprop="dateCreated datePublished" datetime="2018-07-16T19:18:19+08:00">2018-07-16</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">更新于</span>
                
                <time title="修改时间：2018-07-31 19:30:40" itemprop="dateModified" datetime="2018-07-31T19:30:40+08:00">2018-07-31</time>
              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/read/" itemprop="url" rel="index"><span itemprop="name">read</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>肺癌T细胞 scRNAseq 分析</p>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/29942094" target="_blank" rel="noopener"><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/singleTcell_NSCLC.png" alt=""></a></p>
<h3 id="Methods"><a href="#Methods" class="headerlink" title="Methods"></a>Methods</h3><ol>
<li><p>we performed deep single-cell RNA sequencing on T cells isolated from <strong>tumor</strong>, <strong>adjacent normal tissues</strong> and <strong>peripheral blood</strong> for 14 treatment-naïve patients, including <strong>11 adenocarcinomas</strong> and <strong>three squamous cell carcinomas</strong>    </p>
</li>
<li><p>After confirming the existence of T cell infiltration in NSCLC tumors (Supplementary Fig. 1a), we sorted various T cell subtypes based on <strong>cell surface markers CD3/ CD4/CD8/CD25 by fluorescence-activated cell sorting</strong>    </p>
</li>
<li><p>A total of 12,346 cells were sequenced at an <strong>average depth of 1.04 million uniquely-mapped read pairs</strong> per cell</p>
<blockquote>
<p>The final expression table contained <strong>12,415 genes</strong> and <strong>11,769 cells</strong>    </p>
</blockquote>
</li>
<li><p>Dimensional reduction analysis (t-SNE) applied to the expression data revealed that T cells clustered primarily based on their tissue origins and subtypes      </p>
</li>
<li><p>To further address the intrinsic T cell heterogeneity, we applied unsupervised clustering based on t-SNE+ densityClust and identified seven CD8 and nine CD4 clusters    </p>
</li>
<li><p>To gain insight into the clonal relationship among individual T cells, we reconstructed their T cell receptor (TCR) sequences    </p>
<blockquote>
<p>We obtained full-length TCRs with both alpha and beta chains for 8,038 cells of the 16 clusters, including 5,015 harboring unique TCRs and 3,023 harboring repeated use of TCRs, indicative of clonal expansion    </p>
</blockquote>
</li>
<li><p>Detection of Treg marker genes    </p>
</li>
<li><p>Survival analysis. </p>
</li>
</ol>
<h3 id="Results"><a href="#Results" class="headerlink" title="Results"></a>Results</h3><ol>
<li>As well as tumor-infiltrating CD8+ T cells undergoing exhaustion, we observed two clusters of cells exhibiting states preceding exhaustion, and a high ratio of “pre-exhausted” to exhausted T cells was associated with better prognosis of lung adenocarcinoma.</li>
<li><p>we observed further heterogeneity within the tumor regulatory T cells (Tregs), characterized by the bimodal distribution of TNFRSF9, an activation marker for antigenspecific Tregs. The gene signature of those activated tumor Tregs, which included IL1R2, correlated with poor prognosis in lung adenocarcinoma. </p>
</li>
<li><p>…</p>
</li>
</ol>
<h3 id="points"><a href="#points" class="headerlink" title="points"></a>points</h3><ul>
<li>通常，CX3CR1在正常组织中表达高于肿瘤组织</li>
</ul>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201807_RNAedit/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201807_RNAedit/" itemprop="url">
                  identification of A-to-I RNA Editing
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2018-07-12 22:31:21" itemprop="dateCreated datePublished" datetime="2018-07-12T22:31:21+08:00">2018-07-12</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">更新于</span>
                
                <time title="修改时间：2018-07-31 19:30:16" itemprop="dateModified" datetime="2018-07-31T19:30:16+08:00">2018-07-31</time>
              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/RNAseq/" itemprop="url" rel="index"><span itemprop="name">RNAseq</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <ol>
<li><p>Quality Contorl of Fastq</p>
</li>
<li><p>Alignment to Reference Genome</p>
<p>Recommended use STAR-2pass to mapping reads to genome</p>
</li>
<li><p>Pre-process of Bam File</p>
<p>remove multi aglin, duplicate reads, splitN, and so on</p>
<p>Detailed operation can refer: <a href="https://software.broadinstitute.org/gatk/documentation/article.php?id=3891" target="_blank" rel="noopener">https://software.broadinstitute.org/gatk/documentation/article.php?id=3891</a></p>
</li>
<li><p>use GATK to call candidate site </p>
</li>
<li><p>Basic Variant filtering(optional)</p>
</li>
<li><p>Identification of editing sites</p>
<ul>
<li>Select single nucleotide variants(SNV) from all variant results , use snpEff to annot SNV</li>
<li>Select A&gt;T changes in the gene interval, <strong>depending on the direction of gene transcription</strong></li>
<li>Remove sites that appear in <a href="https://cancer.sanger.ac.uk/cosmic" target="_blank" rel="noopener">Cosmic</a>, <a href="https://www.ncbi.nlm.nih.gov/snp" target="_blank" rel="noopener">dbSNP</a>,  TCGA somatic mutations database</li>
<li>Remove sites which the editing level  is 1.0 in all sample</li>
<li>At least 10 reads in at least 1 sample</li>
<li>RNA editing sites is exists at least 2 sample</li>
</ul>
</li>
</ol>
<h3 id="ref："><a href="#ref：" class="headerlink" title="ref："></a>ref：</h3><ul>
<li>The Genomic Landscape and Clinical Relevance of A-to-I RNA Editing in Human Cancers </li>
<li>Dynamic regulation of RNA editing in human brain development and disease </li>
<li>A disrupted RNA editing balance mediated by ADARs (Adenosine DeAminases that act on RNA) in human hepatocellular carcinoma </li>
<li>ADAR1 promotes malignant progenitor reprogramming in chronic myeloid leukemia </li>
<li>Genome-wide analysis of A-to-I RNA editing by single-molecule sequencing in Drosophila </li>
</ul>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201807_16S/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201807_16S/" itemprop="url">
                  转:16S基本分析点
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2018-07-07 08:35:21" itemprop="dateCreated datePublished" datetime="2018-07-07T08:35:21+08:00">2018-07-07</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">更新于</span>
                
                <time title="修改时间：2018-07-31 19:29:47" itemprop="dateModified" datetime="2018-07-31T19:29:47+08:00">2018-07-31</time>
              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/microbe/" itemprop="url" rel="index"><span itemprop="name">microbe</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>8/23/2016 8:35 PM </p>
<hr>
<h2 id="基本介绍"><a href="#基本介绍" class="headerlink" title="基本介绍"></a>基本介绍</h2><h3 id="为什么选择16S测序？"><a href="#为什么选择16S测序？" class="headerlink" title="为什么选择16S测序？"></a>为什么选择16S测序？</h3><p>细菌中唯一的细胞器是核糖体，它的沉降系数是70S，由50S的大亚基和30S的小亚基组成，其中rRNA按沉降系数可分为5S、16S和23S三种，5S和23S rRNA在核糖体大亚基中，16S rRNA在核糖体小亚基中。5S rRNA的序列长度最长达400nt，信息量少；23S rRNA的序列长度长达2900nt，序列太长，测序通量、深度等要求高；而16S rRNA序列长度适中，约为1542nt。<br>沉降系数：离心法时，大分子沉降速度的量度，等于每单位离心场的速度。沉降系数10-13秒称为一个Svedberg单位，简写S，量纲为秒。</p>
<h3 id="16S测序区域如何选择？"><a href="#16S测序区域如何选择？" class="headerlink" title="16S测序区域如何选择？"></a>16S测序区域如何选择？</h3><p>16S rDNA是编码16S rRNA的DNA序列，存在于所有的细菌和古菌的基因组中，一般由9个保守区和9个可变区组成，保守区在细菌间无显著差异，可用于构建所有生命的统一进化树，而可变区在不同细菌中存在一定的差异，可将菌群鉴定精细到分类学上属，甚至种的级别。</p>
<h3 id="16S-rDNA测序区域与哪些因素有关？"><a href="#16S-rDNA测序区域与哪些因素有关？" class="headerlink" title="16S rDNA测序区域与哪些因素有关？"></a>16S rDNA测序区域与哪些因素有关？</h3><ul>
<li><a href="http://www.ncbi.nlm.nih.gov/pubmed/19816594" target="_blank" rel="noopener">引物序列</a>， 16S目的片段的引物是基于保守区的序列设计的，但是由于保守序列的碱基存在多态性，可能会忽略一部分微生物，因此要选用覆盖率高的引物进行目的片段的扩增；</li>
<li>测序平台， Illumina MiSeq/HiSeq测序平台的架构决定其短读长，它的读长最长为2*300pb，限制16S只能进行单V区、双V区或者三V区的测定；</li>
<li>目的区域中已知序列的多少， 扩增的目的片段与数据库中已知的片段进行比对，如果数据库中该区域已知片段很少，很不全面，导致测得的序列比对不上，微生物的多样性随之降低；</li>
<li><a href="https://www.researchgate.net/publication/260598361_Evaluation_of_the_RDP_Classifier_Accuracy_Using_16S_rRNA_Gene_Variable_Regions" target="_blank" rel="noopener">不同区域中鉴定物种的准确性</a>， 不同区域的物种比对到RDP数据库时，鉴定物种的准确性也是测序区域选择的约束条件。单个区域的检测发现，V2区和V4区在各个水平上的精确度最高，V5、V6、V7、V8区在门、纲水平上的精确度也较高，与V2区和V4区相似。</li>
</ul>
<h3 id="OTU是什么"><a href="#OTU是什么" class="headerlink" title="OTU是什么"></a>OTU是什么</h3><p>OTU（operational taxonomic units），即操作分类单元。通过一定的距离度量方法计算两两不同序列之间的距离度量或相似性，继而设置特定的分类阈值，获得同一阈值下的距离矩阵，进行聚类操作，形成不同的分类单元。</p>
<h3 id="OTU在16S测序中有何用"><a href="#OTU在16S测序中有何用" class="headerlink" title="OTU在16S测序中有何用"></a>OTU在16S测序中有何用</h3><p>高通量测序得到的16S序列有成千上万条，如果对每条序列都进行物种注释的话，工作量大、耗时长，而且16S扩增、测序等过程中出现的错误会降低结果的准确性。在16S分析中引入OTU，首先对相似性序列进行聚类，分成数量较少的分类单元，基于分类单元进行物种注释。这不仅简化工作量，提高分析效率，而且OTU在聚类过程中会去除一些测序错误的序列，提高分析的准确性。</p>
<h3 id="OTU如何聚类"><a href="#OTU如何聚类" class="headerlink" title="OTU如何聚类"></a>OTU如何聚类</h3><p>OTU聚类的方法多种多样，如Uclust、cd-hit、BLAST、mothur、usearch和 prefix/suffix，这些聚类方法均可以在QIIME软件中实施。不同聚类方法基于不同的算法，得到的聚类结果虽然不同，但是大体的聚类流程都是一致的，挑选非重复序列，与16S数据库比对，去除嵌合体序列，序列之间距离计算，97%相似OTU聚类。</p>
<blockquote>
<p>嵌合体序列是RCR扩增时，两条不同的序列产生杂交、扩增的序列</p>
</blockquote>
<h3 id="OTU跟物种的关系"><a href="#OTU跟物种的关系" class="headerlink" title="OTU跟物种的关系"></a>OTU跟物种的关系</h3><p>OTU聚类后，挑选出每个OTU中的代表序列，与RDP、Sliva或GreenGene等数据库进行比对，进行物种注释。OTU和物种是映射关系，它们一一对应或多对一关系</p>
<hr>
<h2 id="Alpha多样性"><a href="#Alpha多样性" class="headerlink" title="Alpha多样性"></a>Alpha多样性</h2><p><strong>Alpha多样性</strong>：指一个区域或生态系统内的多样性，用来描述单个样品的物种多样性；Alpha多样性指数主要用来表征三方面信息：物种丰度、物种多样性和测序深度。</p>
<ol>
<li>物种丰度</li>
</ol>
<ul>
<li><strong>observed_species指数</strong>：表示实际观测到的OTU数量；</li>
<li><strong>chao1指数</strong>：评估样品中所含OTU的总数。其公式为：<br><img src="http://i.imgur.com/9PBAXVy.png" alt=""><blockquote>
<p>Schao1：估计的OTU数量；<br>Sobs：实际观察到的OTU数量；<br>n1：只含一条序列的OTU的数量；<br>n2：含两条序列的OTU的数量。</p>
</blockquote>
</li>
</ul>
<ol start="2">
<li>物种多样性</li>
</ol>
<ul>
<li><p><strong>Shannon指数</strong>：包含着种数和各种间个体分配的均匀性两个部分。如果每一个体都属于不同的种，多样性指数就最大；如果每一个体都属于同一种，则其多样性指数就最小。用来估算微生物群落的多样性，shannon值越大，物种多样性越高。其计算公式：<br><img src="http://i.imgur.com/VGZejz7.png" alt=""></p>
<blockquote>
<p>Sobs：实际观察到的OTU数量；<br>ni：第i条OTU的序列数量；<br>N：所有的序列数。</p>
</blockquote>
</li>
<li><p><strong>Simpson指数</strong>：随机抽取的两个个体属于不同种的概率，Simpson指数越大，物种多样性越高。其计算公式：<br><img src="http://i.imgur.com/83QuQEB.png" alt=""></p>
<blockquote>
<p>Sobs：实际观察到的OTU数量；<br>ni：第i个OTU的序列数量；<br>N：所有的序列数。</p>
</blockquote>
</li>
</ul>
<p>3 . 测序深度</p>
<ul>
<li><strong>Coverage</strong>：反应测序深度，goods_coverage 指数越接近于1，说明测序深度已经基本覆盖到样品中所有的物种。其计算公式：<br><img src="http://i.imgur.com/zhXuYZM.png" alt=""><blockquote>
<p>Cdepth：goods_coverage指数表示测序深度；<br>n1：只有含一条序列的OTU数目；<br>N：为抽样中出现的总的序列数。</p>
</blockquote>
</li>
</ul>
<h3 id="如何展现Alpha多样性指数"><a href="#如何展现Alpha多样性指数" class="headerlink" title="如何展现Alpha多样性指数"></a>如何展现Alpha多样性指数</h3><p>Alpha多样性指数数值可以以表格的形式展现，还可以以图表的形式展现。随着抽取的reads条数的增加，曲线逐渐趋平，表明测序量是足够的，可以覆盖样品中的大部分微生物；如果曲线呈现上升趋势，则需要增加测序量，保障测序结果的代表性。简单来说，其实Alpha多样性指数就是衡量测序量是否足够的一个标准。它们大都长成这个样子。<br><img src="http://i.imgur.com/wSaLsXJ.png" alt=""></p>
<hr>
<h2 id="Beta多样性"><a href="#Beta多样性" class="headerlink" title="Beta多样性"></a>Beta多样性</h2><p><strong>Beta多样性</strong>：指不同生态系统之间的多样性比较，用来比较组间样品在物种多样性上存在的差异。Beta多样性是基于不同样品序列间的进化关系及丰度信息来计算样品间距离，用来描述不同样品间的相似性和差异性。<br>样本间距离是指样本之间的相似程度，可以通过数学方法估算。如前所述，样本间越相似，距离数值越小。计算微生物群体样本间距离的方法有多种，例如，Jaccard、Bray-Curtis、Unifrac等。这些距离算法主要分为两大类别：<strong>OTU间是否关联</strong>；<strong>OTU是否加权(表)</strong></p>
<table>
<thead>
<tr>
<th>\</th>
<th>基于独立OUT</th>
<th>基于系统发生数</th>
</tr>
</thead>
<tbody>
<tr>
<td>加权</td>
<td>Bray-Curtis</td>
<td>Weighted Unifrac</td>
</tr>
<tr>
<td>非加权</td>
<td>Jaccard</td>
<td>Unweighted Unifrac</td>
</tr>
</tbody>
</table>
<ul>
<li><p><strong>基于独立OTU vs 基于系统发生树</strong><br>二代测序当中，我们对16s rDNA某个区域进行测序后，会根据序列的相似度定义OTU。这个时候，基于独立OTU的计算方式认为OTU之间不存在进化上的联系，每个OTU间的关系平等。而基于系统发生树计算的方法，会根据16s的序列信息对OTU进行进化树分类，因此不同OTU之间的距离实际上有“远近”之分。</p>
</li>
<li><p><strong>加权 vs 非加权</strong><br>利用非加权的计算方法，主要考虑的是物种的有无，即如果两个群体的物种类型都一致，表示两个群体的β多样性最小。而加权方法，则同时考虑物种有无和物种丰度两个问题。如果A群体由3个物种a和2个物种b组成，B群体由2个物种a和3个物种b组成，则通过非加权方法计算，因为A群体与B群体的物种组成完全一致，都只由物种a和b组成，因此它们之间的β多样性为0。但通过加权方法计算，虽然A与B群体的组成一致，但物种a和b的数目却不同，因此两个群体的β多样性则并非一致。</p>
</li>
</ul>
<p>在宏基因组和16s测序的分析中，使用最多的距离算法主要有Bray-Curtis和Weighted 及Unweighted Unifrac。因此，下面我们就这几种常用的微生物多样性算法的特点和应用范围进行简单比较。</p>
<ul>
<li><p><strong>Bray-Curtis距离 vs Unifrac距离</strong><br>Bray-Curtis距离和Unifrac距离的主要区别在于计算β值的时候是否考虑OTU的进化关系。根据表2，显然，只有后者是有考虑。这会影响到它们的：<strong>数值表述意义不同</strong>：虽然两种方法的数值都是在0-1之间，但具体所表示的生物学意义却不一样。在Bray-Curtis算法中，0表示两个微生物群落的OTU结构（包括组成和丰度）完全一致；而在Unifrac中，0更侧重于表示两个群落的进化分类完全一致；<strong>实际应用的合理性</strong>：在实际微生物研究中，如果样本间物种的近源程度较高（温和处理样本与对照样本，生境相似的不同样本等），利用Bray-Curtis这种把OTU都同等对待的方法，更有利于发现样本间的差异。而Unifrac则更适合用于展示此类样本的重复性。</p>
</li>
<li><p><strong>Weighted Unifrac距离 vs Unweighted Unifrac距离</strong><br>Unifrac除了具有考虑OTU之间的进化关系的特点之外，根据有没有考虑OTU丰度的区别，Unifrac分析可以分为加权（WeightedUunifrac）和非加权（Unweighted Unifrac）两种方法。它们的不同在于：<strong>数值表述意义</strong>：Unweighted UniFrac只考虑了物种有无的变化，因此结果中，0表示两个微生物群落间OTU的种类一致。而Weighted UniFrac则同时考虑物种有无和物种丰度的变化，结果中的0则表示群落间OTU的种类和数量都一致。<strong>实际应用的合理性</strong>：在环境样本的检测中，由于影响因素复杂，群落间物种的组成差异更为剧烈，因此往往采用非加权方法进行分析。但如果要研究对照与实验处理组之间的关系，例如研究短期青霉素处理后，人肠道的菌落变化情况，由于处理后群落的组成一般不会发生大改变，但群落的丰度可能会发生大变化，因此更适合用加权方法去计算。</p>
</li>
</ul>
<h3 id="多样性分析方法"><a href="#多样性分析方法" class="headerlink" title="多样性分析方法"></a>多样性分析方法</h3><ol>
<li><p>Unifrac分析<br>基于系统进化，在 OTU水平上反应不同组间微生物群落结构的差异。若两个微生物群落完全相同，它们是没有各自独立的进化过程，UniFrac的值为0；UniFrac值越大，说明不同微生物群落在进化过程中变异越大，两个微生物群落在进化树中完全分开，则它们是两个完全独立的进化过程，UniFrac值最大，为1。<br><img src="http://i.imgur.com/2MsWX3u.jpg" alt=""></p>
<blockquote>
<p>图A和图B是Unifrac的两种展现形式，样品越靠近说明两个样品的组成越相似。</p>
</blockquote>
</li>
<li><p>PCoA分析<br>PCoA（principal coordinate analysis）主坐标分析，基于降维的方法，在尽可能保留原始信息的情况下，找出前两种影响分组的信息，研究其对样品分组的贡献率，将样品的相似性或差异性可视化，这两个坐标轴是没有任何实际意义的。分析结果展示如下，如果两个样品间距离较近，则表示这两个样品的物种组成比较相近。<br><img src="http://i.imgur.com/W4N2aEV.png" alt=""></p>
<blockquote>
<p>横坐标即第一主坐标，表示对样品分开的贡献率是60.91%；纵坐标即第二主坐标，表示对样品分开的贡献率是14.06%。箱线图的添加是本图的一个特色，直观地展现样品在第一主坐标和第二主坐标的分布情况。</p>
</blockquote>
</li>
<li><p>NMDS分析<br>NMDS与PCoA分析的意义一样，都 是用来展示样品间差异和相似的方法，如果两个样品距离较近，则表示这两个样品的物种组成比较相近。不同之处是两种展现形式是基于不同的算法。<br><img src="http://i.imgur.com/ZVl5Pwq.png" alt=""></p>
<blockquote>
<p>横纵坐标是基于进化或者数量距离矩阵的数值在二维表中成图，不同颜色代表不同分组，点代表样品，点与点之间的距离表示差异程度。</p>
</blockquote>
</li>
<li><p>Anosim分析<br>相似性分析Anosim分析是一种非参数检验，用来检验组间(两组或多组)的差异是否显著大于组内差异，从而判断分组是否有意义。首先利用Bray-Curtis算法计算两两样品间的距离，然后将所有距离从小到大进行排序。<br><img src="http://i.imgur.com/k5hcG1f.png" alt=""></p>
<blockquote>
<p>横坐标表示所有样品(Between)以及每个分组(A、B)，纵坐标表示unifrac距离的秩。Between组比其它每个分组的秩较高时，则表明组间差异大于组内差异。R介于(-1，1)之间，R大于0，说明组间差异显著；R小于0，说明组内差异大于组间差异，统计分析的可信度用P表示，P&lt; 0.05表示统计具有显著性。</p>
</blockquote>
</li>
<li><p>UPGMA层次聚类<br>假设在进化过程中所有核苷酸(氨基酸)的变异率相同，基于群落组成和结构的算法计算样本间的距离，根据β多样性距离矩阵进行层次聚类分析，最后通过UPGMA构建系统进化树。此树可以直观的反应样本间进化的差异。<br><img src="http://i.imgur.com/WJ9YvIq.png" alt=""></p>
<blockquote>
<p>UPGMA层次聚类分析。树枝不同颜色代表不同的分组。</p>
</blockquote>
</li>
</ol>
<h3 id="Beta多样性分析还有什么要点？"><a href="#Beta多样性分析还有什么要点？" class="headerlink" title="Beta多样性分析还有什么要点？"></a>Beta多样性分析还有什么要点？</h3><p>上面的Unifra、PCoA、NMDS、Anosim和UPGMA分析均可做考虑物种丰度（加权，Weight）和不考虑物种丰度（非加权，Unweight）的两种展现形式，可以根据你的研究目的选择合适的计算方法。</p>
<hr>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201807_matplotlib/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201807_matplotlib/" itemprop="url">
                  matplotlib 画图的 Qt 问题
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2018-07-02 19:19:21" itemprop="dateCreated datePublished" datetime="2018-07-02T19:19:21+08:00">2018-07-02</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">更新于</span>
                
                <time title="修改时间：2018-07-31 19:02:43" itemprop="dateModified" datetime="2018-07-31T19:02:43+08:00">2018-07-31</time>
              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/python/" itemprop="url" rel="index"><span itemprop="name">python</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <ol>
<li><p>This application failed to start because it could not find or load the Qt platform plugin “xcb” in “”.</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/Snipaste_180702_14-11-26.png" alt=""></p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">export QT_PLUGIN_PATH="/data/software/anaconda2/plugins/"</span><br></pre></td></tr></table></figure>
</li>
<li><p>QXcbConnection: Could not connect to display</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/Snipaste_180702_14-11-09.png" alt=""></p>
<p>方法一：</p>
<ul>
<li>设置环境变量</li>
</ul>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">export QT_QPA_PLATFORM='offscreen'</span><br></pre></td></tr></table></figure>
<p>方法二：</p>
<ul>
<li>python 画图脚本添加下面两行</li>
</ul>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> matplotlib</span><br><span class="line">matplotlib.use(<span class="string">'Agg'</span>)</span><br></pre></td></tr></table></figure>
</li>
</ol>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://chengcz.github.io/2018/201806_HumanGene/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="ChengCZ">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="ChengCZ's Notebook">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2018/201806_HumanGene/" itemprop="url">
                  canonical transcripts of Human protein coding gene
                </a>
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">

            
            
            

            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              

              
                
              

              <time title="创建时间：2018-06-30 23:33:23" itemprop="dateCreated datePublished" datetime="2018-06-30T23:33:23+08:00">2018-06-30</time>
            

            
              

              
                
                <span class="post-meta-divider">|</span>
                

                <span class="post-meta-item-icon">
                  <i class="fa fa-calendar-check-o"></i>
                </span>
                
                  <span class="post-meta-item-text">更新于</span>
                
                <time title="修改时间：2018-07-31 19:10:22" itemprop="dateModified" datetime="2018-07-31T19:10:22+08:00">2018-07-31</time>
              
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/bioinfor/" itemprop="url" rel="index"><span itemprop="name">bioinfor</span></a></span>

                
                
              
            </span>
          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <p>我们使用 <a href="https://chengcz.github.io/2018/201805_snpEff/">snpEff</a>, <a href="https://chengcz.github.io/2018/201806_ANNOVAR/">annovar</a> 对得到的变异集注释，总会得到多个转录本注释结果，而只需要按照 HGVS 规范报告一个经典转录本(canonical transcripts) 对应注释结果，这时候就需要知道每个基因对应的经典转录本。</p>
<p>一般情况下，我们使用<strong>基因对应的最长转录本代表该基因</strong>，具体操作上，首先从 <a href="https://chengcz.github.io/2018/201805_HGVS/">HGNC</a> 下载经过确认的基因对应经典转录本信息，其中不确定基因再使用最长转录本代表。</p>
<ol>
<li><p>HGNC 下载编码蛋白基因对应转录本</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> wget -c ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/locus_types/gene_with_protein_product.txt</span><br></pre></td></tr></table></figure>
<p>根据表格，很方便得到基因名字对应 hgnc_id, entrez_id, ensembl_gene_id, refseq_accession 等多数据库基因ID，使用 refseq_accession 进行 HGVS 表示，这里就只关注 refseq_accession ， 还是可以发现其中一些基因名并不能对应到唯一 refseq_accession，下面就需要我们进行第二步，结合 GFF 文件挑出经典转录本。</p>
</li>
<li><p>NCBI下载人类基因注释GFF文件</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> wget -c ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh37_latest/refseq_identifiers/GRCh37_latest_genomic.gff.gz</span><br></pre></td></tr></table></figure>
<p>使用脚本将 gff 转换为表格样式，方便筛选每个基因对应最长转录本，具体操作参考如下，其中 <a href="https://github.com/chengcz/pyGTF" target="_blank" rel="noopener">pyGTF</a> 为用来解析 gff 文件。</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/gff2tab.png" alt=""></p>
<p>输出文件可以<a href="https://github.com/chengcz/BioResource/tree/master/HumanGene" target="_blank" rel="noopener">下载</a> , 其具体格式为：</p>
<p><img src="http://or5pz0a28.bkt.clouddn.com/images/2018/gff_tab.png" alt=""></p>
<p>表头分别是：gene_id, gene_name, gene_type, transcript_id, transcript_name, transcript_type, transcript_length, chromosome, start, end, strand，根据 gene_name, transcript_name, transcript_length 三列排序，就可以很方便筛选出基因对应最长转录本。</p>
</li>
<li><p>脚本整合 HGNC 和 GFF 两部分结果</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"></span><br><span class="line">gene, length = &#123;&#125;, &#123;&#125;</span><br><span class="line">gff = pd.read_table(<span class="string">'hg37.geneinfo.txt'</span>, header=<span class="keyword">None</span>)</span><br><span class="line">gff = gff[(gff[<span class="number">2</span>]==<span class="string">'protein_coding'</span>) | (gff[<span class="number">5</span>]==<span class="string">'protein_coding'</span>)]</span><br><span class="line">nrow, ncol = gff.shape</span><br><span class="line"><span class="keyword">for</span> index <span class="keyword">in</span> range(nrow):</span><br><span class="line">    transcriptid = gff.iloc[index,<span class="number">3</span>]</span><br><span class="line">    <span class="keyword">try</span>:</span><br><span class="line">        transcriptid = transcriptid[:transcriptid.rindex(<span class="string">'.'</span>)]</span><br><span class="line">    <span class="keyword">except</span>:</span><br><span class="line">        <span class="keyword">pass</span></span><br><span class="line">    gene.setdefault(gff.iloc[index,<span class="number">1</span>], []).append(transcriptid)</span><br><span class="line">    length[transcriptid] = gff.iloc[index,<span class="number">6</span>]</span><br><span class="line"></span><br><span class="line">hgnc = pd.read_table(<span class="string">'gene_with_protein_product.txt'</span>)</span><br><span class="line">hgnc[<span class="string">'symbol'</span>] = hgnc[<span class="string">'symbol'</span>].astype(str)</span><br><span class="line">hgnc[<span class="string">'refseq_accession'</span>] = hgnc[<span class="string">'refseq_accession'</span>].astype(str)</span><br><span class="line">nrow, ncol = hgnc.shape</span><br><span class="line"><span class="keyword">for</span> index <span class="keyword">in</span> range(nrow):</span><br><span class="line">    genename = hgnc.loc[index, <span class="string">'symbol'</span>]</span><br><span class="line">    refseq = hgnc.loc[index, <span class="string">'refseq_accession'</span>]</span><br><span class="line">    tmp = refseq.split(<span class="string">'|'</span>)</span><br><span class="line">    <span class="keyword">if</span> len(tmp) == <span class="number">1</span>:</span><br><span class="line">        print(<span class="string">'&#123;&#125;\t&#123;&#125;\thgnc'</span>.format(genename, refseq))</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="keyword">if</span> len(tmp) == <span class="number">0</span>:</span><br><span class="line">            tmp = gene[genename]</span><br><span class="line">        tmp = sorted(tmp, key=<span class="keyword">lambda</span> x: length.get(x, <span class="number">0</span>), reverse=<span class="keyword">True</span>)</span><br><span class="line">        print(<span class="string">'&#123;&#125;\t&#123;&#125;\tgff'</span>.format(genename, tmp[<span class="number">0</span>]))</span><br></pre></td></tr></table></figure>
<p>结果整理文件可下载：<a href="https://github.com/chengcz/BioResource/tree/master/HumanGene/protein_coding_gene_canonical_transcripts.txt" target="_blank" rel="noopener">https://github.com/chengcz/BioResource/tree/master/HumanGene/protein_coding_gene_canonical_transcripts.txt</a></p>
</li>
</ol>

          
        
      
    </div>

    

    
    
    

    

    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  
  <nav class="pagination">
    <span class="page-number current">1</span><a class="page-number" href="/page/2/">2</a><span class="space">&hellip;</span><a class="page-number" href="/page/4/">4</a><a class="extend next" rel="next" href="/page/2/"><i class="fa fa-angle-right" aria-label="下一页"></i></a>
  </nav>



          </div>
          

        </div>
        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<div class="powered-by">
  <i class="fa fa-user-md"></i>
  <span id="busuanzi_container_site_uv">
    本站访客数:<span id="busuanzi_value_site_uv"></span>
  </span>
  <span id="busuanzi_container_site_pv">
    本站总访问量:<span id="busuanzi_value_site_pv"></span>
  </span>
</div>

<div class="copyright">&copy; <span itemprop="copyrightYear">2018</span>
  <span class="with-love" id="animate">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">ChengCZ</span>

  

  
</div>



<!--

  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动 v3.7.1</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://theme-next.org">NexT.Mist</a> v6.3.0</div>


-->

        








        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    
	
    

    
  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>


























  
  
    <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  


  


  <script type="text/javascript" src="/js/src/utils.js?v=6.3.0"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=6.3.0"></script>



  
  

  

  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=6.3.0"></script>



  



  










  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  

  

  
  

  

  

  

  

  

</body>
</html>
