<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  <title>ChengCZ&#39;s Notebook</title>
  <meta name="viewport" content="width=device-width">
  <meta property="og:type" content="website">
<meta property="og:title" content="ChengCZ&#39;s Notebook">
<meta property="og:url" content="http://chengcz.github.io/index.html">
<meta property="og:site_name" content="ChengCZ&#39;s Notebook">
<meta property="og:locale" content="zh-CN">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="ChengCZ&#39;s Notebook">
  
    <link rel="alternative" href="/atom.xml" title="ChengCZ&#39;s Notebook" type="application/atom+xml">
  
  
    <link rel="icon" href="/favicon.png">
  
  <link rel="stylesheet" href="/css/style.css">
  <!--[if lt IE 9]><script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script><![endif]-->
  
</head>
<body>
  <div id="container">
    <div class="mobile-nav-panel">
	<i class="icon-reorder icon-large"></i>
</div>
<header id="header">
	<h1 class="blog-title">
		<a href="/">ChengCZ&#39;s Notebook</a>
	</h1>
	<nav class="nav">
		<ul>
			<li><a href="/">Home</a></li><li><a href="/archives">Archives</a></li><li><a href="/categories">Categories</a></li><li><a href="/about">About</a></li>
			<li><a id="nav-search-btn" class="nav-icon" title="Search"></a></li>
			<li><a href="/atom.xml" id="nav-rss-link" class="nav-icon" title="RSS Feed"></a></li>
		</ul>
	</nav>
	<div id="search-form-wrap">
		<form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="Search"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://chengcz.github.io"></form>
	</div>
</header>
    <div id="main">
      
  
    <article id="post-20180515_STAR_Fusion" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2018/20180515_STAR_Fusion/" class="article-date">
  <time datetime="2018-05-15T14:14:52.000Z" itemprop="datePublished">2018-05-15</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2018/20180515_STAR_Fusion/">使用 STAR-fusion 分析 Fusion-Gene</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>软件文档：<a href="https://github.com/STAR-Fusion/STAR-Fusion/wiki" target="_blank" rel="noopener">https://github.com/STAR-Fusion/STAR-Fusion/wiki</a></p>
<p>##软件安装</p>
<ol>
<li>STAR，<a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener">https://github.com/alexdobin/STAR</a></li>
<li>运行依赖 perl 模块<ol>
<li>DB_File</li>
<li>URI::Escape</li>
<li>Set::IntervalTree</li>
<li>Carp::Assert</li>
<li>JSON::XS</li>
<li>PerlIO::gzip</li>
<li>common::sense</li>
<li>Types::Serialiser</li>
<li>Canary::Stability</li>
</ol>
</li>
</ol>
<p>##数据库准备</p>
<p>下载一个较小的未处理的参考文件，自己运行 index 命令。要是网速够快也可以直接在 index 好的数据库文件，~27G</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">wget -c https://data.broadinstitute.org/Trinity/CTAT_RESOURCE_LIB/GRCh37_v19_CTAT_lib_Feb092018.source_data.tar.gz</span><br><span class="line"><span class="meta">$</span>STAR_FUSION_HOME/FusionFilter/prep_genome_lib.pl \</span><br><span class="line">    --genome_fa ref_genome.fa \</span><br><span class="line">    --gtf ref_annot.gtf \</span><br><span class="line">    --fusion_annot_lib CTAT_HumanFusionLib.v0.1.0.dat.gz \</span><br><span class="line">    --annot_filter_rule AnnotFilterRule.pm \</span><br><span class="line">    --pfam_db PFAM.domtblout.dat.gz</span><br></pre></td></tr></table></figure>
<p>##运行STAR-Fusion</p>
<p><strong>STAR-Fusion</strong>  对 STAR 输出的嵌合比对分析发现可能存在的基因融合事件</p>
<h3 id="从-fastq-文件开始"><a href="#从-fastq-文件开始" class="headerlink" title="从 fastq 文件开始"></a>从 fastq 文件开始</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span>STAR_FUSION_HOME/STAR-Fusion \</span><br><span class="line">    --genome_lib_dir /path/to/your/CTAT_resource_lib \</span><br><span class="line">    --left_fq reads_1.fq \</span><br><span class="line">    --right_fq reads_2.fq \</span><br><span class="line">    --output_dir star_fusion_outdir \</span><br><span class="line">    --no_remove_dups</span><br></pre></td></tr></table></figure>
<h3 id="从-STAR-产生-Bam-文件开始"><a href="#从-STAR-产生-Bam-文件开始" class="headerlink" title="从 STAR 产生 Bam 文件开始"></a>从 STAR 产生 Bam 文件开始</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">STAR --genomeDir $&#123;star_index_dir&#125; \</span><br><span class="line">    --readFilesIn $&#123;left_fq_filename&#125; $&#123;right_fq_filename&#125; \</span><br><span class="line">    --twopassMode Basic \</span><br><span class="line">    --outReadsUnmapped None \</span><br><span class="line">    --chimSegmentMin 12 \</span><br><span class="line">    --chimJunctionOverhangMin 12 \</span><br><span class="line">    --alignSJDBoverhangMin 10 \</span><br><span class="line">    --alignMatesGapMax 100000 \</span><br><span class="line">    --alignIntronMax 100000 \</span><br><span class="line">    --chimSegmentReadGapMax 3 \</span><br><span class="line">    --alignSJstitchMismatchNmax 5 -1 5 5 \</span><br><span class="line">    --runThreadN $&#123;THREAD_COUNT&#125; \</span><br><span class="line">    --outSAMstrandField intronMotif</span><br><span class="line"></span><br><span class="line">STAR-Fusion --genome_lib_dir /path/to/your/CTAT_resource_lib \</span><br><span class="line">             -J Chimeric.out.junction \</span><br><span class="line">             --output_dir star_fusion_outdir</span><br></pre></td></tr></table></figure>
<h2 id="输出文件"><a href="#输出文件" class="headerlink" title="输出文件"></a>输出文件</h2><p>STAR 速度还是那么让人惊喜，6m reads不到半小时。 融合结果<code>star-fusion.fusion_predictions.abridged.tsv</code></p>
<blockquote>
<p><strong>FusionName</strong>,<br><strong>JunctionReadCount</strong>, split align到融合点的序列片段数<br><strong>SpanningFragCount</strong>, 双端reads跨越融合点的序列片段数<br><strong>SpliceType</strong>, 断点是否在注释文件存在<br><strong>LeftGene</strong>,<br><strong>LeftBreakpoint</strong>,<br><strong>RightGene</strong>,<br><strong>RightBreakpoint</strong>,<br><strong>LargeAnchorSupport</strong>,<br><strong>FFPM</strong>, fusion fragments per million total reads<br><strong>LeftBreakDinuc</strong>,<br><strong>LeftBreakEntropy</strong>,<br><strong>RightBreakDinuc</strong>,<br><strong>RightBreakEntropy</strong>,<br><strong>annots</strong>,</p>
</blockquote>
<h2 id="结果比较"><a href="#结果比较" class="headerlink" title="结果比较"></a>结果比较</h2><p>真是一个悲伤的故事，根据官网给的两种运行方式，结果差别这么大。查看 STAR-Fusion 脚本，使用的 mapping 参数差异有点大啊，哪一个比较合理呢（一个新坑）？？？</p>
<p><img src="./images/2018/star_fusion.results.png" alt=""></p>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/bioinfor/">bioinfor</a>
  </div>

			</span>
			<span class="tags">
				
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/Fusion-Gene/">Fusion Gene</a></li></ul>

			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-20180509_snpEff" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2018/20180509_snpEff/" class="article-date">
  <time datetime="2018-05-09T13:45:52.000Z" itemprop="datePublished">2018-05-09</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2018/20180509_snpEff/">VCF 注释 —— SnpEff</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>日常工作之读文档，了解参数改变、新方法功能，对已有流程进行修改。</p>
<p>各种文档看一边遍一遍，使用需要又一次看 SnpEff （<a href="http://snpeff.sourceforge.net/index.html" target="_blank" rel="noopener">http://snpeff.sourceforge.net/index.html</a> ），进行一些整理如下，不过还是建议<a href="http://snpeff.sourceforge.net/SnpEff_manual.html" target="_blank" rel="noopener">原文</a>，鬼知道我会在哪里写错理解错。</p>
<h3 id="运行"><a href="#运行" class="headerlink" title="运行"></a>运行</h3><ol>
<li><p>annotation</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">java -jar snpEff.jar eff genome	input.vcf &gt;output.vcf</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> -hgvs, 默认开启，使用 HGVS 变异描述语法对蛋白替换进行描述</span><br><span class="line"><span class="meta">#</span> -fi, 提供bed文件，只对bed区域内位点进行注释</span><br><span class="line"><span class="meta">#</span> -no-downstream, -no-intergenic, -no-intron, -no-upstream, -no-utr, -no [effectType], 参数还可对根据注释类型对注释进行初过滤</span><br><span class="line"><span class="meta">#</span> -canon, 只使用基因对应最长转录本（作为权威转录本）进行注释；参数-canonList可以自行提供基因权威转录本，格式为“GeneID    transcript_ID”</span><br><span class="line"><span class="meta">#</span> -noStats, 不输出注释统计信息</span><br><span class="line"><span class="meta">#</span> -v, 话痨模式，输出各种运行信息</span><br></pre></td></tr></table></figure>
<p>​</p>
</li>
<li><p>database</p>
<p>人、小鼠等物种都有已经 build 的数据库，可以直接下载，链接：<a href="https://sourceforge.net/projects/snpeff/files/databases/v4_3/" target="_blank" rel="noopener">https://sourceforge.net/projects/snpeff/files/databases/v4_3/</a>，注释选择基因组版本。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">java -jar snpEff.jar databases    # 查看可用数据库</span><br><span class="line">java -jar snpEff.jar download -v GRCh37.75    # 下载数据库 GRCh37.75</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> Q, 使用执行程序下载数据库时容易断，一般建议手动下载，或者wget等其他方式</span><br></pre></td></tr></table></figure>
<p>软件还提供了自行构建数据库，详细用法如下：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> 添加配置信息</span><br><span class="line">echo 'GRCh37.70.genome : Human ' &gt;&gt;~/snpEff/snpEff.config</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> Create directoy for this new genome</span><br><span class="line">mkdir -p ~/snpEff/data/GRCh37.70 &amp;&amp; cd ~/snpEff/data/GRCh37.70</span><br><span class="line"><span class="meta">#</span> Get annotation files</span><br><span class="line">wget -c -O genes.gtf.gz ftp://ftp.ensembl.org/pub/release-70/gtf/homo_sapiens/Homo_sapiens.GRCh37.70.gtf.gz</span><br><span class="line"><span class="meta">#</span> Get the genome</span><br><span class="line">wget -c -O sequences.fa.gz ftp://ftp.ensembl.org/pub/release-70/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.70.dna.toplevel.fa.gz</span><br><span class="line"><span class="meta">#</span> Optional, Download CDSs</span><br><span class="line">wget -c -O cds.fa.gz ftp://ftp.ensembl.org/pub/release-70/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh37.70.cdna.all.fa.gz</span><br><span class="line"><span class="meta">#</span> Optional, Download proteins</span><br><span class="line">wget -c -O protein.fa.gz ftp://ftp.ensembl.org/pub/release-70/fasta/homo_sapiens/pep/Homo_sapiens.GRCh37.70.pep.all.fa.gz</span><br><span class="line"><span class="meta">#</span> Optional, Download regulatory annotations</span><br><span class="line">wget -c -O regulation.gff.gz ftp://ftp.ensembl.org/pub/release-70/regulation/homo_sapiens/AnnotatedFeatures.gff.gz</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> Building a database from GFF files</span><br><span class="line">cd ~/snpEff &amp;&amp; java -jar snpEff.jar build -gff3 -v GRCh37.70</span><br><span class="line"><span class="meta">#</span> rm -r ~/snpEff/data/GRCh37.70</span><br></pre></td></tr></table></figure>
<p>此外，snpEff 构建数据库还支持 RefSeq, GenBank 等多种数据输入格式，具体用法查看原文档。</p>
<p><strong>注意物种 codon 选择</strong>，codon 可以在 <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi" target="_blank" rel="noopener">https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi</a> 中查看，codon可以在<code>~/snpEff/snpEff.config</code>文件中设置：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># 设置特异 codon</span><br><span class="line">codon.Invertebrate_Mitochondrial:  TTT/F, TTC/F, TAC/Y, TAA/*, ATG/M+, ATG/M+, ACT/T, ...</span><br><span class="line"># 物种选择恰当的 codontable</span><br><span class="line">dm3.M.codonTable : Invertebrate_Mitochondrial    # the chromosome &apos;M&apos; from fly genome (dm3) uses Invertebrate_Mitochondrial codon table</span><br></pre></td></tr></table></figure>
<p>​</p>
</li>
<li><p>dump</p>
<p>从数据库中提取出注释信息</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">java -jar snpEff.jar dump -v -bed GRCh37.70 &gt; GRCh37.70.bed</span><br><span class="line"><span class="meta">#</span> [-bed |-txt ] 设置数据格式</span><br><span class="line"><span class="meta">#</span> [-0 |-1 ] 设置坐标系统，0-based 或者 1-based</span><br></pre></td></tr></table></figure>
<p>​</p>
</li>
</ol>
<h3 id="注释结果"><a href="#注释结果" class="headerlink" title="注释结果"></a>注释结果</h3><p><img src="./images/2018/snpEff.png" alt=""></p>
<p>可以看到注释信息被添加到了 VCF 中每个变异 INFO 信息中，以 ANN= 特征开始，详细的注释说明可查看<a href="./appendix/snpEff.pdf">官方文档</a> <a href="http://snpeff.sourceforge.net/VCFannotationformat_v1.0.pdf" target="_blank" rel="noopener">http://snpeff.sourceforge.net/VCFannotationformat_v1.0.pdf</a>。因为基因多个转录本、相互重叠基因等原因，可以看到变异位点被多次注释，需根据实际需求进行筛选。</p>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/bioinfor/">bioinfor</a>
  </div>

			</span>
			<span class="tags">
				
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/annotation/">annotation</a></li></ul>

			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-20170927_umount" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2017/20170927_umount/" class="article-date">
  <time datetime="2017-09-27T03:19:21.000Z" itemprop="datePublished">2017-09-27</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2017/20170927_umount/">卸载一个假忙的硬盘</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<ol>
<li><p><code>df -h</code>查看卸载硬盘信息</p>
</li>
<li><p><code>umount /dev/sdf</code>，停止使用的硬盘假装忙碌状态</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">umount: /mnt/usbhd2: device is busy.</span><br><span class="line">        (In some cases useful info about processes that use</span><br><span class="line">         the device is found by lsof(8) or fuser(1))</span><br></pre></td></tr></table></figure>
</li>
<li><p><code>fuser /dev/sdf</code>，查看占用硬盘的进程</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">/mnt/sdf:         106657c</span><br></pre></td></tr></table></figure>
</li>
<li><p><code>kill -9 106657</code></p>
</li>
<li><p>重新执行<code>umount /dev/sdf</code></p>
</li>
</ol>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/linux/">linux</a>
  </div>

			</span>
			<span class="tags">
				
			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-20170901_submit_data_to_sra" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2017/20170901_submit_data_to_sra/" class="article-date">
  <time datetime="2017-09-01T13:29:21.000Z" itemprop="datePublished">2017-09-01</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2017/20170901_submit_data_to_sra/">提交测序数据到SRA</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>九月一号开学了，暑假作业写完了没有？但是跟我并没有关系，不开学，没作业，但是好多事情啊，，，最近提交了一些数据到SRA，每次过程总是那么曲折离奇，对一些遇到的问题进行整理填坑吧。</p>
<p>简单说，申请<strong>BioProject</strong>，申请<strong>BioSample</strong>，填写<strong>SRA_metadata</strong>，上传数据，虽然过程清晰，坑还没填完，仍在邮件NCBI，过程供参考。</p>
<h2 id="1-BioProject"><a href="#1-BioProject" class="headerlink" title="1. BioProject"></a>1. <a href="https://submit.ncbi.nlm.nih.gov/subs/bioproject/" target="_blank" rel="noopener">BioProject</a></h2><p>申请一个项目编号，填写submitter 信息，选择项目类型（Raw sequence reads），物种名称，释放时间，项目描述，BioSample和文章信息（跳过，跳过，，），核对提交</p>
<h2 id="2-BioSample"><a href="#2-BioSample" class="headerlink" title="2. BioSample"></a>2. <a href="https://submit.ncbi.nlm.nih.gov/subs/biosample/" target="_blank" rel="noopener">BioSample</a></h2><p>申请一个样品编号，submitter 信息，释放时间，样品类型，<strong>填写样品属性表格</strong>，核对提交</p>
<h2 id="3-SRA"><a href="#3-SRA" class="headerlink" title="3. SRA"></a>3. <a href="https://submit.ncbi.nlm.nih.gov/subs/sra/" target="_blank" rel="noopener">SRA</a></h2><p>主要就填写<strong>SRA_metadata</strong>，测序平台，文库类型等等，选择批量提交，填写表格注意看要求啊，除了样品名称、登录号等信息，其他属性组合每行也要是唯一的。</p>
<h2 id="4-submit-data"><a href="#4-submit-data" class="headerlink" title="4. submit data"></a>4. submit data</h2><p>看数据大小，选择合适上传方法吧，小于2G用aspera浏览器插件，大数据选择ftp或者aspera命令行。这几次都用了aspera cmd上传，霸占几乎全部网速，很快上传好</p>
<p>具体命令为：<code>ascp -i &lt;path/to/key_file&gt; -QT -l100m -k1 -d &lt;path/to/folder/containing files&gt; subasp@upload.ncbi.nlm.nih.gov:uploads/XXX@xxx.com_g3O1FgOE</code>， 其中key必须是全路径，-d接包括原始数据的文件夹（<strong>不再包含其他文件夹</strong>），<a href="mailto:XXX@xxx.com" target="_blank" rel="noopener">XXX@xxx.com</a>为注册账号邮箱</p>
<p>上传结束后，SRA中选择上传数据文件夹，，，</p>
<blockquote>
<p><em>好了，装逼结束，文件没仔细检查，传了两不完整的fastq，发邮件给ncbi沟通呢，md5值也还没有用到，不知什么情况，，，too young, too naive…</em></p>
</blockquote>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/bioinfor/">bioinfor</a>
  </div>

			</span>
			<span class="tags">
				
			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-20170828_new_object" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2017/20170828_new_object/" class="article-date">
  <time datetime="2017-08-28T14:47:06.000Z" itemprop="datePublished">2017-08-28</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2017/20170828_new_object/">python class实现fasta文件解析</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>没有对象怎么办（object）？？？当然是new一个啊</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">Sequence</span><span class="params">(object)</span>:</span></span><br><span class="line">    <span class="comment"># 使用__init__对Sequence进行初始化，Sequence对象具有三个属性</span></span><br><span class="line">    <span class="function"><span class="keyword">def</span> <span class="title">__init__</span><span class="params">(self, name, seq, descr=<span class="string">''</span>)</span>:</span></span><br><span class="line">        self.name = name</span><br><span class="line">        self.seq = seq</span><br><span class="line">        self.descr = descr</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 给Sequence对象定义一个反向互补方法</span></span><br><span class="line">    <span class="function"><span class="keyword">def</span> <span class="title">reverse_complement</span><span class="params">(self)</span>:</span></span><br><span class="line">        RC = &#123;<span class="string">'A'</span>:<span class="string">'T'</span>,<span class="string">'T'</span>:<span class="string">'A'</span>,<span class="string">'C'</span>:<span class="string">'G'</span>,<span class="string">'G'</span>:<span class="string">'C'</span>,<span class="string">'N'</span>:<span class="string">'N'</span>,</span><br><span class="line">              <span class="string">'a'</span>:<span class="string">'t'</span>,<span class="string">'t'</span>:<span class="string">'a'</span>,<span class="string">'c'</span>:<span class="string">'g'</span>,<span class="string">'g'</span>:<span class="string">'c'</span>,<span class="string">'n'</span>:<span class="string">'n'</span>&#125;</span><br><span class="line">        <span class="keyword">return</span> Sequence(self.name, <span class="string">''</span>.join([RC[i] <span class="keyword">for</span> i <span class="keyword">in</span> self.seq[::<span class="number">-1</span>]]), self.descr)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 定义输出到文件的方法，接受文件句柄输入</span></span><br><span class="line">    <span class="function"><span class="keyword">def</span> <span class="title">write_to_fasta_file</span><span class="params">(self, file_handle)</span>:</span></span><br><span class="line">        <span class="function"><span class="keyword">def</span> <span class="title">_SeqFormat</span><span class="params">(seq, chara=<span class="number">80</span>)</span>:</span></span><br><span class="line">            tmp = <span class="string">''</span></span><br><span class="line">            <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">0</span>,len(seq),chara):</span><br><span class="line">                tmp += (seq[i:(i+chara)]+<span class="string">'\n'</span>)</span><br><span class="line">            <span class="keyword">return</span> tmp</span><br><span class="line">        file_handle.write(<span class="string">'&gt;&#123;&#125; &#123;&#125;\n'</span>.format(self.name, self.descr))</span><br><span class="line">        file_handle.write(<span class="string">'&#123;&#125;'</span>.format(_SeqFormat(self.seq)))</span><br><span class="line"></span><br><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">FastaReader</span><span class="params">(object)</span>:</span></span><br><span class="line">    <span class="function"><span class="keyword">def</span> <span class="title">__init__</span><span class="params">(self, fastafile)</span>:</span></span><br><span class="line">        self.fasta = fastafile</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 使用__iter__方法实现对fasta文件的循环解析，其中用到yeild构造了生成器</span></span><br><span class="line">    <span class="function"><span class="keyword">def</span> <span class="title">__iter__</span><span class="params">(self)</span>:</span></span><br><span class="line">        <span class="keyword">with</span> open(self.fasta) <span class="keyword">as</span> f:</span><br><span class="line">            seq = <span class="keyword">None</span></span><br><span class="line">            <span class="keyword">for</span> line <span class="keyword">in</span> f:</span><br><span class="line">                <span class="keyword">if</span> line.startswith( <span class="string">"&gt;"</span> ):</span><br><span class="line">                    <span class="keyword">if</span> seq:</span><br><span class="line">                        <span class="keyword">yield</span> Sequence( name, seq, descr )</span><br><span class="line">                    name, sep, descr = line[<span class="number">1</span>:<span class="number">-1</span>].partition(<span class="string">' '</span>)</span><br><span class="line">                    seq = <span class="string">""</span></span><br><span class="line">                <span class="keyword">else</span>:</span><br><span class="line">                    <span class="keyword">assert</span> seq <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>, <span class="string">"FASTA file does not start with '&gt;'."</span></span><br><span class="line">                    seq += line[:<span class="number">-1</span>].encode()</span><br><span class="line">        <span class="keyword">if</span> seq <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            <span class="keyword">yield</span> Sequence( name, seq, descr )</span><br><span class="line"></span><br><span class="line"><span class="comment"># example</span></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> FastaReader(<span class="string">'ref.fa'</span>):</span><br><span class="line">    <span class="comment"># 使用</span></span><br><span class="line">    i.write_to_fasta_file(file_handle)</span><br><span class="line">    i.seq</span><br><span class="line">    i.name</span><br><span class="line">    i.descr</span><br></pre></td></tr></table></figure>
<p>懒癌发作了，就这，，，</p>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/python/">python</a>
  </div>

			</span>
			<span class="tags">
				
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/fasta/">fasta</a></li></ul>

			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-20170728_mount_HardDisk" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2017/20170728_mount_HardDisk/" class="article-date">
  <time datetime="2017-07-28T02:20:21.000Z" itemprop="datePublished">2017-07-28</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2017/20170728_mount_HardDisk/">服务器挂载4T硬盘</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>因为一些我并不理解的原因，服务器不能正确识别4T的NTFS移动硬盘，需要对其格式化成ext4格式，就我的一些操作记录。</p>
<ol>
<li><p><code>fdisk -l</code>，查看硬盘分配的硬件编号</p>
<p><img src="./images/2017/fdisk.png" alt=""></p>
</li>
<li><p><code>parted /dev/sdg</code>，对使用parted工具对硬盘进行分区</p>
<ul>
<li><p><code>mklable gpt</code>，把磁盘格式化为 gpt分区表</p>
</li>
<li><p><code>mkpart primary 0 4000G</code>，创建一个主分区，起始位置-结束位置</p>
</li>
<li><p><code>mkpart extended 2000G 4000G</code>，可选，创建一个扩展分区</p>
</li>
<li><p><code>print</code>，打印出当前设备的信息</p>
</li>
<li><p><code>quit</code>，推出parted</p>
</li>
</ul>
<p><img src="./images/2017/linux_parted.png" alt=""></p>
</li>
<li><p><code>mkfs.ext4 /dev/sdg1</code>，对硬盘进行格式化，如果存在多个分区，每个分区都需要进行格式化</p>
<p><img src="./images/2017/mkfs.ext4.png" alt=""></p>
</li>
<li><p><code>mount /dev/sdg1 /mnt/usbmount1</code>，挂载硬盘到服务器</p>
</li>
<li><p><code>umount /dev/sdg1</code>，硬盘卸载</p>
</li>
</ol>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/linux/">linux</a>
  </div>

			</span>
			<span class="tags">
				
			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-20170619_python_dict" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2017/20170619_python_dict/" class="article-date">
  <time datetime="2017-06-19T08:38:01.000Z" itemprop="datePublished">2017-06-19</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2017/20170619_python_dict/">python字典</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>字典也是哈希表，键(key)必须是唯一的可哈希对象，但值(value)则不必。python中键值间使用<strong>“:”</strong>分割，不同键值对间使用<strong>“,”</strong>分割，整个字典包括在<strong>“{}”</strong>中。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#初始化定义</span></span><br><span class="line">seq = &#123;&#125;    <span class="comment">#seq = dict()</span></span><br><span class="line">seq = &#123;<span class="string">'key1'</span>:<span class="string">'value1'</span>, <span class="string">'key2'</span>:<span class="string">'value2'</span>&#125;</span><br></pre></td></tr></table></figure>
<h2 id="设置字典默认值"><a href="#设置字典默认值" class="headerlink" title="设置字典默认值"></a>设置字典默认值</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#key不存在时，返回自定义value</span></span><br><span class="line">seq.get(key, value)</span><br><span class="line"></span><br><span class="line"><span class="comment">#key不存在时，设置key对应键值为list；key存在时返回seq[key]</span></span><br><span class="line">seq.setdefault(key,[])</span><br><span class="line"></span><br><span class="line"><span class="comment">#设置默认字典，读取seq[key]时进行初始化</span></span><br><span class="line"><span class="keyword">from</span> collections <span class="keyword">import</span> defaultdict</span><br><span class="line">seq = defaultdict(list)</span><br></pre></td></tr></table></figure>
<ul>
<li><p>树结构</p>
 <figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> collections <span class="keyword">import</span> defaultdict</span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">tree</span><span class="params">()</span>:</span> <span class="keyword">return</span> defaultdict(tree)</span><br><span class="line"></span><br><span class="line">users = tree()</span><br><span class="line">users[<span class="string">'codingpy'</span>][<span class="string">'username'</span>] = <span class="string">'earlgrey'</span></span><br><span class="line"><span class="comment">#甚至还能不赋值</span></span><br><span class="line">users[<span class="string">'Python'</span>][<span class="string">'Standard Library'</span>][<span class="string">'os'</span>]</span><br></pre></td></tr></table></figure>
</li>
</ul>
<h2 id="计算list中某元素出现次数"><a href="#计算list中某元素出现次数" class="headerlink" title="计算list中某元素出现次数"></a>计算list中某元素出现次数</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 方法一</span></span><br><span class="line">list.count(<span class="string">'X'</span>)    <span class="comment"># int</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 方法二</span></span><br><span class="line"><span class="keyword">from</span> collections <span class="keyword">import</span> Counter</span><br><span class="line">Counter(list)[<span class="string">'X'</span>]    <span class="comment"># dict['X'], int</span></span><br></pre></td></tr></table></figure>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/python/">python</a>
  </div>

			</span>
			<span class="tags">
				
			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-20170521_linux_find" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2017/20170521_linux_find/" class="article-date">
  <time datetime="2017-05-21T04:39:31.000Z" itemprop="datePublished">2017-05-21</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2017/20170521_linux_find/">Linux命令——find</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>find，Linux常用文件搜索命令简要整理</p>
<h1 id="基本用法"><a href="#基本用法" class="headerlink" title="基本用法"></a>基本用法</h1><p>find基本用法如：<code>find [path] &lt;search regular&gt; [action]</code></p>
<ul>
<li>可选参数path，支持空格分隔的多路径查找，默认为当前路径</li>
<li>可选参数action<ul>
<li>-print，匹配文件输出到stdout，默认</li>
<li>-exec，对find匹配到的文件执行该参数所给出的shell命令，格式为<strong>command { } \;</strong>，其中{}表示匹配文件</li>
<li>-ok，于-exec类似，执行前给出提示是否执行</li>
</ul>
</li>
</ul>
<ol>
<li><p>文件名</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">find [path] -name &lt;name&gt;</span><br><span class="line">find [path] -iname &lt;name&gt;    #不区分大小写</span><br><span class="line"><span class="meta">#</span> 其中&lt;name&gt;支持通配符，一般使用时建议增加单引号</span><br></pre></td></tr></table></figure>
</li>
<li><p>文件类型，一般配合其他搜索参数一起使用</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">find [path] -type [bdcplf]</span><br><span class="line"><span class="meta">#</span> 其中d表示目录，l表示链接，f表示文件</span><br></pre></td></tr></table></figure>
</li>
<li><p>文件大小</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> -size 参数支持+-表示大于小于，ckMG表示数据单位</span><br><span class="line">find [path] -size +10M    #path下大于10M的文件</span><br><span class="line">find [path] -size -1G    #path下小于1G的文件</span><br></pre></td></tr></table></figure>
</li>
<li><p>时间戳</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">find [path] -mtime -n +m    #更改时间为m天前n天内的文件</span><br><span class="line">find [path] -atime -n +m    #访问时间</span><br><span class="line">find [path] -ctime -n +m    #创建时间</span><br><span class="line"><span class="meta">#</span> mtime参数后数字表示天， mmin参数后数字表示分钟</span><br></pre></td></tr></table></figure>
</li>
<li><p>文件属主</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">find [path] -user &lt;user_name&gt;    #path下属主为user_name的文件</span><br><span class="line">find [path] -group &lt;group_name&gt;    #path下组名为group_name的文件</span><br><span class="line">find [path] -nouser    #用户ID不存在的文件</span><br><span class="line">find [path] -nogroup    #组ID不存在的文件</span><br></pre></td></tr></table></figure>
</li>
<li><p>文件权限</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">find [path] -perm 755    #权限设置为755的文件</span><br><span class="line">find [path] -perm -u=r    #文件属主有读权限的目录或文件</span><br><span class="line">find [path] -perm -g=r    #用户组有读权限的目录或文件</span><br><span class="line">find [path] -perm -o=r    #其它用户有读权限的目录或文件</span><br></pre></td></tr></table></figure>
</li>
<li><p>其他搜索参数</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">find [path] &lt;search_regular&gt; [option]</span><br><span class="line"><span class="meta">#</span> -follow 追踪链接文件</span><br><span class="line"><span class="meta">#</span> -mount 不跨越文件系统mount点</span><br><span class="line"><span class="meta">#</span> -empty 空文件</span><br></pre></td></tr></table></figure>
</li>
</ol>
<h1 id="示例"><a href="#示例" class="headerlink" title="示例"></a>示例</h1><ol>
<li><p>多筛选条件组合</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">find [path] -name '*R' -a -mtime -1    #修改时间一天内的R脚本文件</span><br><span class="line"><span class="meta">#</span>-a | -and    # 同时满足</span><br><span class="line"><span class="meta">#</span>-o | -or    # 或</span><br><span class="line"><span class="meta">#</span>-not    # 条件取反</span><br></pre></td></tr></table></figure>
</li>
<li><p>几个例子，来自微博@linux命令行精选网</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">find . ! -name &lt;NAME&gt; -delete    #用find删除文件时候排除特定文件</span><br><span class="line">find . -type l -xtype l    #查找失效的符号链接</span><br><span class="line">find . -iname '*.jpg' | sed 's/.*/&lt;img src="&amp;"&gt;/' &gt; gallery.html    #生成html 相册</span><br><span class="line">find . -size 0 -exec rm '&#123;&#125;' \;    #清理空文件</span><br><span class="line">find . -type d -exec mkdir -p $DESTDIR/&#123;&#125; \;    #复制一个目录的结构，忽略文件</span><br><span class="line">find . -type f -size +500M -exec ls -ls &#123;&#125; \; | sort -n    #找出所有大于500M的文件</span><br></pre></td></tr></table></figure>
</li>
</ol>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/linux/">linux</a>
  </div>

			</span>
			<span class="tags">
				
			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-20170511_parse_fasta" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2017/20170511_parse_fasta/" class="article-date">
  <time datetime="2017-05-11T03:40:15.000Z" itemprop="datePublished">2017-05-11</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2017/20170511_parse_fasta/">fasta序列解析</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>fasta格式最常见，也最常用，也是分析中最简单的格式吧，以<code>&gt;</code>开始后续字符表示序列名称，<code>空格</code>后对序列进行描述（可能不存在），换行即为序列信息，可包括换行。下面是一个fasta格式序列实例：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&gt;gi|187608668|ref|NM_001043364.2| Bombyx mori moricin (Mor), mRNA</span><br><span class="line">AAACCGCGCAGTTATTTAAAATATGAATATTTTAAAACTTTTCTTTGTTTTTA</span><br><span class="line">CAACGACAACTGTGTACTATTTTTTATATTTGGTTCGAAAAGTTGCATTATTA</span><br><span class="line">ACGATTTTAGAAAATAAAACTACTTTACTTTTACACG</span><br></pre></td></tr></table></figure>
<p>这个实例中序列名字为<strong>gi|187608668|ref|NM_001043364.2|</strong>，下面就直接贴上常用的fasta解析脚本：</p>
<h2 id="python手动解析"><a href="#python手动解析" class="headerlink" title="python手动解析"></a>python手动解析</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">parse_fasta</span><span class="params">(fa)</span>:</span></span><br><span class="line">    seq = &#123;&#125;</span><br><span class="line">    <span class="keyword">with</span> oepn(fa) <span class="keyword">as</span> fa:</span><br><span class="line">        <span class="keyword">for</span> i <span class="keyword">in</span> fa:</span><br><span class="line">            <span class="keyword">if</span> i.startswith(<span class="string">"&gt;"</span>):</span><br><span class="line">                name = i[<span class="number">1</span>:].partition(<span class="string">' '</span>)[<span class="number">0</span>]</span><br><span class="line">            <span class="keyword">elif</span>:</span><br><span class="line">                seq[name] = seq.get(name, <span class="string">''</span>) + i.strip()</span><br><span class="line">    <span class="keyword">return</span> seq</span><br></pre></td></tr></table></figure>
<h2 id="使用HTSeq模块"><a href="#使用HTSeq模块" class="headerlink" title="使用HTSeq模块"></a>使用HTSeq模块</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">parse_fasta_by_HTSeq</span><span class="params">(fa)</span>:</span></span><br><span class="line">    <span class="keyword">import</span> HTSeq</span><br><span class="line">    <span class="keyword">return</span> &#123;i.name:i.seq <span class="keyword">for</span> i <span class="keyword">in</span> HTSeq.FastaReader(fa)&#125;</span><br></pre></td></tr></table></figure>
<p>当然还能使用biopython进行解析，我不太常用这一模块，就不写了</p>
<h2 id="fasta输出格式化"><a href="#fasta输出格式化" class="headerlink" title="fasta输出格式化"></a>fasta输出格式化</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">format_fasta</span><span class="params">(seq, len=<span class="number">60</span>)</span>:</span></span><br><span class="line">    tmp = <span class="string">''</span></span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> range((len(seq)//len)+<span class="number">1</span>):</span><br><span class="line">        tmp += <span class="string">'&#123;&#125;\n'</span>.format(seq[(i*<span class="number">60</span>):((i+<span class="number">1</span>)*<span class="number">60</span>)])</span><br><span class="line">    <span class="keyword">return</span> tmp</span><br></pre></td></tr></table></figure>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/python/">python</a>
  </div>

			</span>
			<span class="tags">
				
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/fasta/">fasta</a></li></ul>

			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-20170510_RefSeq_GFF2GTF" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2017/20170510_RefSeq_GFF2GTF/" class="article-date">
  <time datetime="2017-05-10T14:33:43.000Z" itemprop="datePublished">2017-05-10</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2017/20170510_RefSeq_GFF2GTF/">RefSeq数据库GFF格式转换</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>最近一项目被RefSeq数据库中gff注释格式折磨死，流程总是不能正确识别，各种方法改了格式，勉强运行，后续步骤继续报错，想着以后肯定还有更多这种情况，自己对文件格式的理解，写了格式转换脚本。</p>
<p>GFF和GTF格式非常类似，都被用来基因结构注释，由9列表示不同意义的特征组成，多行共同注释一个基因的详细结构。然而两者又有很大差异，脚本处理起来也是不同。相比而言GTF处理起来简单些，第三列<code>feature</code>可能是<strong>gene, transcript, exon, CDS</strong>这少数几种情况，最重要的描述基因类型名称的第九列<code>attr</code>很统一，都会包括一般常用的<strong>gene_id, gene_biotype, transcript_id, transcript_biotype</strong>等特征。而GFF文件包括一个明确的层级关系，<strong>gene</strong>是<strong>transcript</strong>（存在mRNA，lncRNA，ncRNA，rRNA，tRNA等多种不同组织方式）的父节点，而<strong>transcript</strong>又是<strong>exon</strong>和<strong>CDS</strong>的父节点，特征层级不同也决定了第九列<code>attr</code>的注释信息不同，gene id, name等信息也是层次关系关联，文件处理较麻烦。</p>
<p>NCBI RefSeq数据库中GFF注释格式感觉更是复杂怪异，废话不多说，直接上代码，愚蠢方法手撕GFF</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#!/usr/bin/env python</span></span><br><span class="line"><span class="comment"># -*- coding:utf-8 -*-</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">import</span> gzip</span><br><span class="line"><span class="keyword">from</span> collections <span class="keyword">import</span> defaultdict</span><br><span class="line"><span class="keyword">from</span> collections <span class="keyword">import</span> Counter</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">GFF_parser</span><span class="params">(file)</span>:</span></span><br><span class="line">    openfile = gzip.open <span class="keyword">if</span> file.endswith(<span class="string">'gz'</span>) <span class="keyword">else</span> open</span><br><span class="line">    <span class="keyword">with</span> openfile(file) <span class="keyword">as</span> anno_file:</span><br><span class="line">        <span class="keyword">for</span> line <span class="keyword">in</span> anno_file:</span><br><span class="line">            <span class="keyword">if</span> line.startswith(<span class="string">'#'</span>):</span><br><span class="line">                <span class="keyword">continue</span></span><br><span class="line">            seqname, source, feature, start, end, score, strand, \</span><br><span class="line">            frame, attributeStr = line.strip().split(<span class="string">'\t'</span>)</span><br><span class="line">            attribute = [i.strip() <span class="keyword">for</span> i <span class="keyword">in</span> attributeStr.split(<span class="string">';'</span>) <span class="keyword">if</span> i.strip() != <span class="string">''</span>]</span><br><span class="line">            attr = &#123;i.partition(<span class="string">'='</span>)[<span class="number">0</span>]:i.partition(<span class="string">'='</span>)[<span class="number">2</span>] <span class="keyword">for</span> i <span class="keyword">in</span> attribute&#125;</span><br><span class="line">            <span class="keyword">yield</span> seqname, source, feature, int(start), int(end), score, strand, frame, attr</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">format_gtf_line</span><span class="params">(chro, feature, start, end, strand, attr, \</span></span></span><br><span class="line"><span class="function"><span class="params">    source=None, score=None, frame=None)</span>:</span></span><br><span class="line">    source = source <span class="keyword">if</span> source <span class="keyword">else</span> <span class="string">'RefSeq'</span></span><br><span class="line">    score = score <span class="keyword">if</span> score <span class="keyword">else</span> <span class="string">'.'</span></span><br><span class="line">    frame = frame <span class="keyword">if</span> frame <span class="keyword">else</span> <span class="string">'.'</span></span><br><span class="line">    attrstr = <span class="string">''</span>.join([<span class="string">'&#123;&#125; "&#123;&#125;"; '</span>.format(i[<span class="number">0</span>], i[<span class="number">1</span>]) <span class="keyword">for</span> i <span class="keyword">in</span> sorted(attr.items())])</span><br><span class="line">    gtf_line = <span class="string">'&#123;chro&#125;\t&#123;source&#125;\t&#123;feature&#125;\t&#123;start&#125;\t&#123;end&#125;\t&#123;score&#125;\t\</span></span><br><span class="line"><span class="string">        &#123;strand&#125;\t&#123;frame&#125;\t&#123;attrstr&#125;'</span>.format(**locals())</span><br><span class="line">    <span class="keyword">return</span> gtf_line</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">main</span><span class="params">(gff, output)</span>:</span></span><br><span class="line">    gtf_out = open(output, <span class="string">'w'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># RefSeq数据库中GFF文件feature收集，或去除无意义项，或替换</span></span><br><span class="line">    skip = [<span class="string">'region'</span>, <span class="string">'miRNA'</span>, <span class="string">'cDNA_match'</span>, <span class="string">'repeat_region'</span>, <span class="string">'D_loop'</span>, \</span><br><span class="line">            <span class="string">'match'</span>, <span class="string">'origin_of_replication'</span>, <span class="string">'centromere'</span>, <span class="string">'sequence_feature'</span>]</span><br><span class="line">    transcript = [<span class="string">'mRNA'</span>, <span class="string">'lnc_RNA'</span>, <span class="string">'primary_transcript'</span>, <span class="string">'tRNA'</span>, <span class="string">'rRNA'</span>, \</span><br><span class="line">            <span class="string">'C_gene_segment'</span>, <span class="string">'V_gene_segment'</span>, <span class="string">'SRP_RNA'</span>, <span class="string">'ncRNA'</span>, <span class="string">'transcript'</span>]</span><br><span class="line">    biotype = [<span class="string">'lnc_RNA'</span>, <span class="string">'tRNA'</span>, <span class="string">'rRNA'</span>, <span class="string">'SRP_RNA'</span>]</span><br><span class="line"></span><br><span class="line">    order = []</span><br><span class="line">    gene_info = defaultdict(dict)</span><br><span class="line">    transcript_info = defaultdict(dict)</span><br><span class="line">    check_gene_name = &#123;&#125;    <span class="comment">#检查重复gene_name</span></span><br><span class="line">    check_transcript_name = &#123;&#125;</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> GFF_parser(gff):</span><br><span class="line">        chro, source, feature, start, end, score, strand, frame, attr = i</span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> feature <span class="keyword">in</span> skip:</span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line">        <span class="keyword">elif</span> feature == <span class="string">'gene'</span>:</span><br><span class="line">            gene_id = attr[<span class="string">'ID'</span>]</span><br><span class="line">            gene_info[gene_id][<span class="string">'name'</span>] = attr[<span class="string">'Name'</span>]</span><br><span class="line">            gene_info[gene_id][<span class="string">'biotype'</span>] = attr[<span class="string">'gene_biotype'</span>]</span><br><span class="line">            check_gene_name[gene_id] = attr[<span class="string">'Name'</span>]</span><br><span class="line">        <span class="comment">#使用字典transcript_info储存转录本信息</span></span><br><span class="line">        <span class="keyword">elif</span> feature <span class="keyword">in</span> transcript:</span><br><span class="line">            transcript_id = attr[<span class="string">'ID'</span>]</span><br><span class="line">            gene_id = attr[<span class="string">'Parent'</span>] <span class="keyword">if</span> <span class="string">'Parent'</span> <span class="keyword">in</span> attr <span class="keyword">else</span> transcript_id</span><br><span class="line">            transcript_name = attr[<span class="string">'transcript_id'</span>] <span class="keyword">if</span> <span class="string">'transcript_id'</span> \</span><br><span class="line">                <span class="keyword">in</span> attr <span class="keyword">else</span> transcript_id</span><br><span class="line">            order.append(transcript_id)</span><br><span class="line">            transcript_info[transcript_id][<span class="string">'gtf_info'</span>] = [chro, start, end, strand]</span><br><span class="line">            transcript_info[transcript_id][<span class="string">'name'</span>] = transcript_name</span><br><span class="line">            transcript_info[transcript_id][<span class="string">'gene'</span>] = gene_id</span><br><span class="line">            <span class="comment"># 部分转录本不存在对应gene信息，直接关联转录类型</span></span><br><span class="line">            <span class="keyword">if</span> feature <span class="keyword">in</span> biotype:</span><br><span class="line">                transcript_info[transcript_id][<span class="string">'biotype'</span>] = feature</span><br><span class="line">            check_transcript_name[transcript_id] = transcript_name</span><br><span class="line">        <span class="keyword">elif</span> feature == <span class="string">'exon'</span>:</span><br><span class="line">            transcript_id = attr[<span class="string">'Parent'</span>]</span><br><span class="line">            <span class="keyword">try</span>:</span><br><span class="line">                transcript_info[transcript_id][<span class="string">'exon'</span>][<span class="number">1</span>].append(start)</span><br><span class="line">                transcript_info[transcript_id][<span class="string">'exon'</span>][<span class="number">2</span>].append(end)</span><br><span class="line">            <span class="keyword">except</span> KeyError:</span><br><span class="line">                transcript_info[transcript_id][<span class="string">'exon'</span>] = [chro, [start,], \</span><br><span class="line">                    [end,], strand]</span><br><span class="line">        <span class="keyword">elif</span> feature == <span class="string">'CDS'</span>:</span><br><span class="line">            transcript_id = attr[<span class="string">'Parent'</span>]</span><br><span class="line">            <span class="keyword">try</span>:</span><br><span class="line">                transcript_info[transcript_id][<span class="string">'CDS'</span>][<span class="number">1</span>].append(start)</span><br><span class="line">                transcript_info[transcript_id][<span class="string">'CDS'</span>][<span class="number">2</span>].append(end)</span><br><span class="line">                transcript_info[transcript_id][<span class="string">'CDS'</span>][<span class="number">4</span>].append(frame)</span><br><span class="line">            <span class="keyword">except</span> KeyError:</span><br><span class="line">                transcript_info[transcript_id][<span class="string">'CDS'</span>] = [chro, [start,], \</span><br><span class="line">                    [end,], strand, [frame,]]</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            sys.stderr.write(<span class="string">'Warning: Annotation Feature "&#123;&#125;" is not collected, '</span>\</span><br><span class="line">                <span class="string">'Unknown error may occur.\n'</span>.format(feature))</span><br><span class="line"></span><br><span class="line"><span class="comment"># 检查重复gene_name， transcript_names，进行替换</span></span><br><span class="line">    dup_name = <span class="keyword">lambda</span> list_in: set([i[<span class="number">0</span>] <span class="keyword">for</span> i <span class="keyword">in</span> Counter(list_in).items() <span class="keyword">if</span> i[<span class="number">1</span>] &gt; <span class="number">1</span>])</span><br><span class="line">    gene_name = [i[<span class="number">1</span>] <span class="keyword">for</span> i <span class="keyword">in</span> check_gene_name.items()]</span><br><span class="line">    dup_gene_name = dup_name(gene_name)</span><br><span class="line"></span><br><span class="line">    transcript_name = [i[<span class="number">1</span>] <span class="keyword">for</span> i <span class="keyword">in</span> check_transcript_name.items()]</span><br><span class="line">    dup_transcript_name = dup_name(transcript_name)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 输出GTF文件，order控制输出顺序</span></span><br><span class="line">    <span class="keyword">for</span> transcript <span class="keyword">in</span> order:</span><br><span class="line">        attr = &#123;&#125;</span><br><span class="line">        gene_id = transcript_info[transcript][<span class="string">'gene'</span>]</span><br><span class="line">        <span class="keyword">try</span>:</span><br><span class="line">            gene_name = gene_info[gene_id][<span class="string">'name'</span>]</span><br><span class="line">            gene_name = gene_id <span class="keyword">if</span> gene_name <span class="keyword">in</span> dup_gene_name <span class="keyword">else</span> gene_name</span><br><span class="line">        <span class="keyword">except</span> KeyError:</span><br><span class="line">            gene_name = gene_id</span><br><span class="line">        attr[<span class="string">'gene_id'</span>] = gene_name</span><br><span class="line">        transcript_name = transcript_info[transcript][<span class="string">'name'</span>]</span><br><span class="line">        transcript_name = transcript <span class="keyword">if</span> transcript_name <span class="keyword">in</span> 、</span><br><span class="line">            dup_transcript_name <span class="keyword">else</span> transcript_name</span><br><span class="line">        attr[<span class="string">'transcript_id'</span>] = transcript_name</span><br><span class="line">        <span class="keyword">try</span>:</span><br><span class="line">            attr[<span class="string">'transcript_biotype'</span>] = transcript_info[transcript][<span class="string">'biotype'</span>]</span><br><span class="line">        <span class="keyword">except</span> KeyError:</span><br><span class="line">            attr[<span class="string">'transcript_biotype'</span>] = gene_info[gene_id][<span class="string">'biotype'</span>]</span><br><span class="line">        <span class="keyword">try</span>:</span><br><span class="line">            attr[<span class="string">'gene_biotype'</span>] =  gene_info[gene_id][<span class="string">'biotype'</span>]</span><br><span class="line">        <span class="keyword">except</span> KeyError:</span><br><span class="line">            attr[<span class="string">'gene_biotype'</span>] =  transcript_info[transcript][<span class="string">'biotype'</span>]</span><br><span class="line">        chro, start, end, strand = transcript_info[transcript][<span class="string">'gtf_info'</span>]</span><br><span class="line">        gtf_out.write(<span class="string">'&#123;&#125;\n'</span>.format(format_gtf_line(chro,<span class="string">'transcript'</span>,、</span><br><span class="line">            start,end,strand,attr)))</span><br><span class="line">        <span class="keyword">try</span>:</span><br><span class="line">            exon_feature = transcript_info[transcript][<span class="string">'exon'</span>]</span><br><span class="line">        <span class="keyword">except</span> KeyError:</span><br><span class="line">            exon_feature = transcript_info[transcript][<span class="string">'CDS'</span>][:<span class="number">-1</span>]</span><br><span class="line">        chro, start, end, strand = exon_feature</span><br><span class="line">        <span class="keyword">for</span> i,j <span class="keyword">in</span> enumerate(start):</span><br><span class="line">            gtf_out.write(<span class="string">'&#123;&#125;\n'</span>.format(format_gtf_line(chro,<span class="string">'exon'</span>,、</span><br><span class="line">                start[i],end[i],strand,attr)))</span><br><span class="line">        <span class="keyword">try</span>:</span><br><span class="line">            cds_feature = transcript_info[transcript][<span class="string">'CDS'</span>]</span><br><span class="line">            chro, start, end, strand, frame = cds_feature</span><br><span class="line">            <span class="keyword">for</span> i,j <span class="keyword">in</span> enumerate(start):</span><br><span class="line">                gtf_out.write(<span class="string">'&#123;&#125;\n'</span>.format(format_gtf_line(chro, <span class="string">'CDS'</span>, start[i], \</span><br><span class="line">                        end[i], strand, attr, frame=frame[i])))</span><br><span class="line">        <span class="keyword">except</span> KeyError:</span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line">        <span class="comment">#部分注释信息不完善，多次使用try... except...输出替代的对应注释信息</span></span><br><span class="line">    gtf_out.close()</span><br></pre></td></tr></table></figure>
<p>GFF格式包括<strong>gene, transcript, exon(CDS)</strong>三个层级，其中<strong>transcript</strong>进行上下关联最为重要，这里脚本以<strong>transcript</strong>作为键值进行数据处理，另外脚本还有一个假设就是<strong>gene, transcript</strong>对应ID在文件中唯一。</p>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/bioinfor/">bioinfor</a>
  </div>

			</span>
			<span class="tags">
				
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/GFF/">GFF</a></li><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/GTF/">GTF</a></li></ul>

			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
  
    <nav id="page-nav">
      <span class="page-number current">1</span><a class="page-number" href="/page/2/">2</a><a class="extend next" rel="next" href="/page/2/">Next &raquo;</a>
    </nav>
  

    </div>
    <div class="mb-search">
  <form action="//google.com/search" method="get" accept-charset="utf-8">
    <input type="search" name="q" results="0" placeholder="Search">
    <input type="hidden" name="q" value="site:chengcz.github.io">
  </form>
</div>
<footer id="footer">
	<!---<h1 class="footer-blog-title">
		<a href="/">ChengCZ&#39;s Notebook</a>
	</h1>--->
	<span class="copyright">
		&copy; 2018 ChengCZ<br>
		<!---Modify from <a href="http://sanographix.github.io/tumblr/apollo/" target="_blank">Apollo</a> theme, designed by <a href="http://www.sanographix.net/" target="_blank">SANOGRAPHIX.NET</a><br>
		Powered by <a href="http://hexo.io/" target="_blank">Hexo</a>--->
	</span>
</footer>

    

<script src="//ajax.googleapis.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>


<link rel="stylesheet" href="/fancybox/jquery.fancybox.css" media="screen" type="text/css">
<script src="/fancybox/jquery.fancybox.pack.js"></script>


<script src="/js/script.js"></script>
  </div>
</body>
</html>