<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  <title>ChengCZ&#39;s Notebook</title>
  <meta name="viewport" content="width=device-width">
  <meta property="og:type" content="website">
<meta property="og:title" content="ChengCZ&#39;s Notebook">
<meta property="og:url" content="http://chengcz.github.io/index.html">
<meta property="og:site_name" content="ChengCZ&#39;s Notebook">
<meta property="og:locale" content="zh-CN">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="ChengCZ&#39;s Notebook">
  
    <link rel="alternative" href="/atom.xml" title="ChengCZ&#39;s Notebook" type="application/atom+xml">
  
  
    <link rel="icon" href="/favicon.png">
  
  <link rel="stylesheet" href="/css/style.css">
  <!--[if lt IE 9]><script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script><![endif]-->
  
</head>
<body>
  <div id="container">
    <div class="mobile-nav-panel">
	<i class="icon-reorder icon-large"></i>
</div>
<header id="header">
	<h1 class="blog-title">
		<a href="/">ChengCZ&#39;s Notebook</a>
	</h1>
	<nav class="nav">
		<ul>
			<li><a href="/">Home</a></li><li><a href="/archives">Archives</a></li><li><a href="/about">About</a></li>
			<li><a id="nav-search-btn" class="nav-icon" title="Search"></a></li>
			<li><a href="/atom.xml" id="nav-rss-link" class="nav-icon" title="RSS Feed"></a></li>
		</ul>
	</nav>
	<div id="search-form-wrap">
		<form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="Search"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://chengcz.github.io"></form>
	</div>
</header>
    <div id="main">
      
  
    <article id="post-201805_subprocess" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2018/201805_subprocess/" class="article-date">
  <time datetime="2018-05-17T11:15:52.000Z" itemprop="datePublished">2018-05-17</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2018/201805_subprocess/">python 执行 linux 命令 —— subprocess</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>python 脚本内执行 linux 命令有多种方式，os.system(), os.popen()也可以，官方最新推荐方式为subprocess</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> subprocess</span><br><span class="line"></span><br><span class="line">job = subprocess.Popen(<span class="string">'ls -l'</span>,    <span class="comment"># string of cmd</span></span><br><span class="line">					   shell=<span class="keyword">True</span>,    <span class="comment"># 传入参数作为 shell 进行执行</span></span><br><span class="line">					   stdout=subprocess.PIPE,    <span class="comment"># 输出标准输出到subprocess流程</span></span><br><span class="line">  					   stderr=subprocess.PIPE,    <span class="comment"># 输出标准错误到subprocess流程</span></span><br><span class="line">  					   universal_newlines=<span class="keyword">True</span>)    <span class="comment"># stdout 返回时，使用 string 代替 bytes 类型</span></span><br><span class="line">pid = job.pid    <span class="comment"># 获得 cmd 本地执行的 pid</span></span><br><span class="line">job.wait()    <span class="comment"># 等待工作结束，默认Popen不等待 cmd 执行完毕即退出</span></span><br><span class="line">stdout, stderr = job.communicate()    <span class="comment"># 获得 cmd 执行返回标准输出、标准错误</span></span><br><span class="line">job.returncode    <span class="comment"># 获得 cmd 执行状态码</span></span><br></pre></td></tr></table></figure>
<p>os 模块实现 linux cmd 执行</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> os</span><br><span class="line"></span><br><span class="line">os.system(<span class="string">'ls -l'</span>)    <span class="comment"># 直接返回 cmd 执行状态码</span></span><br><span class="line"><span class="comment"># assert not os.system('ls')</span></span><br><span class="line">os.popen(<span class="string">'ls -l'</span>)    <span class="comment"># 以文件句柄的方式返回 cmd 执行的stdout信息</span></span><br></pre></td></tr></table></figure>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/python/">python</a>
  </div>

			</span>
			<span class="tags">
				
			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-201805_argparse" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2018/201805_argparse/" class="article-date">
  <time datetime="2018-05-17T11:15:52.000Z" itemprop="datePublished">2018-05-17</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2018/201805_argparse/">python 参数解析 —— argsparse</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>argsparse 解析命令行参数，一些常用参数整理。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> argparse</span><br><span class="line"></span><br><span class="line">parser = argparse.ArgumentParser(<span class="comment"># prog=__file__,    # 设置项目名称，__file__ 与 sys.argv[0] 一致，默认显示为 basename(__file__)</span></span><br><span class="line">      usage=<span class="string">'''</span></span><br><span class="line"><span class="string">      python %(prog)s [option]  '''</span>,    <span class="comment"># 设置脚本输出默认 USAGE 信息</span></span><br><span class="line">      description=<span class="string">''' some script description '''</span>,   <span class="comment"># 脚本功能描述等等</span></span><br><span class="line">      formatter_class=argparse.RawTextHelpFormatter,    <span class="comment"># 以脚本内原始文档格式输出描述等信息</span></span><br><span class="line">      epilog=<span class="string">'''</span></span><br><span class="line"><span class="string">      Note and input example, and so on.  '''</span>)    <span class="comment"># 脚本描述以外的其他信息</span></span><br><span class="line"></span><br><span class="line">parser.add_argument(<span class="string">'-v'</span>, <span class="string">'--version'</span>,    <span class="comment"># 添加脚本版本信息参数</span></span><br><span class="line">                    action=<span class="string">'version'</span>,</span><br><span class="line">                    version=<span class="string">'%(prog)s 2.0'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 添加一个普通的可选参数</span></span><br><span class="line">parser.add_argument(<span class="string">'-e'</span>,    <span class="comment"># 短参数</span></span><br><span class="line">                    <span class="string">'--example'</span>,    <span class="comment"># 长参数</span></span><br><span class="line">                    dest=<span class="string">'example'</span>,     <span class="comment"># 参数对应模块内部，唯一标识符</span></span><br><span class="line">                    metavar=<span class="string">''</span>,    <span class="comment"># 设置help信息中参数后字符</span></span><br><span class="line">                    action=<span class="string">'store'</span>,    <span class="comment"># 参数处理方式</span></span><br><span class="line">                    type=str,    <span class="comment"># 传入参数类型</span></span><br><span class="line">                    choices=[<span class="string">'a'</span>, <span class="string">'b'</span>, <span class="string">'c'</span>],    <span class="comment"># 参数选择范围</span></span><br><span class="line">                    required=<span class="keyword">True</span>,    <span class="comment"># 表示该参数必须被设置</span></span><br><span class="line">                    default=<span class="string">'a'</span>,    <span class="comment"># 参数缺省时，该参数默认值</span></span><br><span class="line">                    nargs=<span class="string">'+'</span>,    <span class="comment"># 参数个数，*+？，表示至少一个输入</span></span><br><span class="line">                    help=<span class="string">'some help infor'</span>)    <span class="comment"># help信息</span></span><br><span class="line"><span class="comment"># action属性为store_true, store_false, store_const时，参数不接任何输入信息</span></span><br><span class="line">parser.add_argument(<span class="string">'--sum'</span>,</span><br><span class="line">                    dest=<span class="string">'accumulate'</span>,</span><br><span class="line">                    action=<span class="string">'store_const'</span>,    <span class="comment"># 该参数被设置时，参数取值成 const 对应值</span></span><br><span class="line">                    const=sum,</span><br><span class="line">                    default=max,</span><br><span class="line">                    help=<span class="string">'sum the integers (default: find the max)'</span>)</span><br><span class="line"><span class="comment"># 位置参数</span></span><br><span class="line">parser.add_argument(<span class="string">'pos'</span>,</span><br><span class="line">                    dest=<span class="string">'pos_argument'</span>,</span><br><span class="line">                    metavar=<span class="string">''</span>,</span><br><span class="line">                    nargs=<span class="string">'+'</span>,    <span class="comment"># 参数输入文件个数</span></span><br><span class="line">                    help=<span class="string">'some help infor'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 互斥参数组，组内参数不可同时出现，required=True表示组内必须有参数被设置</span></span><br><span class="line">exclusive = parser.add_mutually_exclusive_group(required=<span class="keyword">True</span>)    </span><br><span class="line">exclusive.add_argument(<span class="string">'-a'</span>, )</span><br><span class="line">exclusive.add_argument(<span class="string">'-b'</span>, )</span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置多个参数为一组，无实际意义，仅在打印help信息时，分组显示</span></span><br><span class="line">group = parser.add_argument_group(<span class="string">'argument group info'</span>)    </span><br><span class="line">group.add_argument(<span class="string">'-c'</span>, )</span><br><span class="line">group.add_argument(<span class="string">'-d'</span>, )</span><br><span class="line"></span><br><span class="line"><span class="comment"># parse.print_help()</span></span><br><span class="line">args = parser.parse_args()    <span class="comment"># 参数解析，生成可调用对象</span></span><br><span class="line"></span><br><span class="line">example = args.example</span><br><span class="line"><span class="comment"># ...</span></span><br></pre></td></tr></table></figure>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/python/">python</a>
  </div>

			</span>
			<span class="tags">
				
			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-201805_TophatFusion" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2018/201805_TophatFusion/" class="article-date">
  <time datetime="2018-05-17T10:45:52.000Z" itemprop="datePublished">2018-05-17</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2018/201805_TophatFusion/">使用 topaht2-fusion 分析 Fusion-Gene</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>tophat2-fusion 文档搬运工：<a href="https://ccb.jhu.edu/software/tophat/fusion_tutorial.shtml" target="_blank" rel="noopener">https://ccb.jhu.edu/software/tophat/fusion_tutorial.shtml</a></p>
<p>###step1.tophat</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">tophat2 -p 8 -o $OUTPUT --fusion-search --keep-fasta-order --bowtie1 --no-coverage-search --fusion-anchor-length 10 --fusion-min-dist 100000 --mate-inner-dist 40 --mate-std-dev 100 $hg_bowtie1_index $read1.fastq.gz $read2.fastq.gz</span><br></pre></td></tr></table></figure>
<p><strong>Note:</strong></p>
<p><em>–fusion-search</em>，融合 reads 搜索</p>
<p><em>–keep-fasta-order</em>,</p>
<p><em>–bowtie1</em>，使用 bowtie1 代替 bowtie2 进行 reads 比对。需要使用 bowtie-1.1.2，最新版本 1.2.2 会报错</p>
<p><img src="./images/2018/bowtie1.2_TophatFusion.png" alt=""></p>
<p><em>–no-coverage-search</em>,</p>
<p><em>–fusion-anchor-length 10</em> ,</p>
<p><em>–fusion-min-dist 100000</em> ,</p>
<p><em>–mate-inner-dist 40</em> ,</p>
<p><em>–mate-std-dev 100</em> ,</p>
<p>###step2.tophat-fusion-post</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">tophat-fusion-post -p 8 --num-fusion-reads 1 --num-fusion-pairs 2 --num-fusion-both 5 $hg_bowtie1_index</span><br></pre></td></tr></table></figure>
<p><strong>Note:</strong></p>
<p>–num-fusion-reads* , 支持融合的最少 reads 数目</p>
<p><em>–num-fusion-pairs</em>, 支持融合事件的最少配对 reads 数，表示为跨越融合点的测序片段</p>
<p><em>–num-fusion-both</em>,  支持融合的最少 reads，包括跨越的 reads 对和 split reads</p>
<p><em>–fusion-read-mismatches</em>, Reads support fusions if they map across fusion with at most this many mismatches. The default is 2.</p>
<p><em>–fusion-multireads</em>, Reads that map to more than this many places will be ignored. The default is 2.</p>
<p><em>–non-human</em>, If your annotation is different from that of human, use the option.</p>
<p><em>–fusion-pair-dist</em>, Pairs with this specified distance are counted as supporting evidence for the fusion. The default threshold for the inner distance is 250.</p>
<h3 id="results"><a href="#results" class="headerlink" title="results"></a>results</h3><p>测试了多个样品，多个参数，也根据官网数据、命令运行项目，都没有发现任何融合结果，bug还是打开方式不对</p>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/bioinfor/">bioinfor</a>
  </div>

			</span>
			<span class="tags">
				
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/FusionGene/">FusionGene</a></li></ul>

			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-201805_STARFusion" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2018/201805_STARFusion/" class="article-date">
  <time datetime="2018-05-15T14:14:52.000Z" itemprop="datePublished">2018-05-15</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2018/201805_STARFusion/">使用 STAR-fusion 分析 Fusion-Gene</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>软件文档：<a href="https://github.com/STAR-Fusion/STAR-Fusion/wiki" target="_blank" rel="noopener">https://github.com/STAR-Fusion/STAR-Fusion/wiki</a></p>
<p>##软件安装</p>
<ol>
<li>STAR，<a href="https://github.com/alexdobin/STAR" target="_blank" rel="noopener">https://github.com/alexdobin/STAR</a></li>
<li>运行依赖 perl 模块<ol>
<li>DB_File</li>
<li>URI::Escape</li>
<li>Set::IntervalTree</li>
<li>Carp::Assert</li>
<li>JSON::XS</li>
<li>PerlIO::gzip</li>
<li>common::sense</li>
<li>Types::Serialiser</li>
<li>Canary::Stability</li>
</ol>
</li>
</ol>
<p>##数据库准备</p>
<p>下载一个较小的未处理的参考文件，自己运行 index 命令。要是网速够快也可以直接在 index 好的数据库文件，~27G</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">wget -c https://data.broadinstitute.org/Trinity/CTAT_RESOURCE_LIB/GRCh37_v19_CTAT_lib_Feb092018.source_data.tar.gz</span><br><span class="line"><span class="meta">$</span>STAR_FUSION_HOME/FusionFilter/prep_genome_lib.pl \</span><br><span class="line">    --genome_fa ref_genome.fa \</span><br><span class="line">    --gtf ref_annot.gtf \</span><br><span class="line">    --fusion_annot_lib CTAT_HumanFusionLib.v0.1.0.dat.gz \</span><br><span class="line">    --annot_filter_rule AnnotFilterRule.pm \</span><br><span class="line">    --pfam_db PFAM.domtblout.dat.gz</span><br></pre></td></tr></table></figure>
<p>##运行STAR-Fusion</p>
<p><strong>STAR-Fusion</strong>  对 STAR 输出的嵌合比对分析发现可能存在的基因融合事件</p>
<h3 id="从-fastq-文件开始"><a href="#从-fastq-文件开始" class="headerlink" title="从 fastq 文件开始"></a>从 fastq 文件开始</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">$</span>STAR_FUSION_HOME/STAR-Fusion \</span><br><span class="line">    --genome_lib_dir /path/to/your/CTAT_resource_lib \</span><br><span class="line">    --left_fq reads_1.fq \</span><br><span class="line">    --right_fq reads_2.fq \</span><br><span class="line">    --output_dir star_fusion_outdir \</span><br><span class="line">    --no_remove_dups</span><br></pre></td></tr></table></figure>
<h3 id="从-STAR-产生-Bam-文件开始"><a href="#从-STAR-产生-Bam-文件开始" class="headerlink" title="从 STAR 产生 Bam 文件开始"></a>从 STAR 产生 Bam 文件开始</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">STAR --genomeDir $&#123;star_index_dir&#125; \</span><br><span class="line">    --readFilesIn $&#123;left_fq_filename&#125; $&#123;right_fq_filename&#125; \</span><br><span class="line">    --twopassMode Basic \</span><br><span class="line">    --outReadsUnmapped None \</span><br><span class="line">    --chimSegmentMin 12 \</span><br><span class="line">    --chimJunctionOverhangMin 12 \</span><br><span class="line">    --alignSJDBoverhangMin 10 \</span><br><span class="line">    --alignMatesGapMax 100000 \</span><br><span class="line">    --alignIntronMax 100000 \</span><br><span class="line">    --chimSegmentReadGapMax 3 \</span><br><span class="line">    --alignSJstitchMismatchNmax 5 -1 5 5 \</span><br><span class="line">    --runThreadN $&#123;THREAD_COUNT&#125; \</span><br><span class="line">    --outSAMstrandField intronMotif</span><br><span class="line"></span><br><span class="line">STAR-Fusion --genome_lib_dir /path/to/your/CTAT_resource_lib \</span><br><span class="line">             -J Chimeric.out.junction \</span><br><span class="line">             --output_dir star_fusion_outdir</span><br></pre></td></tr></table></figure>
<h2 id="输出文件"><a href="#输出文件" class="headerlink" title="输出文件"></a>输出文件</h2><p>STAR 速度还是那么让人惊喜，6m reads不到半小时。 融合结果<code>star-fusion.fusion_predictions.abridged.tsv</code></p>
<blockquote>
<p><strong>FusionName</strong>,<br><strong>JunctionReadCount</strong>, split align到融合点的序列片段数<br><strong>SpanningFragCount</strong>, 双端reads跨越融合点的序列片段数<br><strong>SpliceType</strong>, 断点是否在注释文件存在<br><strong>LeftGene</strong>,<br><strong>LeftBreakpoint</strong>,<br><strong>RightGene</strong>,<br><strong>RightBreakpoint</strong>,<br><strong>LargeAnchorSupport</strong>,<br><strong>FFPM</strong>, fusion fragments per million total reads<br><strong>LeftBreakDinuc</strong>,<br><strong>LeftBreakEntropy</strong>,<br><strong>RightBreakDinuc</strong>,<br><strong>RightBreakEntropy</strong>,<br><strong>annots</strong>,</p>
</blockquote>
<h2 id="结果比较"><a href="#结果比较" class="headerlink" title="结果比较"></a>结果比较</h2><p>真是一个悲伤的故事，根据官网给的两种运行方式，结果差别这么大。查看 STAR-Fusion 脚本，使用的 mapping 参数差异有点大啊，哪一个比较合理呢（一个新坑）？？？</p>
<p><img src="./images/2018/star_fusion.results.png" alt=""></p>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/bioinfor/">bioinfor</a>
  </div>

			</span>
			<span class="tags">
				
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/FusionGene/">FusionGene</a></li></ul>

			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-201805_trimmomatic" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2018/201805_trimmomatic/" class="article-date">
  <time datetime="2018-05-13T07:47:30.000Z" itemprop="datePublished">2018-05-13</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2018/201805_trimmomatic/">使用 trimmomatic 进行数据质控</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>trimmomatic 使用 java 编写，免安装多平台运行，同时运行速度非常快。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> paired end </span><br><span class="line">java -jar trimmomatic-0.32.jar PE -threads 8 -phred33 \</span><br><span class="line">    sample_R1.fastq.gz sample_R2.fastq.gz \</span><br><span class="line">    sample_R1_paired.fastq.gz sample_R1_unpaired.fastq.gz \</span><br><span class="line">    sample_R2_paired.fastq.gz sample_R2_unpaired.fastq.gz \</span><br><span class="line">    ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> single end</span><br><span class="line">java -jar trimmomatic-0.32.jar PE -threads 8 -phred33 \</span><br><span class="line">    sample.fastq.gz sample_clean.fastq.gz \</span><br><span class="line">    ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36</span><br></pre></td></tr></table></figure>
<p>参数选择<br><strong>PE/SE</strong></p>
<blockquote>
<p>设定对 Paired-End 或 Single-End 的 reads 进行处理，其输入和输出参数稍有不一样</p>
</blockquote>
<p><strong>threads</strong></p>
<blockquote>
<p>设置多线程运行数</p>
</blockquote>
<p><strong>phred33/phred64</strong></p>
<blockquote>
<p>设置碱基的质量格式，可选 phred64</p>
</blockquote>
<p><strong>ILLUMINACLIP</strong>:<fastawithadaptersetc>:<seed mismatches="">:<palindrome clip="" threshold="">:<simple clip="" threshold="">:<minadapterlength>:<keepbothreads> </keepbothreads></minadapterlength></simple></palindrome></seed></fastawithadaptersetc></p>
<blockquote>
<p>切除 adapter 序列。参数后分别接 adapter 序列的 fasta 文件，允许的最大 mismatch 数， palindrome 模式下匹配碱基数阈值： simple 模式下的匹配碱基数阈值</p>
<p><em>minAdapterLength</em>：只对 PE 测序的 palindrome clip 模式有效，指定 palindrome 模式下可以切除的接头序列最短长度，由于历史的原因，默认值是 8，但实际上 palindrome 模式可以切除短至 1bp 的接头污染，所以可以设置为 1 。<br><em>keepBothReads</em>：只对 PE 测序的 palindrome clip 模式有效，这个参数很重要，在上图中 D 模式下， R1 和 R2 在去除了接头序列之后剩余的部分是完全反向互补的，默认参数 false，意味着整条去除与 R1 完全反向互补的 R2，当做重复去除掉，但在有些情况下，例如需要用到 paired reads 的 bowtie2 流程，就要将这个参数改为 true，否则会损失一部分 paired reads。</p>
</blockquote>
<p><strong>SLIDINGWINDOW</strong></p>
<blockquote>
<p>从 reads 首端（ 5’端）开始进行滑动，当滑动位点周围一段序列(window)的平均碱基低于阈值，则从该处进行切除。 Windows 的 size 是 4 bp， 若其平均碱基质量小于15，则切除</p>
</blockquote>
<p>MAXINFO：<br><strong>LEADING/TRAILING&gt;</strong></p>
<blockquote>
<p>切除 reads 首端（ 5’端） / reads 末端（ 3’端）碱基质量小于指定值的碱基</p>
</blockquote>
<p><strong>CROP/HEADCROP</strong></p>
<blockquote>
<p>从 reads 末端（ 3’端）/reads 首端（ 5’端）切除碱基到指定长度</p>
</blockquote>
<p><strong>MINLEN</strong></p>
<blockquote>
<p>抛弃低于指定长度的 reads</p>
</blockquote>
<p><strong>TOPHRED33/TOPHRED64</strong></p>
<blockquote>
<p>转换碱基质量格式，Illumina HiSeq 2000质量系统为phred-64，可用该参数转换到phred-33</p>
</blockquote>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/bioinfor/">bioinfor</a>
  </div>

			</span>
			<span class="tags">
				
			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-201805_snpEff" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2018/201805_snpEff/" class="article-date">
  <time datetime="2018-05-09T13:45:52.000Z" itemprop="datePublished">2018-05-09</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2018/201805_snpEff/">VCF 注释 —— SnpEff</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>日常工作之读文档，了解参数改变、新方法功能，对已有流程进行修改。</p>
<p>各种文档看一边遍一遍，使用需要又一次看 SnpEff （<a href="http://snpeff.sourceforge.net/index.html" target="_blank" rel="noopener">http://snpeff.sourceforge.net/index.html</a> ），进行一些整理如下，不过还是建议<a href="http://snpeff.sourceforge.net/SnpEff_manual.html" target="_blank" rel="noopener">原文</a>，鬼知道我会在哪里写错理解错。</p>
<h3 id="运行"><a href="#运行" class="headerlink" title="运行"></a>运行</h3><ol>
<li><p>annotation</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">java -jar snpEff.jar eff genome	input.vcf &gt;output.vcf</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> -hgvs, 默认开启，使用 HGVS 变异描述语法对蛋白替换进行描述</span><br><span class="line"><span class="meta">#</span> -fi, 提供bed文件，只对bed区域内位点进行注释</span><br><span class="line"><span class="meta">#</span> -no-downstream, -no-intergenic, -no-intron, -no-upstream, -no-utr, -no [effectType], 参数还可对根据注释类型对注释进行初过滤</span><br><span class="line"><span class="meta">#</span> -canon, 只使用基因对应最长转录本（作为权威转录本）进行注释；参数-canonList可以自行提供基因权威转录本，格式为“GeneID    transcript_ID”</span><br><span class="line"><span class="meta">#</span> -noStats, 不输出注释统计信息</span><br><span class="line"><span class="meta">#</span> -v, 话痨模式，输出各种运行信息</span><br></pre></td></tr></table></figure>
<p>​</p>
</li>
<li><p>database</p>
<p>人、小鼠等物种都有已经 build 的数据库，可以直接下载，链接：<a href="https://sourceforge.net/projects/snpeff/files/databases/v4_3/" target="_blank" rel="noopener">https://sourceforge.net/projects/snpeff/files/databases/v4_3/</a>，注释选择基因组版本。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">java -jar snpEff.jar databases    # 查看可用数据库</span><br><span class="line">java -jar snpEff.jar download -v GRCh37.75    # 下载数据库 GRCh37.75</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> Q, 使用执行程序下载数据库时容易断，一般建议手动下载，或者wget等其他方式</span><br></pre></td></tr></table></figure>
<p>软件还提供了自行构建数据库，详细用法如下：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span> 添加配置信息</span><br><span class="line">echo 'GRCh37.70.genome : Human ' &gt;&gt;~/snpEff/snpEff.config</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> Create directoy for this new genome</span><br><span class="line">mkdir -p ~/snpEff/data/GRCh37.70 &amp;&amp; cd ~/snpEff/data/GRCh37.70</span><br><span class="line"><span class="meta">#</span> Get annotation files</span><br><span class="line">wget -c -O genes.gtf.gz ftp://ftp.ensembl.org/pub/release-70/gtf/homo_sapiens/Homo_sapiens.GRCh37.70.gtf.gz</span><br><span class="line"><span class="meta">#</span> Get the genome</span><br><span class="line">wget -c -O sequences.fa.gz ftp://ftp.ensembl.org/pub/release-70/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.70.dna.toplevel.fa.gz</span><br><span class="line"><span class="meta">#</span> Optional, Download CDSs</span><br><span class="line">wget -c -O cds.fa.gz ftp://ftp.ensembl.org/pub/release-70/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh37.70.cdna.all.fa.gz</span><br><span class="line"><span class="meta">#</span> Optional, Download proteins</span><br><span class="line">wget -c -O protein.fa.gz ftp://ftp.ensembl.org/pub/release-70/fasta/homo_sapiens/pep/Homo_sapiens.GRCh37.70.pep.all.fa.gz</span><br><span class="line"><span class="meta">#</span> Optional, Download regulatory annotations</span><br><span class="line">wget -c -O regulation.gff.gz ftp://ftp.ensembl.org/pub/release-70/regulation/homo_sapiens/AnnotatedFeatures.gff.gz</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> Building a database from GFF files</span><br><span class="line">cd ~/snpEff &amp;&amp; java -jar snpEff.jar build -gff3 -v GRCh37.70</span><br><span class="line"><span class="meta">#</span> rm -r ~/snpEff/data/GRCh37.70</span><br></pre></td></tr></table></figure>
<p>此外，snpEff 构建数据库还支持 RefSeq, GenBank 等多种数据输入格式，具体用法查看原文档。</p>
<p><strong>注意物种 codon 选择</strong>，codon 可以在 <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi" target="_blank" rel="noopener">https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi</a> 中查看，codon可以在<code>~/snpEff/snpEff.config</code>文件中设置：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># 设置特异 codon</span><br><span class="line">codon.Invertebrate_Mitochondrial:  TTT/F, TTC/F, TAC/Y, TAA/*, ATG/M+, ATG/M+, ACT/T, ...</span><br><span class="line"># 物种选择恰当的 codontable</span><br><span class="line">dm3.M.codonTable : Invertebrate_Mitochondrial    # the chromosome &apos;M&apos; from fly genome (dm3) uses Invertebrate_Mitochondrial codon table</span><br></pre></td></tr></table></figure>
<p>​</p>
</li>
<li><p>dump</p>
<p>从数据库中提取出注释信息</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">java -jar snpEff.jar dump -v -bed GRCh37.70 &gt; GRCh37.70.bed</span><br><span class="line"><span class="meta">#</span> [-bed |-txt ] 设置数据格式</span><br><span class="line"><span class="meta">#</span> [-0 |-1 ] 设置坐标系统，0-based 或者 1-based</span><br></pre></td></tr></table></figure>
<p>​</p>
</li>
</ol>
<h3 id="注释结果"><a href="#注释结果" class="headerlink" title="注释结果"></a>注释结果</h3><p><img src="./images/2018/snpEff.png" alt=""></p>
<p>可以看到注释信息被添加到了 VCF 中每个变异 INFO 信息中，以 ANN= 特征开始，详细的注释说明可查看<a href="./appendix/snpEff.pdf">官方文档</a> <a href="http://snpeff.sourceforge.net/VCFannotationformat_v1.0.pdf" target="_blank" rel="noopener">http://snpeff.sourceforge.net/VCFannotationformat_v1.0.pdf</a>。因为基因多个转录本、相互重叠基因等原因，可以看到变异位点被多次注释，需根据实际需求进行筛选。</p>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/bioinfor/">bioinfor</a>
  </div>

			</span>
			<span class="tags">
				
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/vcf/">vcf</a></li></ul>

			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-201709_umount" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2017/201709_umount/" class="article-date">
  <time datetime="2017-09-27T03:19:21.000Z" itemprop="datePublished">2017-09-27</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2017/201709_umount/">卸载一个假忙的硬盘</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<ol>
<li><p><code>df -h</code>查看卸载硬盘信息</p>
</li>
<li><p><code>umount /dev/sdf</code>，停止使用的硬盘假装忙碌状态</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">umount: /mnt/usbhd2: device is busy.</span><br><span class="line">        (In some cases useful info about processes that use</span><br><span class="line">         the device is found by lsof(8) or fuser(1))</span><br></pre></td></tr></table></figure>
</li>
<li><p><code>fuser /dev/sdf</code>，查看占用硬盘的进程</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">/mnt/sdf:         106657c</span><br></pre></td></tr></table></figure>
</li>
<li><p><code>kill -9 106657</code></p>
</li>
<li><p>重新执行<code>umount /dev/sdf</code></p>
</li>
</ol>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/linux/">linux</a>
  </div>

			</span>
			<span class="tags">
				
			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-201709_ShellNote" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2017/201709_ShellNote/" class="article-date">
  <time datetime="2017-09-15T08:38:30.000Z" itemprop="datePublished">2017-09-15</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2017/201709_ShellNote/">shell 命令使用笔记</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<ol>
<li><p><code>&amp;&amp;</code>，cmd1 &amp;&amp; cmd2</p>
<p>如果cmd1成功执行（返回0，注意设置中脚本执行返回值），那么执行cmd2</p>
<p>​</p>
</li>
<li><p><code>||</code>，cmd1 || cmd2</p>
<p>如果cmd1执行失败，那么执行cmd2</p>
<p>​</p>
</li>
<li><p><code>()</code>, <code>{}</code>组合逻辑控制</p>
<p>()组合来控制命令； {}组合控制子shell</p>
<p>​</p>
</li>
<li><p>条件判断</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">if [ condition ]; then</span><br><span class="line"><span class="meta">#</span> if test condition; then</span><br><span class="line">	echo 'yes'</span><br><span class="line">fi</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span>使用[]判断条件时，[]和condition之间必须用空格分隔开</span><br><span class="line"><span class="meta">#</span> -d		判断是不是文件夹</span><br><span class="line"><span class="meta">#</span> -f		判断是不是文件</span><br><span class="line"><span class="meta">#</span> -L		判断是不是链接</span><br><span class="line"><span class="meta">#</span> -s		判断文件是否存在且非空</span><br><span class="line"><span class="meta">#</span> -r		判断文件是否可读</span><br><span class="line"><span class="meta">#</span> -w 		判断文件是否可写</span><br><span class="line"><span class="meta">#</span> -x		判断文件是否可执行</span><br><span class="line"><span class="meta">#</span> -z		判断是否空字符串</span><br><span class="line"><span class="meta">#</span> -n		判断是否非空字符串</span><br><span class="line"><span class="meta">#</span> =		判断字符串是否相等</span><br><span class="line"><span class="meta">#</span> !=		判断字符串是否不等</span><br><span class="line"><span class="meta">#</span> -eq		判断数值是否相等</span><br><span class="line"><span class="meta">#</span> -ne		判断数值是否不等</span><br><span class="line"><span class="meta">#</span> -gt		第一个数大于第二个数</span><br><span class="line"><span class="meta">#</span> -ge		第一个数大于等于第二个数</span><br><span class="line"><span class="meta">#</span> -lt		第一个数小于第二个数</span><br><span class="line"><span class="meta">#</span> -le		第一个数小于等于第二个数</span><br><span class="line"><span class="meta">#</span> -a		逻辑与</span><br><span class="line"><span class="meta">#</span> -o		逻辑或</span><br><span class="line"><span class="meta">#</span> !		逻辑否</span><br></pre></td></tr></table></figure>
<p>​</p>
</li>
</ol>
<ol start="5">
<li><p>推出状态</p>
<p><code>exit n</code>，其中n表示数字，0表示脚本执行成功无错误，1表示执行失败某处有错误</p>
<blockquote>
<p>可以使用 <strong>$?</strong> 获得最后执行命令推出状态</p>
</blockquote>
<p>​</p>
</li>
<li><p>shell脚本</p>
<ul>
<li style="list-style: none"><input type="checkbox"> <strong>$#</strong> 获取脚本输出参数长度</li>
<li style="list-style: none"><input type="checkbox"> <strong>$@</strong> 获取脚本所有参数</li>
</ul>
</li>
<li><p>循环</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">for i in `ls `</span><br><span class="line">	do</span><br><span class="line">		echo 'yes'</span><br><span class="line">	done</span><br></pre></td></tr></table></figure>
<p>​</p>
</li>
<li><p>序列生成</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">seq 3    # 1 2 3</span><br><span class="line">seq 3 5    # 3 4 5</span><br><span class="line">seq 5 2 10    # 5 7 9</span><br></pre></td></tr></table></figure>
<p>​</p>
</li>
<li><p>随机数生成</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">echo $RANDOM</span><br><span class="line"><span class="meta">#</span> RANDOM是bash的一个内建函数，会返回一个[0, 32676]内的整数</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span> 生成一定范围内的随机数</span><br><span class="line">beg=5</span><br><span class="line">end=15</span><br><span class="line">echo $((RANDOM % ($end - $beg) + $beg))</span><br></pre></td></tr></table></figure>
<p>​</p>
</li>
<li><p>shell 数学计算</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">echo $((9-4))    # 5</span><br><span class="line">echo $((9/4))    # 2</span><br><span class="line">echo $((9/-4))    # -2</span><br></pre></td></tr></table></figure>
<p>​</p>
</li>
</ol>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/linux/">linux</a>
  </div>

			</span>
			<span class="tags">
				
			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-201709_submitDataToSRA" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2017/201709_submitDataToSRA/" class="article-date">
  <time datetime="2017-09-01T13:29:21.000Z" itemprop="datePublished">2017-09-01</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2017/201709_submitDataToSRA/">提交测序数据到 SRA</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>九月一号开学了，暑假作业写完了没有？但是跟我并没有关系，不开学，没作业，但是好多事情啊，，，最近提交了一些数据到SRA，每次过程总是那么曲折离奇，对一些遇到的问题进行整理填坑吧。</p>
<p>简单说，申请<strong>BioProject</strong>，申请<strong>BioSample</strong>，填写<strong>SRA_metadata</strong>，上传数据，虽然过程清晰，坑还没填完，仍在邮件NCBI，过程供参考。</p>
<h2 id="1-BioProject"><a href="#1-BioProject" class="headerlink" title="1. BioProject"></a>1. <a href="https://submit.ncbi.nlm.nih.gov/subs/bioproject/" target="_blank" rel="noopener">BioProject</a></h2><p>申请一个项目编号，填写submitter 信息，选择项目类型（Raw sequence reads），物种名称，释放时间，项目描述，BioSample和文章信息（跳过，跳过，，），核对提交</p>
<h2 id="2-BioSample"><a href="#2-BioSample" class="headerlink" title="2. BioSample"></a>2. <a href="https://submit.ncbi.nlm.nih.gov/subs/biosample/" target="_blank" rel="noopener">BioSample</a></h2><p>申请一个样品编号，submitter 信息，释放时间，样品类型，<strong>填写样品属性表格</strong>，核对提交</p>
<h2 id="3-SRA"><a href="#3-SRA" class="headerlink" title="3. SRA"></a>3. <a href="https://submit.ncbi.nlm.nih.gov/subs/sra/" target="_blank" rel="noopener">SRA</a></h2><p>主要就填写<strong>SRA_metadata</strong>，测序平台，文库类型等等，选择批量提交，填写表格注意看要求啊，除了样品名称、登录号等信息，其他属性组合每行也要是唯一的。</p>
<h2 id="4-submit-data"><a href="#4-submit-data" class="headerlink" title="4. submit data"></a>4. submit data</h2><p>看数据大小，选择合适上传方法吧，小于2G用aspera浏览器插件，大数据选择ftp或者aspera命令行。这几次都用了aspera cmd上传，霸占几乎全部网速，很快上传好</p>
<p>具体命令为：<code>ascp -i &lt;path/to/key_file&gt; -QT -l100m -k1 -d &lt;path/to/folder/containing files&gt; subasp@upload.ncbi.nlm.nih.gov:uploads/XXX@xxx.com_g3O1FgOE</code>， 其中key必须是全路径，-d接包括原始数据的文件夹（<strong>不再包含其他文件夹</strong>），<a href="mailto:XXX@xxx.com" target="_blank" rel="noopener">XXX@xxx.com</a>为注册账号邮箱</p>
<p>上传结束后，SRA中选择上传数据文件夹，，，</p>
<blockquote>
<p><em>好了，装逼结束，文件没仔细检查，传了两不完整的fastq，发邮件给ncbi沟通呢，md5值也还没有用到，不知什么情况，，，too young, too naive…</em></p>
</blockquote>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/bioinfor/">bioinfor</a>
  </div>

			</span>
			<span class="tags">
				
			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
    <article id="post-201708_object" class="post">
	<footer class="entry-meta-header">
		<span class="meta-elements date">
			<a href="/2017/201708_object/" class="article-date">
  <time datetime="2017-08-28T14:47:06.000Z" itemprop="datePublished">2017-08-28</time>
</a>
		</span>
		<span class="meta-elements author">ChengCZ</span>
		<div class="commentscount">
			
		</div>
	</footer>
	
	<header class="entry-header">
		
  
    <h1 itemprop="name" class="entry-title">
      <a class="article-title" href="/2017/201708_object/">python class 实现 fasta 文件解析</a>
    </h1>
  

	</header>
	<div class="entry-content">
		
    	<p>没有对象怎么办（object）？？？当然是new一个啊</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">Sequence</span><span class="params">(object)</span>:</span></span><br><span class="line">    <span class="comment"># 使用__init__对Sequence进行初始化，Sequence对象具有三个属性</span></span><br><span class="line">    <span class="function"><span class="keyword">def</span> <span class="title">__init__</span><span class="params">(self, name, seq, descr=<span class="string">''</span>)</span>:</span></span><br><span class="line">        self.name = name</span><br><span class="line">        self.seq = seq</span><br><span class="line">        self.descr = descr</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 给Sequence对象定义一个反向互补方法</span></span><br><span class="line">    <span class="function"><span class="keyword">def</span> <span class="title">reverse_complement</span><span class="params">(self)</span>:</span></span><br><span class="line">        RC = &#123;<span class="string">'A'</span>:<span class="string">'T'</span>,<span class="string">'T'</span>:<span class="string">'A'</span>,<span class="string">'C'</span>:<span class="string">'G'</span>,<span class="string">'G'</span>:<span class="string">'C'</span>,<span class="string">'N'</span>:<span class="string">'N'</span>,</span><br><span class="line">              <span class="string">'a'</span>:<span class="string">'t'</span>,<span class="string">'t'</span>:<span class="string">'a'</span>,<span class="string">'c'</span>:<span class="string">'g'</span>,<span class="string">'g'</span>:<span class="string">'c'</span>,<span class="string">'n'</span>:<span class="string">'n'</span>&#125;</span><br><span class="line">        <span class="keyword">return</span> Sequence(self.name, <span class="string">''</span>.join([RC[i] <span class="keyword">for</span> i <span class="keyword">in</span> self.seq[::<span class="number">-1</span>]]), self.descr)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 定义输出到文件的方法，接受文件句柄输入</span></span><br><span class="line">    <span class="function"><span class="keyword">def</span> <span class="title">write_to_fasta_file</span><span class="params">(self, file_handle)</span>:</span></span><br><span class="line">        <span class="function"><span class="keyword">def</span> <span class="title">_SeqFormat</span><span class="params">(seq, chara=<span class="number">80</span>)</span>:</span></span><br><span class="line">            tmp = <span class="string">''</span></span><br><span class="line">            <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">0</span>,len(seq),chara):</span><br><span class="line">                tmp += (seq[i:(i+chara)]+<span class="string">'\n'</span>)</span><br><span class="line">            <span class="keyword">return</span> tmp</span><br><span class="line">        file_handle.write(<span class="string">'&gt;&#123;&#125; &#123;&#125;\n'</span>.format(self.name, self.descr))</span><br><span class="line">        file_handle.write(<span class="string">'&#123;&#125;'</span>.format(_SeqFormat(self.seq)))</span><br><span class="line"></span><br><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">FastaReader</span><span class="params">(object)</span>:</span></span><br><span class="line">    <span class="function"><span class="keyword">def</span> <span class="title">__init__</span><span class="params">(self, fastafile)</span>:</span></span><br><span class="line">        self.fasta = fastafile</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 使用__iter__方法实现对fasta文件的循环解析，其中用到yeild构造了生成器</span></span><br><span class="line">    <span class="function"><span class="keyword">def</span> <span class="title">__iter__</span><span class="params">(self)</span>:</span></span><br><span class="line">        <span class="keyword">with</span> open(self.fasta) <span class="keyword">as</span> f:</span><br><span class="line">            seq = <span class="keyword">None</span></span><br><span class="line">            <span class="keyword">for</span> line <span class="keyword">in</span> f:</span><br><span class="line">                <span class="keyword">if</span> line.startswith( <span class="string">"&gt;"</span> ):</span><br><span class="line">                    <span class="keyword">if</span> seq:</span><br><span class="line">                        <span class="keyword">yield</span> Sequence( name, seq, descr )</span><br><span class="line">                    name, sep, descr = line[<span class="number">1</span>:<span class="number">-1</span>].partition(<span class="string">' '</span>)</span><br><span class="line">                    seq = <span class="string">""</span></span><br><span class="line">                <span class="keyword">else</span>:</span><br><span class="line">                    <span class="keyword">assert</span> seq <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>, <span class="string">"FASTA file does not start with '&gt;'."</span></span><br><span class="line">                    seq += line[:<span class="number">-1</span>].encode()</span><br><span class="line">        <span class="keyword">if</span> seq <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">            <span class="keyword">yield</span> Sequence( name, seq, descr )</span><br><span class="line"></span><br><span class="line"><span class="comment"># example</span></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> FastaReader(<span class="string">'ref.fa'</span>):</span><br><span class="line">    <span class="comment"># 使用</span></span><br><span class="line">    i.write_to_fasta_file(file_handle)</span><br><span class="line">    i.seq</span><br><span class="line">    i.name</span><br><span class="line">    i.descr</span><br></pre></td></tr></table></figure>
<p>懒癌发作了，就这，，，</p>

    
	</div>
	<footer class="entry-footer">
		<div class="entry-meta-footer">
			<span class="category">
				
  <div class="article-category">
    <a class="article-category-link" href="/categories/python/">python</a>
  </div>

			</span>
			<span class="tags">
				
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/fasta/">fasta</a></li></ul>

			</span>
		</div>
	</footer>
	
</article>


	<hr class="article-devider">



  
  
    <nav id="page-nav">
      <span class="page-number current">1</span><a class="page-number" href="/page/2/">2</a><a class="page-number" href="/page/3/">3</a><a class="extend next" rel="next" href="/page/2/">Next &raquo;</a>
    </nav>
  

    </div>
    <div class="mb-search">
  <form action="//google.com/search" method="get" accept-charset="utf-8">
    <input type="search" name="q" results="0" placeholder="Search">
    <input type="hidden" name="q" value="site:chengcz.github.io">
  </form>
</div>
<footer id="footer">
	<!---<h1 class="footer-blog-title">
		<a href="/">ChengCZ&#39;s Notebook</a>
	</h1>--->
	<span class="copyright">
		&copy; 2018 ChengCZ<br>
		<!---Modify from <a href="http://sanographix.github.io/tumblr/apollo/" target="_blank">Apollo</a> theme, designed by <a href="http://www.sanographix.net/" target="_blank">SANOGRAPHIX.NET</a><br>
		Powered by <a href="http://hexo.io/" target="_blank">Hexo</a>--->
	</span>
</footer>

    

<script src="//ajax.googleapis.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>


<link rel="stylesheet" href="/fancybox/jquery.fancybox.css" media="screen" type="text/css">
<script src="/fancybox/jquery.fancybox.pack.js"></script>


<script src="/js/script.js"></script>
  </div>
</body>
</html>